<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>24&nbsp; Dimensionality Reduction and Multivariate Methods – Statistics for Bioengineering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/24-high-performance-computing.html" rel="next">
<link href="../chapters/22-statistical-learning.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-485d01fc63b59abcd3ee1bf1e8e2748d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/19-anova.html">Advanced Topics</a></li><li class="breadcrumb-item"><a href="../chapters/23-dimensionality-reduction.html"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Statistics for Bioengineering</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Preface</span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Foundations of Data Science</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/01-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/02-unix-command-line.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Unix and the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/03-r-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/04-markdown-latex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Markdown and LaTeX</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/05-tidy-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Tidy Data and Data Wrangling</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/06-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Data Visualization</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Probability and Distributions</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/07-probability-foundations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Foundations of Probability</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/08-discrete-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Discrete Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/09-continuous-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Continuous Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/10-sampling-estimation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Sampling and Parameter Estimation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Inference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/11-hypothesis-testing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/12-t-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">T-Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/13-nonparametric-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Nonparametric Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/14-bootstrapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Bootstrapping</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/14b-what-are-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">What are Models?</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/15-correlation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Correlation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/16-simple-linear-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Simple Linear Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/17-residual-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Residual Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/18-statistical-power.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Statistical Power</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Advanced Topics</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/19-anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Analysis of Variance</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/20-multiple-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Multiple Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/21-glm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Generalized Linear Models</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/22-statistical-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Statistical Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/23-dimensionality-reduction.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/24-high-performance-computing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">High Performance Computing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Scientific Communication</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/25-presenting-results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Presenting Statistical Results</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A1-eugenics-history.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">The Eugenics History of Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A2-r-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">R Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A3-probability-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Common Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A4-greek-letters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Greek Letters in Mathematics and Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A5-sampling-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Sampling Distributions in Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A6-unix-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Unix Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A7-latex-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">LaTeX Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A8-keyboard-shortcuts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Keyboard Shortcuts Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A9-exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Practice Exercises</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#the-challenge-of-high-dimensional-data" id="toc-the-challenge-of-high-dimensional-data" class="nav-link active" data-scroll-target="#the-challenge-of-high-dimensional-data"><span class="header-section-number">24.1</span> The Challenge of High-Dimensional Data</a></li>
  <li><a href="#principal-component-analysis-pca" id="toc-principal-component-analysis-pca" class="nav-link" data-scroll-target="#principal-component-analysis-pca"><span class="header-section-number">24.2</span> Principal Component Analysis (PCA)</a>
  <ul class="collapse">
  <li><a href="#the-eigenanalysis-foundation" id="toc-the-eigenanalysis-foundation" class="nav-link" data-scroll-target="#the-eigenanalysis-foundation"><span class="header-section-number">24.2.1</span> The Eigenanalysis Foundation</a></li>
  <li><a href="#pca-in-r" id="toc-pca-in-r" class="nav-link" data-scroll-target="#pca-in-r"><span class="header-section-number">24.2.2</span> PCA in R</a></li>
  <li><a href="#loadings-what-variables-drive-each-pc" id="toc-loadings-what-variables-drive-each-pc" class="nav-link" data-scroll-target="#loadings-what-variables-drive-each-pc"><span class="header-section-number">24.2.3</span> Loadings: What Variables Drive Each PC?</a></li>
  <li><a href="#interpreting-pca-results" id="toc-interpreting-pca-results" class="nav-link" data-scroll-target="#interpreting-pca-results"><span class="header-section-number">24.2.4</span> Interpreting PCA Results</a></li>
  <li><a href="#biplot-visualization" id="toc-biplot-visualization" class="nav-link" data-scroll-target="#biplot-visualization"><span class="header-section-number">24.2.5</span> Biplot Visualization</a></li>
  </ul></li>
  <li><a href="#principal-coordinate-analysis-pcoa" id="toc-principal-coordinate-analysis-pcoa" class="nav-link" data-scroll-target="#principal-coordinate-analysis-pcoa"><span class="header-section-number">24.3</span> Principal Coordinate Analysis (PCoA)</a>
  <ul class="collapse">
  <li><a href="#when-to-use-pcoa-vs.-pca" id="toc-when-to-use-pcoa-vs.-pca" class="nav-link" data-scroll-target="#when-to-use-pcoa-vs.-pca"><span class="header-section-number">24.3.1</span> When to Use PCoA vs.&nbsp;PCA</a></li>
  </ul></li>
  <li><a href="#non-metric-multidimensional-scaling-nmds" id="toc-non-metric-multidimensional-scaling-nmds" class="nav-link" data-scroll-target="#non-metric-multidimensional-scaling-nmds"><span class="header-section-number">24.4</span> Non-Metric Multidimensional Scaling (NMDS)</a></li>
  <li><a href="#cluster-analysis" id="toc-cluster-analysis" class="nav-link" data-scroll-target="#cluster-analysis"><span class="header-section-number">24.5</span> Cluster Analysis</a>
  <ul class="collapse">
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering"><span class="header-section-number">24.5.1</span> Hierarchical Clustering</a></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering"><span class="header-section-number">24.5.2</span> K-Means Clustering</a></li>
  <li><a href="#choosing-the-number-of-clusters" id="toc-choosing-the-number-of-clusters" class="nav-link" data-scroll-target="#choosing-the-number-of-clusters"><span class="header-section-number">24.5.3</span> Choosing the Number of Clusters</a></li>
  </ul></li>
  <li><a href="#manova-multivariate-analysis-of-variance" id="toc-manova-multivariate-analysis-of-variance" class="nav-link" data-scroll-target="#manova-multivariate-analysis-of-variance"><span class="header-section-number">24.6</span> MANOVA: Multivariate Analysis of Variance</a>
  <ul class="collapse">
  <li><a href="#why-not-multiple-anovas" id="toc-why-not-multiple-anovas" class="nav-link" data-scroll-target="#why-not-multiple-anovas"><span class="header-section-number">24.6.1</span> Why Not Multiple ANOVAs?</a></li>
  <li><a href="#the-manova-framework" id="toc-the-manova-framework" class="nav-link" data-scroll-target="#the-manova-framework"><span class="header-section-number">24.6.2</span> The MANOVA Framework</a></li>
  <li><a href="#test-statistics" id="toc-test-statistics" class="nav-link" data-scroll-target="#test-statistics"><span class="header-section-number">24.6.3</span> Test Statistics</a></li>
  <li><a href="#manova-in-r" id="toc-manova-in-r" class="nav-link" data-scroll-target="#manova-in-r"><span class="header-section-number">24.6.4</span> MANOVA in R</a></li>
  <li><a href="#follow-up-analyses" id="toc-follow-up-analyses" class="nav-link" data-scroll-target="#follow-up-analyses"><span class="header-section-number">24.6.5</span> Follow-Up Analyses</a></li>
  <li><a href="#manova-assumptions" id="toc-manova-assumptions" class="nav-link" data-scroll-target="#manova-assumptions"><span class="header-section-number">24.6.6</span> MANOVA Assumptions</a></li>
  </ul></li>
  <li><a href="#discriminant-function-analysis-dfa" id="toc-discriminant-function-analysis-dfa" class="nav-link" data-scroll-target="#discriminant-function-analysis-dfa"><span class="header-section-number">24.7</span> Discriminant Function Analysis (DFA)</a>
  <ul class="collapse">
  <li><a href="#the-goal-of-dfa" id="toc-the-goal-of-dfa" class="nav-link" data-scroll-target="#the-goal-of-dfa"><span class="header-section-number">24.7.1</span> The Goal of DFA</a></li>
  <li><a href="#dfa-in-r" id="toc-dfa-in-r" class="nav-link" data-scroll-target="#dfa-in-r"><span class="header-section-number">24.7.2</span> DFA in R</a></li>
  <li><a href="#interpreting-dfa-output" id="toc-interpreting-dfa-output" class="nav-link" data-scroll-target="#interpreting-dfa-output"><span class="header-section-number">24.7.3</span> Interpreting DFA Output</a></li>
  <li><a href="#using-dfa-for-prediction" id="toc-using-dfa-for-prediction" class="nav-link" data-scroll-target="#using-dfa-for-prediction"><span class="header-section-number">24.7.4</span> Using DFA for Prediction</a></li>
  <li><a href="#cross-validated-classification" id="toc-cross-validated-classification" class="nav-link" data-scroll-target="#cross-validated-classification"><span class="header-section-number">24.7.5</span> Cross-Validated Classification</a></li>
  <li><a href="#dfa-for-biomarker-discovery" id="toc-dfa-for-biomarker-discovery" class="nav-link" data-scroll-target="#dfa-for-biomarker-discovery"><span class="header-section-number">24.7.6</span> DFA for Biomarker Discovery</a></li>
  </ul></li>
  <li><a href="#comparing-methods" id="toc-comparing-methods" class="nav-link" data-scroll-target="#comparing-methods"><span class="header-section-number">24.8</span> Comparing Methods</a></li>
  <li><a href="#using-ordination-scores-in-further-analyses" id="toc-using-ordination-scores-in-further-analyses" class="nav-link" data-scroll-target="#using-ordination-scores-in-further-analyses"><span class="header-section-number">24.9</span> Using Ordination Scores in Further Analyses</a></li>
  <li><a href="#practical-workflow" id="toc-practical-workflow" class="nav-link" data-scroll-target="#practical-workflow"><span class="header-section-number">24.10</span> Practical Workflow</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">24.11</span> Summary</a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="header-section-number">24.12</span> Additional Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/19-anova.html">Advanced Topics</a></li><li class="breadcrumb-item"><a href="../chapters/23-dimensionality-reduction.html"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-dimensionality-reduction" class="quarto-section-identifier"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="the-challenge-of-high-dimensional-data" class="level2" data-number="24.1">
<h2 data-number="24.1" class="anchored" data-anchor-id="the-challenge-of-high-dimensional-data"><span class="header-section-number">24.1</span> The Challenge of High-Dimensional Data</h2>
<p>Modern biology generates datasets with many variables: gene expression across thousands of genes, metabolomic profiles with hundreds of compounds, morphological measurements on many traits. When datasets have many variables, visualization becomes impossible and statistical analysis becomes complicated.</p>
<p><strong>Dimensionality reduction</strong> creates a smaller set of new variables that capture most of the information in the original data. These techniques help us:</p>
<ul>
<li>Visualize high-dimensional data in 2D or 3D</li>
<li>Identify patterns and clusters</li>
<li>Remove noise and redundancy</li>
<li>Create composite variables for downstream analysis</li>
</ul>
</section>
<section id="principal-component-analysis-pca" class="level2" data-number="24.2">
<h2 data-number="24.2" class="anchored" data-anchor-id="principal-component-analysis-pca"><span class="header-section-number">24.2</span> Principal Component Analysis (PCA)</h2>
<p><strong>Principal Component Analysis (PCA)</strong> is the most widely used dimensionality reduction technique. It finds new variables (principal components) that are linear combinations of the originals, chosen to capture maximum variance.</p>
<section id="the-eigenanalysis-foundation" class="level3" data-number="24.2.1">
<h3 data-number="24.2.1" class="anchored" data-anchor-id="the-eigenanalysis-foundation"><span class="header-section-number">24.2.1</span> The Eigenanalysis Foundation</h3>
<p>The key insight involves <strong>eigenanalysis</strong>: decomposing the covariance (or correlation) matrix of variables to find directions of maximum variation.</p>
<p>Given a covariance matrix <span class="math inline">\(\Sigma\)</span>, eigenanalysis finds:</p>
<ul>
<li><strong>Eigenvectors</strong>: The directions of the principal components</li>
<li><strong>Eigenvalues</strong>: The variance explained by each component</li>
</ul>
<p>The first principal component points in the direction of maximum variance. Each subsequent component is orthogonal (uncorrelated) and captures remaining variance in decreasing order.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../images/Images_7a.006.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</section>
<section id="pca-in-r" class="level3" data-number="24.2.2">
<h3 data-number="24.2.2" class="anchored" data-anchor-id="pca-in-r"><span class="header-section-number">24.2.2</span> PCA in R</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA on iris data</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>iris_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance explained</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris_pca)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Importance of components:
                          PC1    PC2     PC3     PC4
Standard deviation     1.7084 0.9560 0.38309 0.14393
Proportion of Variance 0.7296 0.2285 0.03669 0.00518
Cumulative Proportion  0.7296 0.9581 0.99482 1.00000</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Scree plot</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris_pca, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">main =</span> <span class="st">"Scree Plot"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># PC scores colored by species</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris_pca<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>],</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>,</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCA of Iris Data"</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="loadings-what-variables-drive-each-pc" class="level3" data-number="24.2.3">
<h3 data-number="24.2.3" class="anchored" data-anchor-id="loadings-what-variables-drive-each-pc"><span class="header-section-number">24.2.3</span> Loadings: What Variables Drive Each PC?</h3>
<p>Each principal component is defined by its <strong>loadings</strong>—the coefficients showing how much each original variable contributes:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Loadings (rotation matrix)</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>iris_pca<span class="sc">$</span>rotation</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                    PC1         PC2        PC3        PC4
Sepal.Length  0.5210659 -0.37741762  0.7195664  0.2612863
Sepal.Width  -0.2693474 -0.92329566 -0.2443818 -0.1235096
Petal.Length  0.5804131 -0.02449161 -0.1421264 -0.8014492
Petal.Width   0.5648565 -0.06694199 -0.6342727  0.5235971</code></pre>
</div>
</div>
<p>Large absolute loadings indicate that a variable strongly influences that component. The sign indicates the direction of the relationship.</p>
</section>
<section id="interpreting-pca-results" class="level3" data-number="24.2.4">
<h3 data-number="24.2.4" class="anchored" data-anchor-id="interpreting-pca-results"><span class="header-section-number">24.2.4</span> Interpreting PCA Results</h3>
<p>Key elements of PCA output:</p>
<ul>
<li><strong>Eigenvalues</strong>: Variance explained by each component (shown in scree plot)</li>
<li><strong>Proportion of variance</strong>: How much of total variance each PC captures</li>
<li><strong>Loadings</strong>: Coefficients relating original variables to PCs</li>
<li><strong>Scores</strong>: Values of the new variables for each observation</li>
</ul>
<p>The first few PCs often capture most of the meaningful variation, allowing you to reduce many variables to just 2-3 for visualization and analysis.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
How Many Components to Keep?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Common approaches:</p>
<ul>
<li>Keep components with eigenvalues &gt; 1 (Kaiser criterion)</li>
<li>Keep enough to explain 80-90% of variance</li>
<li>Look for an “elbow” in the scree plot</li>
<li>Use cross-validation if using PCs for prediction</li>
</ul>
</div>
</div>
</section>
<section id="biplot-visualization" class="level3" data-number="24.2.5">
<h3 data-number="24.2.5" class="anchored" data-anchor-id="biplot-visualization"><span class="header-section-number">24.2.5</span> Biplot Visualization</h3>
<p>A <strong>biplot</strong> shows both observations (scores) and variables (loadings) on the same plot:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(iris_pca, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"gray50"</span>, <span class="st">"red"</span>), <span class="at">cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="st">"PCA Biplot of Iris Data"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Arrows show variable loadings—their direction and length indicate how each variable relates to the principal components.</p>
</section>
</section>
<section id="principal-coordinate-analysis-pcoa" class="level2" data-number="24.3">
<h2 data-number="24.3" class="anchored" data-anchor-id="principal-coordinate-analysis-pcoa"><span class="header-section-number">24.3</span> Principal Coordinate Analysis (PCoA)</h2>
<p>While PCA uses correlations among variables, <strong>Principal Coordinate Analysis (PCoA)</strong> (also called Metric Multidimensional Scaling) starts with a dissimilarity matrix among observations. This is valuable when:</p>
<ul>
<li>You have a meaningful distance metric (e.g., genetic distances)</li>
<li>Variables are mixed types or non-numeric</li>
<li>The data are counts (e.g., microbiome data)</li>
</ul>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCoA example using Euclidean distances</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="ot">&lt;-</span> <span class="fu">dist</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>pcoa_result <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(dist_matrix, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Proportion of variance explained</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>eig_vals <span class="ot">&lt;-</span> pcoa_result<span class="sc">$</span>eig[pcoa_result<span class="sc">$</span>eig <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a>var_explained <span class="ot">&lt;-</span> eig_vals <span class="sc">/</span> <span class="fu">sum</span>(eig_vals)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Variance explained by first two axes:"</span>,</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">round</span>(<span class="fu">sum</span>(var_explained[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Variance explained by first two axes: 97.8 %</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pcoa_result<span class="sc">$</span>points,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">paste0</span>(<span class="st">"PCoA1 ("</span>, <span class="fu">round</span>(var_explained[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>),</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="fu">paste0</span>(<span class="st">"PCoA2 ("</span>, <span class="fu">round</span>(var_explained[<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>),</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCoA of Iris Data"</span>)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<section id="when-to-use-pcoa-vs.-pca" class="level3" data-number="24.3.1">
<h3 data-number="24.3.1" class="anchored" data-anchor-id="when-to-use-pcoa-vs.-pca"><span class="header-section-number">24.3.1</span> When to Use PCoA vs.&nbsp;PCA</h3>
<ul>
<li><strong>PCA</strong>: Variables are measured on a common scale; interested in variable contributions</li>
<li><strong>PCoA</strong>: Have a distance matrix; want to preserve distances among samples</li>
<li>For Euclidean distances, PCA and PCoA give equivalent results</li>
</ul>
</section>
</section>
<section id="non-metric-multidimensional-scaling-nmds" class="level2" data-number="24.4">
<h2 data-number="24.4" class="anchored" data-anchor-id="non-metric-multidimensional-scaling-nmds"><span class="header-section-number">24.4</span> Non-Metric Multidimensional Scaling (NMDS)</h2>
<p><strong>NMDS</strong> is an ordination technique that preserves rank-order of distances rather than exact distances. It’s widely used in ecology because it makes no assumptions about the data distribution.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># NMDS example</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>nmds_result <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">k =</span> <span class="dv">2</span>, <span class="at">trymax =</span> <span class="dv">100</span>, <span class="at">trace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Stress value indicates fit (&lt; 0.1 is good, &lt; 0.2 is acceptable)</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Stress:"</span>, <span class="fu">round</span>(nmds_result<span class="sc">$</span>stress, <span class="dv">3</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Stress: 0.038 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nmds_result<span class="sc">$</span>points,</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"NMDS1"</span>, <span class="at">ylab =</span> <span class="st">"NMDS2"</span>,</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"NMDS of Iris Data"</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Metric vs.&nbsp;Non-Metric Methods
</div>
</div>
<div class="callout-body-container callout-body">
<p>PCA and metric PCoA produce scores on a ratio scale—differences between scores are meaningful. These can be used directly in linear models.</p>
<p>Non-metric multidimensional scaling (NMDS) produces ordinal rankings only. NMDS scores should <strong>not</strong> be used in parametric analyses like ANOVA or regression.</p>
</div>
</div>
</section>
<section id="cluster-analysis" class="level2" data-number="24.5">
<h2 data-number="24.5" class="anchored" data-anchor-id="cluster-analysis"><span class="header-section-number">24.5</span> Cluster Analysis</h2>
<p><strong>Cluster analysis</strong> groups observations based on their similarity. Unlike PCA, which creates new continuous variables, clustering assigns observations to discrete groups.</p>
<section id="hierarchical-clustering" class="level3" data-number="24.5.1">
<h3 data-number="24.5.1" class="anchored" data-anchor-id="hierarchical-clustering"><span class="header-section-number">24.5.1</span> Hierarchical Clustering</h3>
<p><strong>Hierarchical clustering</strong> builds a tree (dendrogram) of nested clusters. At each step, it either combines the most similar observations/clusters (agglomerative) or splits the most dissimilar (divisive).</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical clustering</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>iris_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(iris_scaled), <span class="at">method =</span> <span class="st">"complete"</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc, <span class="at">labels =</span> <span class="cn">FALSE</span>, <span class="at">main =</span> <span class="st">"Hierarchical Clustering of Iris"</span>)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rect.hclust</span>(hc, <span class="at">k =</span> <span class="dv">3</span>, <span class="at">border =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
<p>Different <strong>linkage methods</strong> define how cluster distances are calculated:</p>
<ul>
<li><strong>Complete</strong>: Maximum distance between points in different clusters</li>
<li><strong>Single</strong>: Minimum distance (tends to chain)</li>
<li><strong>Average</strong>: Mean distance between all pairs</li>
<li><strong>Ward’s</strong>: Minimizes within-cluster variance</li>
</ul>
</section>
<section id="k-means-clustering" class="level3" data-number="24.5.2">
<h3 data-number="24.5.2" class="anchored" data-anchor-id="k-means-clustering"><span class="header-section-number">24.5.2</span> K-Means Clustering</h3>
<p><strong>K-means</strong> partitions data into K groups by minimizing within-cluster variance. It requires specifying K in advance.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare to true species</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Cluster =</span> km<span class="sc">$</span>cluster, <span class="at">Species =</span> iris<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>       Species
Cluster setosa versicolor virginica
      1     50          0         0
      2      0         39        14
      3      0         11        36</code></pre>
</div>
</div>
<p>The <code>nstart</code> parameter runs the algorithm multiple times with different starting points, reducing sensitivity to initialization.</p>
</section>
<section id="choosing-the-number-of-clusters" class="level3" data-number="24.5.3">
<h3 data-number="24.5.3" class="anchored" data-anchor-id="choosing-the-number-of-clusters"><span class="header-section-number">24.5.3</span> Choosing the Number of Clusters</h3>
<p>Several methods help determine the optimal number of clusters:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Elbow method: look for bend in within-cluster sum of squares</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>wss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)<span class="sc">$</span>tot.withinss</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, wss, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>, <span class="at">ylab =</span> <span class="st">"Within-cluster SS"</span>,</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Elbow Method"</span>)</span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Silhouette method</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>sil_width <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb16-14"><a href="#cb16-14" aria-hidden="true" tabindex="-1"></a>  km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb16-15"><a href="#cb16-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="fu">silhouette</span>(km<span class="sc">$</span>cluster, <span class="fu">dist</span>(iris_scaled))[, <span class="st">"sil_width"</span>])</span>
<span id="cb16-16"><a href="#cb16-16" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb16-17"><a href="#cb16-17" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, sil_width, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb16-18"><a href="#cb16-18" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>, <span class="at">ylab =</span> <span class="st">"Average Silhouette Width"</span>,</span>
<span id="cb16-19"><a href="#cb16-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Silhouette Method"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="manova-multivariate-analysis-of-variance" class="level2" data-number="24.6">
<h2 data-number="24.6" class="anchored" data-anchor-id="manova-multivariate-analysis-of-variance"><span class="header-section-number">24.6</span> MANOVA: Multivariate Analysis of Variance</h2>
<p>When you have multiple response variables and want to test for group differences, <strong>MANOVA</strong> (Multivariate Analysis of Variance) is the appropriate technique. It extends ANOVA to multiple dependent variables simultaneously.</p>
<section id="why-not-multiple-anovas" class="level3" data-number="24.6.1">
<h3 data-number="24.6.1" class="anchored" data-anchor-id="why-not-multiple-anovas"><span class="header-section-number">24.6.1</span> Why Not Multiple ANOVAs?</h3>
<p>Running separate ANOVAs on each variable:</p>
<ul>
<li>Ignores correlations among response variables</li>
<li>Inflates Type I error rate with multiple tests</li>
<li>May miss differences only apparent when variables are considered together</li>
</ul>
<p>MANOVA tests whether group centroids differ in multivariate space.</p>
</section>
<section id="the-manova-framework" class="level3" data-number="24.6.2">
<h3 data-number="24.6.2" class="anchored" data-anchor-id="the-manova-framework"><span class="header-section-number">24.6.2</span> The MANOVA Framework</h3>
<p>MANOVA decomposes the total multivariate variation:</p>
<p><span class="math display">\[\mathbf{T} = \mathbf{H} + \mathbf{E}\]</span></p>
<p>where:</p>
<ul>
<li><strong>T</strong>: Total sum of squares and cross-products matrix</li>
<li><strong>H</strong>: Hypothesis (between-groups) matrix</li>
<li><strong>E</strong>: Error (within-groups) matrix</li>
</ul>
<p>These are matrices because we have multiple response variables.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../images/Images_7a.011.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</section>
<section id="test-statistics" class="level3" data-number="24.6.3">
<h3 data-number="24.6.3" class="anchored" data-anchor-id="test-statistics"><span class="header-section-number">24.6.3</span> Test Statistics</h3>
<p>Several test statistics exist for MANOVA, each a function of the eigenvalues of <span class="math inline">\(\mathbf{HE}^{-1}\)</span>:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 45%">
<col style="width: 54%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Statistic</th>
<th style="text-align: left;">Description</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>Wilks’ Lambda (Λ)</strong></td>
<td style="text-align: left;">Product of 1/(1+λᵢ); most commonly used</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Hotelling-Lawley Trace</strong></td>
<td style="text-align: left;">Sum of eigenvalues</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Pillai’s Trace</strong></td>
<td style="text-align: left;">Sum of λᵢ/(1+λᵢ); most robust</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Roy’s Largest Root</strong></td>
<td style="text-align: left;">Maximum eigenvalue; most powerful but sensitive</td>
</tr>
</tbody>
</table>
<p><strong>Pillai’s Trace</strong> is generally recommended because it’s most robust to violations of assumptions.</p>
</section>
<section id="manova-in-r" class="level3" data-number="24.6.4">
<h3 data-number="24.6.4" class="anchored" data-anchor-id="manova-in-r"><span class="header-section-number">24.6.4</span> MANOVA in R</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># MANOVA on iris data</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>manova_model <span class="ot">&lt;-</span> <span class="fu">manova</span>(<span class="fu">cbind</span>(Sepal.Length, Sepal.Width,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>                              Petal.Length, Petal.Width) <span class="sc">~</span> Species,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">data =</span> iris)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary with different test statistics</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(manova_model, <span class="at">test =</span> <span class="st">"Pillai"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>           Df Pillai approx F num Df den Df    Pr(&gt;F)    
Species     2 1.1919   53.466      8    290 &lt; 2.2e-16 ***
Residuals 147                                            
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(manova_model, <span class="at">test =</span> <span class="st">"Wilks"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>           Df    Wilks approx F num Df den Df    Pr(&gt;F)    
Species     2 0.023439   199.15      8    288 &lt; 2.2e-16 ***
Residuals 147                                              
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
<p>The significant result tells us that species differ in their multivariate centroid—the combination of all four measurements.</p>
</section>
<section id="follow-up-analyses" class="level3" data-number="24.6.5">
<h3 data-number="24.6.5" class="anchored" data-anchor-id="follow-up-analyses"><span class="header-section-number">24.6.5</span> Follow-Up Analyses</h3>
<p>A significant MANOVA should be followed by:</p>
<ol type="1">
<li><strong>Univariate ANOVAs</strong> to see which variables differ</li>
<li><strong>Discriminant Function Analysis</strong> to understand how groups differ</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Univariate follow-ups</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">summary.aov</span>(manova_model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code> Response Sepal.Length :
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
Species       2 63.212  31.606  119.26 &lt; 2.2e-16 ***
Residuals   147 38.956   0.265                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

 Response Sepal.Width :
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
Species       2 11.345  5.6725   49.16 &lt; 2.2e-16 ***
Residuals   147 16.962  0.1154                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

 Response Petal.Length :
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
Species       2 437.10 218.551  1180.2 &lt; 2.2e-16 ***
Residuals   147  27.22   0.185                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1

 Response Petal.Width :
             Df Sum Sq Mean Sq F value    Pr(&gt;F)    
Species       2 80.413  40.207  960.01 &lt; 2.2e-16 ***
Residuals   147  6.157   0.042                      
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</section>
<section id="manova-assumptions" class="level3" data-number="24.6.6">
<h3 data-number="24.6.6" class="anchored" data-anchor-id="manova-assumptions"><span class="header-section-number">24.6.6</span> MANOVA Assumptions</h3>
<p>MANOVA assumes:</p>
<ol type="1">
<li><strong>Multivariate normality</strong> within groups</li>
<li><strong>Homogeneity of covariance matrices</strong> across groups</li>
<li><strong>Independence</strong> of observations</li>
<li><strong>No multicollinearity</strong> among response variables</li>
</ol>
<p>Test homogeneity of covariance matrices with Box’s M test (though it’s sensitive to non-normality):</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Box's M test (requires biotools package)</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="co"># library(biotools)</span></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># boxM(iris[, 1:4], iris$Species)</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="discriminant-function-analysis-dfa" class="level2" data-number="24.7">
<h2 data-number="24.7" class="anchored" data-anchor-id="discriminant-function-analysis-dfa"><span class="header-section-number">24.7</span> Discriminant Function Analysis (DFA)</h2>
<p><strong>Discriminant Function Analysis</strong> (DFA, also called Linear Discriminant Analysis or LDA) finds linear combinations of variables that best separate groups. It complements MANOVA by showing <em>how</em> groups differ.</p>
<section id="the-goal-of-dfa" class="level3" data-number="24.7.1">
<h3 data-number="24.7.1" class="anchored" data-anchor-id="the-goal-of-dfa"><span class="header-section-number">24.7.1</span> The Goal of DFA</h3>
<p>DFA finds discriminant functions—weighted combinations of original variables—that maximize separation between groups while minimizing variation within groups.</p>
<p>The first discriminant function captures the most separation, the second captures remaining separation orthogonal to the first, and so on.</p>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="../images/Images_7a.012.jpeg" class="img-fluid quarto-figure quarto-figure-center figure-img"></p>
</figure>
</div>
</section>
<section id="dfa-in-r" class="level3" data-number="24.7.2">
<h3 data-number="24.7.2" class="anchored" data-anchor-id="dfa-in-r"><span class="header-section-number">24.7.2</span> DFA in R</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Linear Discriminant Analysis</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>lda_model <span class="ot">&lt;-</span> <span class="fu">lda</span>(Species <span class="sc">~</span> ., <span class="at">data =</span> iris)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co"># View the model</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>lda_model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Call:
lda(Species ~ ., data = iris)

Prior probabilities of groups:
    setosa versicolor  virginica 
 0.3333333  0.3333333  0.3333333 

Group means:
           Sepal.Length Sepal.Width Petal.Length Petal.Width
setosa            5.006       3.428        1.462       0.246
versicolor        5.936       2.770        4.260       1.326
virginica         6.588       2.974        5.552       2.026

Coefficients of linear discriminants:
                    LD1         LD2
Sepal.Length  0.8293776 -0.02410215
Sepal.Width   1.5344731 -2.16452123
Petal.Length -2.2012117  0.93192121
Petal.Width  -2.8104603 -2.83918785

Proportion of trace:
   LD1    LD2 
0.9912 0.0088 </code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Discriminant scores</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>lda_scores <span class="ot">&lt;-</span> <span class="fu">predict</span>(lda_model)<span class="sc">$</span>x</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lda_scores,</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Discriminant Function Scores"</span>,</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"LD1"</span>, <span class="at">ylab =</span> <span class="st">"LD2"</span>)</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="interpreting-dfa-output" class="level3" data-number="24.7.3">
<h3 data-number="24.7.3" class="anchored" data-anchor-id="interpreting-dfa-output"><span class="header-section-number">24.7.3</span> Interpreting DFA Output</h3>
<p>Key components:</p>
<ul>
<li><strong>Coefficients of linear discriminants</strong>: Weights for creating discriminant scores</li>
<li><strong>Proportion of trace</strong>: Variance explained by each discriminant function</li>
<li><strong>Group means</strong>: Average score on each discriminant function for each group</li>
</ul>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Coefficients (loadings)</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>lda_model<span class="sc">$</span>scaling</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>                    LD1         LD2
Sepal.Length  0.8293776 -0.02410215
Sepal.Width   1.5344731 -2.16452123
Petal.Length -2.2012117  0.93192121
Petal.Width  -2.8104603 -2.83918785</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Proportion of separation explained</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>lda_model<span class="sc">$</span>svd<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>(lda_model<span class="sc">$</span>svd<span class="sc">^</span><span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.991212605 0.008787395</code></pre>
</div>
</div>
</section>
<section id="using-dfa-for-prediction" class="level3" data-number="24.7.4">
<h3 data-number="24.7.4" class="anchored" data-anchor-id="using-dfa-for-prediction"><span class="header-section-number">24.7.4</span> Using DFA for Prediction</h3>
<p>DFA can classify new observations into groups based on their discriminant scores:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification accuracy</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">predict</span>(lda_model)<span class="sc">$</span>class</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Predicted =</span> predictions, <span class="at">Actual =</span> iris<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            Actual
Predicted    setosa versicolor virginica
  setosa         50          0         0
  versicolor      0         48         1
  virginica       0          2        49</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification accuracy</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(predictions <span class="sc">==</span> iris<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.98</code></pre>
</div>
</div>
</section>
<section id="cross-validated-classification" class="level3" data-number="24.7.5">
<h3 data-number="24.7.5" class="anchored" data-anchor-id="cross-validated-classification"><span class="header-section-number">24.7.5</span> Cross-Validated Classification</h3>
<p>For honest estimates of classification accuracy, use leave-one-out cross-validation:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated LDA</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>lda_cv <span class="ot">&lt;-</span> <span class="fu">lda</span>(Species <span class="sc">~</span> ., <span class="at">data =</span> iris, <span class="at">CV =</span> <span class="cn">TRUE</span>)</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb35-4"><a href="#cb35-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated classification table</span></span>
<span id="cb35-5"><a href="#cb35-5" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Predicted =</span> lda_cv<span class="sc">$</span>class, <span class="at">Actual =</span> iris<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>            Actual
Predicted    setosa versicolor virginica
  setosa         50          0         0
  versicolor      0         48         1
  virginica       0          2        49</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated accuracy</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(lda_cv<span class="sc">$</span>class <span class="sc">==</span> iris<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.98</code></pre>
</div>
</div>
</section>
<section id="dfa-for-biomarker-discovery" class="level3" data-number="24.7.6">
<h3 data-number="24.7.6" class="anchored" data-anchor-id="dfa-for-biomarker-discovery"><span class="header-section-number">24.7.6</span> DFA for Biomarker Discovery</h3>
<p>DFA is valuable for identifying which variables best distinguish groups—useful in biomarker discovery:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Which variables contribute most to separation?</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>scaling_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Variable =</span> <span class="fu">rownames</span>(lda_model<span class="sc">$</span>scaling),</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">LD1 =</span> <span class="fu">abs</span>(lda_model<span class="sc">$</span>scaling[, <span class="dv">1</span>]),</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">LD2 =</span> <span class="fu">abs</span>(lda_model<span class="sc">$</span>scaling[, <span class="dv">2</span>])</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(scaling_df<span class="sc">$</span>LD1, <span class="at">names.arg =</span> scaling_df<span class="sc">$</span>Variable,</span>
<span id="cb39-9"><a href="#cb39-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Variable Contributions to LD1"</span>,</span>
<span id="cb39-10"><a href="#cb39-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Absolute Coefficient"</span>,</span>
<span id="cb39-11"><a href="#cb39-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"steelblue"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="23-dimensionality-reduction_files/figure-html/unnamed-chunk-17-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="comparing-methods" class="level2" data-number="24.8">
<h2 data-number="24.8" class="anchored" data-anchor-id="comparing-methods"><span class="header-section-number">24.8</span> Comparing Methods</h2>
<table class="caption-top table">
<colgroup>
<col style="width: 17%">
<col style="width: 15%">
<col style="width: 17%">
<col style="width: 28%">
<col style="width: 21%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Method</th>
<th style="text-align: left;">Input</th>
<th style="text-align: left;">Output</th>
<th style="text-align: left;">Supervision</th>
<th style="text-align: left;">Best For</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">PCA</td>
<td style="text-align: left;">Variables</td>
<td style="text-align: left;">Continuous scores</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">Reducing correlated variables</td>
</tr>
<tr class="even">
<td style="text-align: left;">PCoA</td>
<td style="text-align: left;">Distance matrix</td>
<td style="text-align: left;">Continuous scores</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">Preserving sample distances</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NMDS</td>
<td style="text-align: left;">Distance matrix</td>
<td style="text-align: left;">Ordinal scores</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">Ecological community data</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cluster Analysis</td>
<td style="text-align: left;">Variables or distances</td>
<td style="text-align: left;">Group assignments</td>
<td style="text-align: left;">None</td>
<td style="text-align: left;">Finding natural groupings</td>
</tr>
<tr class="odd">
<td style="text-align: left;">MANOVA</td>
<td style="text-align: left;">Variables + groups</td>
<td style="text-align: left;">Test statistics</td>
<td style="text-align: left;">Groups known</td>
<td style="text-align: left;">Testing group differences</td>
</tr>
<tr class="even">
<td style="text-align: left;">DFA</td>
<td style="text-align: left;">Variables + groups</td>
<td style="text-align: left;">Discriminant scores</td>
<td style="text-align: left;">Groups known</td>
<td style="text-align: left;">Classifying observations</td>
</tr>
</tbody>
</table>
</section>
<section id="using-ordination-scores-in-further-analyses" class="level2" data-number="24.9">
<h2 data-number="24.9" class="anchored" data-anchor-id="using-ordination-scores-in-further-analyses"><span class="header-section-number">24.9</span> Using Ordination Scores in Further Analyses</h2>
<p>PC scores and discriminant scores are legitimate new variables that can be used in downstream analysis:</p>
<ul>
<li>Regression of scores on other continuous variables</li>
<li>ANOVA comparing groups on ordination scores</li>
<li>Correlation of scores with environmental gradients</li>
</ul>
<p>This is valuable when you have many correlated variables and want to reduce dimensionality before hypothesis testing.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use PC scores in ANOVA</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>pc_scores <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC1 =</span> iris_pca<span class="sc">$</span>x[, <span class="dv">1</span>],</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC2 =</span> iris_pca<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Species =</span> iris<span class="sc">$</span>Species</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">aov</span>(PC1 <span class="sc">~</span> Species, <span class="at">data =</span> pc_scores))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>             Df Sum Sq Mean Sq F value Pr(&gt;F)    
Species       2  406.4  203.21    1051 &lt;2e-16 ***
Residuals   147   28.4    0.19                   
---
Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</code></pre>
</div>
</div>
</section>
<section id="practical-workflow" class="level2" data-number="24.10">
<h2 data-number="24.10" class="anchored" data-anchor-id="practical-workflow"><span class="header-section-number">24.10</span> Practical Workflow</h2>
<ol type="1">
<li><p><strong>Explore data</strong>: Check for outliers, missing values, scaling issues</p></li>
<li><p><strong>Standardize if needed</strong>: Especially important when variables are on different scales</p></li>
<li><p><strong>Choose appropriate method</strong>: Based on your data type and question</p></li>
<li><p><strong>Examine output</strong>: Scree plots, loadings, clustering diagnostics</p></li>
<li><p><strong>Validate</strong>: Cross-validation for classification; permutation tests for significance</p></li>
<li><p><strong>Interpret biologically</strong>: What do the patterns mean in your system?</p></li>
</ol>
</section>
<section id="summary" class="level2" data-number="24.11">
<h2 data-number="24.11" class="anchored" data-anchor-id="summary"><span class="header-section-number">24.11</span> Summary</h2>
<ul>
<li>Dimensionality reduction creates fewer variables that capture most information</li>
<li>PCA finds linear combinations maximizing variance; useful for correlated variables</li>
<li>PCoA works from distance matrices; useful for ecological and genetic data</li>
<li>Cluster analysis groups similar observations together</li>
<li>MANOVA tests whether groups differ on multiple response variables simultaneously</li>
<li>DFA finds combinations that best discriminate known groups</li>
<li>These methods can be combined: use PCA to reduce dimensions, then cluster or classify</li>
</ul>
</section>
<section id="additional-resources" class="level2" data-number="24.12">
<h2 data-number="24.12" class="anchored" data-anchor-id="additional-resources"><span class="header-section-number">24.12</span> Additional Resources</h2>
<ul>
<li><span class="citation" data-cites="james2023islr">James et al. (<a href="../references.html#ref-james2023islr" role="doc-biblioref">2023</a>)</span> - Modern treatment of dimensionality reduction and clustering</li>
<li><span class="citation" data-cites="logan2010biostatistical">Logan (<a href="../references.html#ref-logan2010biostatistical" role="doc-biblioref">2010</a>)</span> - MANOVA and DFA in biological research contexts</li>
<li>Borcard, D., Gillet, F., &amp; Legendre, P. (2018). <em>Numerical Ecology with R</em> - Comprehensive ordination methods</li>
</ul>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-james2023islr" class="csl-entry" role="listitem">
James, Gareth, Daniela Witten, Trevor Hastie, and Robert Tibshirani. 2023. <em>An Introduction to Statistical Learning with Applications in r</em>. 2nd ed. Springer. <a href="https://www.statlearning.com">https://www.statlearning.com</a>.
</div>
<div id="ref-logan2010biostatistical" class="csl-entry" role="listitem">
Logan, Murray. 2010. <em>Biostatistical Design and Analysis Using r</em>. Wiley-Blackwell.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/22-statistical-learning.html" class="pagination-link" aria-label="Statistical Learning">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Statistical Learning</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/24-high-performance-computing.html" class="pagination-link" aria-label="High Performance Computing">
        <span class="nav-page-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">High Performance Computing</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb42" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Dimensionality Reduction and Multivariate Methods {#sec-dimensionality-reduction}</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-5"><a href="#cb42-5" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-6"><a href="#cb42-6" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb42-7"><a href="#cb42-7" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb42-8"><a href="#cb42-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb42-9"><a href="#cb42-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb42-10"><a href="#cb42-10" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span>
<span id="cb42-11"><a href="#cb42-11" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-12"><a href="#cb42-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-13"><a href="#cb42-13" aria-hidden="true" tabindex="-1"></a><span class="fu">## The Challenge of High-Dimensional Data</span></span>
<span id="cb42-14"><a href="#cb42-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-15"><a href="#cb42-15" aria-hidden="true" tabindex="-1"></a>Modern biology generates datasets with many variables: gene expression across thousands of genes, metabolomic profiles with hundreds of compounds, morphological measurements on many traits. When datasets have many variables, visualization becomes impossible and statistical analysis becomes complicated.</span>
<span id="cb42-16"><a href="#cb42-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-17"><a href="#cb42-17" aria-hidden="true" tabindex="-1"></a>**Dimensionality reduction** creates a smaller set of new variables that capture most of the information in the original data. These techniques help us:</span>
<span id="cb42-18"><a href="#cb42-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-19"><a href="#cb42-19" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Visualize high-dimensional data in 2D or 3D</span>
<span id="cb42-20"><a href="#cb42-20" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Identify patterns and clusters</span>
<span id="cb42-21"><a href="#cb42-21" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Remove noise and redundancy</span>
<span id="cb42-22"><a href="#cb42-22" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Create composite variables for downstream analysis</span>
<span id="cb42-23"><a href="#cb42-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-24"><a href="#cb42-24" aria-hidden="true" tabindex="-1"></a><span class="fu">## Principal Component Analysis (PCA)</span></span>
<span id="cb42-25"><a href="#cb42-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-26"><a href="#cb42-26" aria-hidden="true" tabindex="-1"></a>**Principal Component Analysis (PCA)** is the most widely used dimensionality reduction technique. It finds new variables (principal components) that are linear combinations of the originals, chosen to capture maximum variance.</span>
<span id="cb42-27"><a href="#cb42-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-28"><a href="#cb42-28" aria-hidden="true" tabindex="-1"></a><span class="fu">### The Eigenanalysis Foundation</span></span>
<span id="cb42-29"><a href="#cb42-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-30"><a href="#cb42-30" aria-hidden="true" tabindex="-1"></a>The key insight involves **eigenanalysis**: decomposing the covariance (or correlation) matrix of variables to find directions of maximum variation.</span>
<span id="cb42-31"><a href="#cb42-31" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-32"><a href="#cb42-32" aria-hidden="true" tabindex="-1"></a>Given a covariance matrix $\Sigma$, eigenanalysis finds:</span>
<span id="cb42-33"><a href="#cb42-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-34"><a href="#cb42-34" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Eigenvectors**: The directions of the principal components</span>
<span id="cb42-35"><a href="#cb42-35" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Eigenvalues**: The variance explained by each component</span>
<span id="cb42-36"><a href="#cb42-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-37"><a href="#cb42-37" aria-hidden="true" tabindex="-1"></a>The first principal component points in the direction of maximum variance. Each subsequent component is orthogonal (uncorrelated) and captures remaining variance in decreasing order.</span>
<span id="cb42-38"><a href="#cb42-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-39"><a href="#cb42-39" aria-hidden="true" tabindex="-1"></a><span class="al">![](../images/Images_7a.006.jpeg)</span>{fig-align="center"}</span>
<span id="cb42-40"><a href="#cb42-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-41"><a href="#cb42-41" aria-hidden="true" tabindex="-1"></a><span class="fu">### PCA in R</span></span>
<span id="cb42-42"><a href="#cb42-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-45"><a href="#cb42-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-46"><a href="#cb42-46" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb42-47"><a href="#cb42-47" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb42-48"><a href="#cb42-48" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA on iris data</span></span>
<span id="cb42-49"><a href="#cb42-49" aria-hidden="true" tabindex="-1"></a>iris_pca <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-50"><a href="#cb42-50" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-51"><a href="#cb42-51" aria-hidden="true" tabindex="-1"></a><span class="co"># Variance explained</span></span>
<span id="cb42-52"><a href="#cb42-52" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(iris_pca)</span>
<span id="cb42-53"><a href="#cb42-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-54"><a href="#cb42-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Scree plot</span></span>
<span id="cb42-55"><a href="#cb42-55" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb42-56"><a href="#cb42-56" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris_pca, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">main =</span> <span class="st">"Scree Plot"</span>)</span>
<span id="cb42-57"><a href="#cb42-57" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-58"><a href="#cb42-58" aria-hidden="true" tabindex="-1"></a><span class="co"># PC scores colored by species</span></span>
<span id="cb42-59"><a href="#cb42-59" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(iris_pca<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>],</span>
<span id="cb42-60"><a href="#cb42-60" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb42-61"><a href="#cb42-61" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>,</span>
<span id="cb42-62"><a href="#cb42-62" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCA of Iris Data"</span>)</span>
<span id="cb42-63"><a href="#cb42-63" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb42-64"><a href="#cb42-64" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb42-65"><a href="#cb42-65" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-66"><a href="#cb42-66" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-67"><a href="#cb42-67" aria-hidden="true" tabindex="-1"></a><span class="fu">### Loadings: What Variables Drive Each PC?</span></span>
<span id="cb42-68"><a href="#cb42-68" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-69"><a href="#cb42-69" aria-hidden="true" tabindex="-1"></a>Each principal component is defined by its **loadings**—the coefficients showing how much each original variable contributes:</span>
<span id="cb42-70"><a href="#cb42-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-73"><a href="#cb42-73" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-74"><a href="#cb42-74" aria-hidden="true" tabindex="-1"></a><span class="co"># Loadings (rotation matrix)</span></span>
<span id="cb42-75"><a href="#cb42-75" aria-hidden="true" tabindex="-1"></a>iris_pca<span class="sc">$</span>rotation</span>
<span id="cb42-76"><a href="#cb42-76" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-77"><a href="#cb42-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-78"><a href="#cb42-78" aria-hidden="true" tabindex="-1"></a>Large absolute loadings indicate that a variable strongly influences that component. The sign indicates the direction of the relationship.</span>
<span id="cb42-79"><a href="#cb42-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-80"><a href="#cb42-80" aria-hidden="true" tabindex="-1"></a><span class="fu">### Interpreting PCA Results</span></span>
<span id="cb42-81"><a href="#cb42-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-82"><a href="#cb42-82" aria-hidden="true" tabindex="-1"></a>Key elements of PCA output:</span>
<span id="cb42-83"><a href="#cb42-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-84"><a href="#cb42-84" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Eigenvalues**: Variance explained by each component (shown in scree plot)</span>
<span id="cb42-85"><a href="#cb42-85" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Proportion of variance**: How much of total variance each PC captures</span>
<span id="cb42-86"><a href="#cb42-86" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Loadings**: Coefficients relating original variables to PCs</span>
<span id="cb42-87"><a href="#cb42-87" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Scores**: Values of the new variables for each observation</span>
<span id="cb42-88"><a href="#cb42-88" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-89"><a href="#cb42-89" aria-hidden="true" tabindex="-1"></a>The first few PCs often capture most of the meaningful variation, allowing you to reduce many variables to just 2-3 for visualization and analysis.</span>
<span id="cb42-90"><a href="#cb42-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-91"><a href="#cb42-91" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip}</span>
<span id="cb42-92"><a href="#cb42-92" aria-hidden="true" tabindex="-1"></a><span class="fu">## How Many Components to Keep?</span></span>
<span id="cb42-93"><a href="#cb42-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-94"><a href="#cb42-94" aria-hidden="true" tabindex="-1"></a>Common approaches:</span>
<span id="cb42-95"><a href="#cb42-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-96"><a href="#cb42-96" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Keep components with eigenvalues &gt; 1 (Kaiser criterion)</span>
<span id="cb42-97"><a href="#cb42-97" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Keep enough to explain 80-90% of variance</span>
<span id="cb42-98"><a href="#cb42-98" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Look for an "elbow" in the scree plot</span>
<span id="cb42-99"><a href="#cb42-99" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Use cross-validation if using PCs for prediction</span>
<span id="cb42-100"><a href="#cb42-100" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb42-101"><a href="#cb42-101" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-102"><a href="#cb42-102" aria-hidden="true" tabindex="-1"></a><span class="fu">### Biplot Visualization</span></span>
<span id="cb42-103"><a href="#cb42-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-104"><a href="#cb42-104" aria-hidden="true" tabindex="-1"></a>A **biplot** shows both observations (scores) and variables (loadings) on the same plot:</span>
<span id="cb42-105"><a href="#cb42-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-108"><a href="#cb42-108" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-109"><a href="#cb42-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb42-110"><a href="#cb42-110" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb42-111"><a href="#cb42-111" aria-hidden="true" tabindex="-1"></a><span class="fu">biplot</span>(iris_pca, <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"gray50"</span>, <span class="st">"red"</span>), <span class="at">cex =</span> <span class="fl">0.7</span>,</span>
<span id="cb42-112"><a href="#cb42-112" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="st">"PCA Biplot of Iris Data"</span>)</span>
<span id="cb42-113"><a href="#cb42-113" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-114"><a href="#cb42-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-115"><a href="#cb42-115" aria-hidden="true" tabindex="-1"></a>Arrows show variable loadings—their direction and length indicate how each variable relates to the principal components.</span>
<span id="cb42-116"><a href="#cb42-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-117"><a href="#cb42-117" aria-hidden="true" tabindex="-1"></a><span class="fu">## Principal Coordinate Analysis (PCoA)</span></span>
<span id="cb42-118"><a href="#cb42-118" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-119"><a href="#cb42-119" aria-hidden="true" tabindex="-1"></a>While PCA uses correlations among variables, **Principal Coordinate Analysis (PCoA)** (also called Metric Multidimensional Scaling) starts with a dissimilarity matrix among observations. This is valuable when:</span>
<span id="cb42-120"><a href="#cb42-120" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-121"><a href="#cb42-121" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You have a meaningful distance metric (e.g., genetic distances)</span>
<span id="cb42-122"><a href="#cb42-122" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Variables are mixed types or non-numeric</span>
<span id="cb42-123"><a href="#cb42-123" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The data are counts (e.g., microbiome data)</span>
<span id="cb42-124"><a href="#cb42-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-127"><a href="#cb42-127" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-128"><a href="#cb42-128" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb42-129"><a href="#cb42-129" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb42-130"><a href="#cb42-130" aria-hidden="true" tabindex="-1"></a><span class="co"># PCoA example using Euclidean distances</span></span>
<span id="cb42-131"><a href="#cb42-131" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="ot">&lt;-</span> <span class="fu">dist</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb42-132"><a href="#cb42-132" aria-hidden="true" tabindex="-1"></a>pcoa_result <span class="ot">&lt;-</span> <span class="fu">cmdscale</span>(dist_matrix, <span class="at">k =</span> <span class="dv">2</span>, <span class="at">eig =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-133"><a href="#cb42-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-134"><a href="#cb42-134" aria-hidden="true" tabindex="-1"></a><span class="co"># Proportion of variance explained</span></span>
<span id="cb42-135"><a href="#cb42-135" aria-hidden="true" tabindex="-1"></a>eig_vals <span class="ot">&lt;-</span> pcoa_result<span class="sc">$</span>eig[pcoa_result<span class="sc">$</span>eig <span class="sc">&gt;</span> <span class="dv">0</span>]</span>
<span id="cb42-136"><a href="#cb42-136" aria-hidden="true" tabindex="-1"></a>var_explained <span class="ot">&lt;-</span> eig_vals <span class="sc">/</span> <span class="fu">sum</span>(eig_vals)</span>
<span id="cb42-137"><a href="#cb42-137" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Variance explained by first two axes:"</span>,</span>
<span id="cb42-138"><a href="#cb42-138" aria-hidden="true" tabindex="-1"></a>    <span class="fu">round</span>(<span class="fu">sum</span>(var_explained[<span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]) <span class="sc">*</span> <span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb42-139"><a href="#cb42-139" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-140"><a href="#cb42-140" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb42-141"><a href="#cb42-141" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pcoa_result<span class="sc">$</span>points,</span>
<span id="cb42-142"><a href="#cb42-142" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb42-143"><a href="#cb42-143" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb42-144"><a href="#cb42-144" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="fu">paste0</span>(<span class="st">"PCoA1 ("</span>, <span class="fu">round</span>(var_explained[<span class="dv">1</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>),</span>
<span id="cb42-145"><a href="#cb42-145" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="fu">paste0</span>(<span class="st">"PCoA2 ("</span>, <span class="fu">round</span>(var_explained[<span class="dv">2</span>]<span class="sc">*</span><span class="dv">100</span>, <span class="dv">1</span>), <span class="st">"%)"</span>),</span>
<span id="cb42-146"><a href="#cb42-146" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCoA of Iris Data"</span>)</span>
<span id="cb42-147"><a href="#cb42-147" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb42-148"><a href="#cb42-148" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb42-149"><a href="#cb42-149" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-150"><a href="#cb42-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-151"><a href="#cb42-151" aria-hidden="true" tabindex="-1"></a><span class="fu">### When to Use PCoA vs. PCA</span></span>
<span id="cb42-152"><a href="#cb42-152" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-153"><a href="#cb42-153" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**PCA**: Variables are measured on a common scale; interested in variable contributions</span>
<span id="cb42-154"><a href="#cb42-154" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**PCoA**: Have a distance matrix; want to preserve distances among samples</span>
<span id="cb42-155"><a href="#cb42-155" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>For Euclidean distances, PCA and PCoA give equivalent results</span>
<span id="cb42-156"><a href="#cb42-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-157"><a href="#cb42-157" aria-hidden="true" tabindex="-1"></a><span class="fu">## Non-Metric Multidimensional Scaling (NMDS)</span></span>
<span id="cb42-158"><a href="#cb42-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-159"><a href="#cb42-159" aria-hidden="true" tabindex="-1"></a>**NMDS** is an ordination technique that preserves rank-order of distances rather than exact distances. It's widely used in ecology because it makes no assumptions about the data distribution.</span>
<span id="cb42-160"><a href="#cb42-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-163"><a href="#cb42-163" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-164"><a href="#cb42-164" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb42-165"><a href="#cb42-165" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb42-166"><a href="#cb42-166" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb42-167"><a href="#cb42-167" aria-hidden="true" tabindex="-1"></a><span class="co"># NMDS example</span></span>
<span id="cb42-168"><a href="#cb42-168" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(vegan)</span>
<span id="cb42-169"><a href="#cb42-169" aria-hidden="true" tabindex="-1"></a>nmds_result <span class="ot">&lt;-</span> <span class="fu">metaMDS</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>], <span class="at">k =</span> <span class="dv">2</span>, <span class="at">trymax =</span> <span class="dv">100</span>, <span class="at">trace =</span> <span class="cn">FALSE</span>)</span>
<span id="cb42-170"><a href="#cb42-170" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-171"><a href="#cb42-171" aria-hidden="true" tabindex="-1"></a><span class="co"># Stress value indicates fit (&lt; 0.1 is good, &lt; 0.2 is acceptable)</span></span>
<span id="cb42-172"><a href="#cb42-172" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Stress:"</span>, <span class="fu">round</span>(nmds_result<span class="sc">$</span>stress, <span class="dv">3</span>), <span class="st">"</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb42-173"><a href="#cb42-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-174"><a href="#cb42-174" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(nmds_result<span class="sc">$</span>points,</span>
<span id="cb42-175"><a href="#cb42-175" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb42-176"><a href="#cb42-176" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">xlab =</span> <span class="st">"NMDS1"</span>, <span class="at">ylab =</span> <span class="st">"NMDS2"</span>,</span>
<span id="cb42-177"><a href="#cb42-177" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"NMDS of Iris Data"</span>)</span>
<span id="cb42-178"><a href="#cb42-178" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb42-179"><a href="#cb42-179" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb42-180"><a href="#cb42-180" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-181"><a href="#cb42-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-182"><a href="#cb42-182" aria-hidden="true" tabindex="-1"></a>::: {.callout-warning}</span>
<span id="cb42-183"><a href="#cb42-183" aria-hidden="true" tabindex="-1"></a><span class="fu">## Metric vs. Non-Metric Methods</span></span>
<span id="cb42-184"><a href="#cb42-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-185"><a href="#cb42-185" aria-hidden="true" tabindex="-1"></a>PCA and metric PCoA produce scores on a ratio scale—differences between scores are meaningful. These can be used directly in linear models.</span>
<span id="cb42-186"><a href="#cb42-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-187"><a href="#cb42-187" aria-hidden="true" tabindex="-1"></a>Non-metric multidimensional scaling (NMDS) produces ordinal rankings only. NMDS scores should **not** be used in parametric analyses like ANOVA or regression.</span>
<span id="cb42-188"><a href="#cb42-188" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb42-189"><a href="#cb42-189" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-190"><a href="#cb42-190" aria-hidden="true" tabindex="-1"></a><span class="fu">## Cluster Analysis</span></span>
<span id="cb42-191"><a href="#cb42-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-192"><a href="#cb42-192" aria-hidden="true" tabindex="-1"></a>**Cluster analysis** groups observations based on their similarity. Unlike PCA, which creates new continuous variables, clustering assigns observations to discrete groups.</span>
<span id="cb42-193"><a href="#cb42-193" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-194"><a href="#cb42-194" aria-hidden="true" tabindex="-1"></a><span class="fu">### Hierarchical Clustering</span></span>
<span id="cb42-195"><a href="#cb42-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-196"><a href="#cb42-196" aria-hidden="true" tabindex="-1"></a>**Hierarchical clustering** builds a tree (dendrogram) of nested clusters. At each step, it either combines the most similar observations/clusters (agglomerative) or splits the most dissimilar (divisive).</span>
<span id="cb42-197"><a href="#cb42-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-200"><a href="#cb42-200" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-201"><a href="#cb42-201" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb42-202"><a href="#cb42-202" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb42-203"><a href="#cb42-203" aria-hidden="true" tabindex="-1"></a><span class="co"># Hierarchical clustering</span></span>
<span id="cb42-204"><a href="#cb42-204" aria-hidden="true" tabindex="-1"></a>iris_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(iris[, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>])</span>
<span id="cb42-205"><a href="#cb42-205" aria-hidden="true" tabindex="-1"></a>hc <span class="ot">&lt;-</span> <span class="fu">hclust</span>(<span class="fu">dist</span>(iris_scaled), <span class="at">method =</span> <span class="st">"complete"</span>)</span>
<span id="cb42-206"><a href="#cb42-206" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(hc, <span class="at">labels =</span> <span class="cn">FALSE</span>, <span class="at">main =</span> <span class="st">"Hierarchical Clustering of Iris"</span>)</span>
<span id="cb42-207"><a href="#cb42-207" aria-hidden="true" tabindex="-1"></a><span class="fu">rect.hclust</span>(hc, <span class="at">k =</span> <span class="dv">3</span>, <span class="at">border =</span> <span class="st">"red"</span>)</span>
<span id="cb42-208"><a href="#cb42-208" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-209"><a href="#cb42-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-210"><a href="#cb42-210" aria-hidden="true" tabindex="-1"></a>Different **linkage methods** define how cluster distances are calculated:</span>
<span id="cb42-211"><a href="#cb42-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-212"><a href="#cb42-212" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Complete**: Maximum distance between points in different clusters</span>
<span id="cb42-213"><a href="#cb42-213" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Single**: Minimum distance (tends to chain)</span>
<span id="cb42-214"><a href="#cb42-214" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Average**: Mean distance between all pairs</span>
<span id="cb42-215"><a href="#cb42-215" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Ward's**: Minimizes within-cluster variance</span>
<span id="cb42-216"><a href="#cb42-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-217"><a href="#cb42-217" aria-hidden="true" tabindex="-1"></a><span class="fu">### K-Means Clustering</span></span>
<span id="cb42-218"><a href="#cb42-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-219"><a href="#cb42-219" aria-hidden="true" tabindex="-1"></a>**K-means** partitions data into K groups by minimizing within-cluster variance. It requires specifying K in advance.</span>
<span id="cb42-220"><a href="#cb42-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-223"><a href="#cb42-223" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-224"><a href="#cb42-224" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb42-225"><a href="#cb42-225" aria-hidden="true" tabindex="-1"></a>km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb42-226"><a href="#cb42-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-227"><a href="#cb42-227" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare to true species</span></span>
<span id="cb42-228"><a href="#cb42-228" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Cluster =</span> km<span class="sc">$</span>cluster, <span class="at">Species =</span> iris<span class="sc">$</span>Species)</span>
<span id="cb42-229"><a href="#cb42-229" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-230"><a href="#cb42-230" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-231"><a href="#cb42-231" aria-hidden="true" tabindex="-1"></a>The <span class="in">`nstart`</span> parameter runs the algorithm multiple times with different starting points, reducing sensitivity to initialization.</span>
<span id="cb42-232"><a href="#cb42-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-233"><a href="#cb42-233" aria-hidden="true" tabindex="-1"></a><span class="fu">### Choosing the Number of Clusters</span></span>
<span id="cb42-234"><a href="#cb42-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-235"><a href="#cb42-235" aria-hidden="true" tabindex="-1"></a>Several methods help determine the optimal number of clusters:</span>
<span id="cb42-236"><a href="#cb42-236" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-239"><a href="#cb42-239" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-240"><a href="#cb42-240" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb42-241"><a href="#cb42-241" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb42-242"><a href="#cb42-242" aria-hidden="true" tabindex="-1"></a><span class="co"># Elbow method: look for bend in within-cluster sum of squares</span></span>
<span id="cb42-243"><a href="#cb42-243" aria-hidden="true" tabindex="-1"></a>wss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb42-244"><a href="#cb42-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)<span class="sc">$</span>tot.withinss</span>
<span id="cb42-245"><a href="#cb42-245" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-246"><a href="#cb42-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-247"><a href="#cb42-247" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb42-248"><a href="#cb42-248" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, wss, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb42-249"><a href="#cb42-249" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>, <span class="at">ylab =</span> <span class="st">"Within-cluster SS"</span>,</span>
<span id="cb42-250"><a href="#cb42-250" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Elbow Method"</span>)</span>
<span id="cb42-251"><a href="#cb42-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-252"><a href="#cb42-252" aria-hidden="true" tabindex="-1"></a><span class="co"># Silhouette method</span></span>
<span id="cb42-253"><a href="#cb42-253" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb42-254"><a href="#cb42-254" aria-hidden="true" tabindex="-1"></a>sil_width <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb42-255"><a href="#cb42-255" aria-hidden="true" tabindex="-1"></a>  km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(iris_scaled, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb42-256"><a href="#cb42-256" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(<span class="fu">silhouette</span>(km<span class="sc">$</span>cluster, <span class="fu">dist</span>(iris_scaled))[, <span class="st">"sil_width"</span>])</span>
<span id="cb42-257"><a href="#cb42-257" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb42-258"><a href="#cb42-258" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, sil_width, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb42-259"><a href="#cb42-259" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>, <span class="at">ylab =</span> <span class="st">"Average Silhouette Width"</span>,</span>
<span id="cb42-260"><a href="#cb42-260" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Silhouette Method"</span>)</span>
<span id="cb42-261"><a href="#cb42-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-262"><a href="#cb42-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-263"><a href="#cb42-263" aria-hidden="true" tabindex="-1"></a><span class="fu">## MANOVA: Multivariate Analysis of Variance</span></span>
<span id="cb42-264"><a href="#cb42-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-265"><a href="#cb42-265" aria-hidden="true" tabindex="-1"></a>When you have multiple response variables and want to test for group differences, **MANOVA** (Multivariate Analysis of Variance) is the appropriate technique. It extends ANOVA to multiple dependent variables simultaneously.</span>
<span id="cb42-266"><a href="#cb42-266" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-267"><a href="#cb42-267" aria-hidden="true" tabindex="-1"></a><span class="fu">### Why Not Multiple ANOVAs?</span></span>
<span id="cb42-268"><a href="#cb42-268" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-269"><a href="#cb42-269" aria-hidden="true" tabindex="-1"></a>Running separate ANOVAs on each variable:</span>
<span id="cb42-270"><a href="#cb42-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-271"><a href="#cb42-271" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Ignores correlations among response variables</span>
<span id="cb42-272"><a href="#cb42-272" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Inflates Type I error rate with multiple tests</span>
<span id="cb42-273"><a href="#cb42-273" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>May miss differences only apparent when variables are considered together</span>
<span id="cb42-274"><a href="#cb42-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-275"><a href="#cb42-275" aria-hidden="true" tabindex="-1"></a>MANOVA tests whether group centroids differ in multivariate space.</span>
<span id="cb42-276"><a href="#cb42-276" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-277"><a href="#cb42-277" aria-hidden="true" tabindex="-1"></a><span class="fu">### The MANOVA Framework</span></span>
<span id="cb42-278"><a href="#cb42-278" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-279"><a href="#cb42-279" aria-hidden="true" tabindex="-1"></a>MANOVA decomposes the total multivariate variation:</span>
<span id="cb42-280"><a href="#cb42-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-281"><a href="#cb42-281" aria-hidden="true" tabindex="-1"></a>$$\mathbf{T} = \mathbf{H} + \mathbf{E}$$</span>
<span id="cb42-282"><a href="#cb42-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-283"><a href="#cb42-283" aria-hidden="true" tabindex="-1"></a>where:</span>
<span id="cb42-284"><a href="#cb42-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-285"><a href="#cb42-285" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**T**: Total sum of squares and cross-products matrix</span>
<span id="cb42-286"><a href="#cb42-286" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**H**: Hypothesis (between-groups) matrix</span>
<span id="cb42-287"><a href="#cb42-287" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**E**: Error (within-groups) matrix</span>
<span id="cb42-288"><a href="#cb42-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-289"><a href="#cb42-289" aria-hidden="true" tabindex="-1"></a>These are matrices because we have multiple response variables.</span>
<span id="cb42-290"><a href="#cb42-290" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-291"><a href="#cb42-291" aria-hidden="true" tabindex="-1"></a><span class="al">![](../images/Images_7a.011.jpeg)</span>{fig-align="center"}</span>
<span id="cb42-292"><a href="#cb42-292" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-293"><a href="#cb42-293" aria-hidden="true" tabindex="-1"></a><span class="fu">### Test Statistics</span></span>
<span id="cb42-294"><a href="#cb42-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-295"><a href="#cb42-295" aria-hidden="true" tabindex="-1"></a>Several test statistics exist for MANOVA, each a function of the eigenvalues of $\mathbf{HE}^{-1}$:</span>
<span id="cb42-296"><a href="#cb42-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-297"><a href="#cb42-297" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Statistic <span class="pp">|</span> Description <span class="pp">|</span></span>
<span id="cb42-298"><a href="#cb42-298" aria-hidden="true" tabindex="-1"></a><span class="pp">|:----------|:------------|</span></span>
<span id="cb42-299"><a href="#cb42-299" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Wilks' Lambda (Λ)** <span class="pp">|</span> Product of 1/(1+λᵢ); most commonly used <span class="pp">|</span></span>
<span id="cb42-300"><a href="#cb42-300" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Hotelling-Lawley Trace** <span class="pp">|</span> Sum of eigenvalues <span class="pp">|</span></span>
<span id="cb42-301"><a href="#cb42-301" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Pillai's Trace** <span class="pp">|</span> Sum of λᵢ/(1+λᵢ); most robust <span class="pp">|</span></span>
<span id="cb42-302"><a href="#cb42-302" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Roy's Largest Root** <span class="pp">|</span> Maximum eigenvalue; most powerful but sensitive <span class="pp">|</span></span>
<span id="cb42-303"><a href="#cb42-303" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-304"><a href="#cb42-304" aria-hidden="true" tabindex="-1"></a>**Pillai's Trace** is generally recommended because it's most robust to violations of assumptions.</span>
<span id="cb42-305"><a href="#cb42-305" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-306"><a href="#cb42-306" aria-hidden="true" tabindex="-1"></a><span class="fu">### MANOVA in R</span></span>
<span id="cb42-307"><a href="#cb42-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-310"><a href="#cb42-310" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-311"><a href="#cb42-311" aria-hidden="true" tabindex="-1"></a><span class="co"># MANOVA on iris data</span></span>
<span id="cb42-312"><a href="#cb42-312" aria-hidden="true" tabindex="-1"></a>manova_model <span class="ot">&lt;-</span> <span class="fu">manova</span>(<span class="fu">cbind</span>(Sepal.Length, Sepal.Width,</span>
<span id="cb42-313"><a href="#cb42-313" aria-hidden="true" tabindex="-1"></a>                              Petal.Length, Petal.Width) <span class="sc">~</span> Species,</span>
<span id="cb42-314"><a href="#cb42-314" aria-hidden="true" tabindex="-1"></a>                       <span class="at">data =</span> iris)</span>
<span id="cb42-315"><a href="#cb42-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-316"><a href="#cb42-316" aria-hidden="true" tabindex="-1"></a><span class="co"># Summary with different test statistics</span></span>
<span id="cb42-317"><a href="#cb42-317" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(manova_model, <span class="at">test =</span> <span class="st">"Pillai"</span>)</span>
<span id="cb42-318"><a href="#cb42-318" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(manova_model, <span class="at">test =</span> <span class="st">"Wilks"</span>)</span>
<span id="cb42-319"><a href="#cb42-319" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-320"><a href="#cb42-320" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-321"><a href="#cb42-321" aria-hidden="true" tabindex="-1"></a>The significant result tells us that species differ in their multivariate centroid—the combination of all four measurements.</span>
<span id="cb42-322"><a href="#cb42-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-323"><a href="#cb42-323" aria-hidden="true" tabindex="-1"></a><span class="fu">### Follow-Up Analyses</span></span>
<span id="cb42-324"><a href="#cb42-324" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-325"><a href="#cb42-325" aria-hidden="true" tabindex="-1"></a>A significant MANOVA should be followed by:</span>
<span id="cb42-326"><a href="#cb42-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-327"><a href="#cb42-327" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Univariate ANOVAs** to see which variables differ</span>
<span id="cb42-328"><a href="#cb42-328" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Discriminant Function Analysis** to understand how groups differ</span>
<span id="cb42-329"><a href="#cb42-329" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-332"><a href="#cb42-332" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-333"><a href="#cb42-333" aria-hidden="true" tabindex="-1"></a><span class="co"># Univariate follow-ups</span></span>
<span id="cb42-334"><a href="#cb42-334" aria-hidden="true" tabindex="-1"></a><span class="fu">summary.aov</span>(manova_model)</span>
<span id="cb42-335"><a href="#cb42-335" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-336"><a href="#cb42-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-337"><a href="#cb42-337" aria-hidden="true" tabindex="-1"></a><span class="fu">### MANOVA Assumptions</span></span>
<span id="cb42-338"><a href="#cb42-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-339"><a href="#cb42-339" aria-hidden="true" tabindex="-1"></a>MANOVA assumes:</span>
<span id="cb42-340"><a href="#cb42-340" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-341"><a href="#cb42-341" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Multivariate normality** within groups</span>
<span id="cb42-342"><a href="#cb42-342" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Homogeneity of covariance matrices** across groups</span>
<span id="cb42-343"><a href="#cb42-343" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Independence** of observations</span>
<span id="cb42-344"><a href="#cb42-344" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**No multicollinearity** among response variables</span>
<span id="cb42-345"><a href="#cb42-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-346"><a href="#cb42-346" aria-hidden="true" tabindex="-1"></a>Test homogeneity of covariance matrices with Box's M test (though it's sensitive to non-normality):</span>
<span id="cb42-347"><a href="#cb42-347" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-350"><a href="#cb42-350" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-351"><a href="#cb42-351" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb42-352"><a href="#cb42-352" aria-hidden="true" tabindex="-1"></a><span class="co"># Box's M test (requires biotools package)</span></span>
<span id="cb42-353"><a href="#cb42-353" aria-hidden="true" tabindex="-1"></a><span class="co"># library(biotools)</span></span>
<span id="cb42-354"><a href="#cb42-354" aria-hidden="true" tabindex="-1"></a><span class="co"># boxM(iris[, 1:4], iris$Species)</span></span>
<span id="cb42-355"><a href="#cb42-355" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-356"><a href="#cb42-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-357"><a href="#cb42-357" aria-hidden="true" tabindex="-1"></a><span class="fu">## Discriminant Function Analysis (DFA)</span></span>
<span id="cb42-358"><a href="#cb42-358" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-359"><a href="#cb42-359" aria-hidden="true" tabindex="-1"></a>**Discriminant Function Analysis** (DFA, also called Linear Discriminant Analysis or LDA) finds linear combinations of variables that best separate groups. It complements MANOVA by showing *how* groups differ.</span>
<span id="cb42-360"><a href="#cb42-360" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-361"><a href="#cb42-361" aria-hidden="true" tabindex="-1"></a><span class="fu">### The Goal of DFA</span></span>
<span id="cb42-362"><a href="#cb42-362" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-363"><a href="#cb42-363" aria-hidden="true" tabindex="-1"></a>DFA finds discriminant functions—weighted combinations of original variables—that maximize separation between groups while minimizing variation within groups.</span>
<span id="cb42-364"><a href="#cb42-364" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-365"><a href="#cb42-365" aria-hidden="true" tabindex="-1"></a>The first discriminant function captures the most separation, the second captures remaining separation orthogonal to the first, and so on.</span>
<span id="cb42-366"><a href="#cb42-366" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-367"><a href="#cb42-367" aria-hidden="true" tabindex="-1"></a><span class="al">![](../images/Images_7a.012.jpeg)</span>{fig-align="center"}</span>
<span id="cb42-368"><a href="#cb42-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-369"><a href="#cb42-369" aria-hidden="true" tabindex="-1"></a><span class="fu">### DFA in R</span></span>
<span id="cb42-370"><a href="#cb42-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-373"><a href="#cb42-373" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-374"><a href="#cb42-374" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb42-375"><a href="#cb42-375" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb42-376"><a href="#cb42-376" aria-hidden="true" tabindex="-1"></a><span class="co"># Linear Discriminant Analysis</span></span>
<span id="cb42-377"><a href="#cb42-377" aria-hidden="true" tabindex="-1"></a>lda_model <span class="ot">&lt;-</span> <span class="fu">lda</span>(Species <span class="sc">~</span> ., <span class="at">data =</span> iris)</span>
<span id="cb42-378"><a href="#cb42-378" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-379"><a href="#cb42-379" aria-hidden="true" tabindex="-1"></a><span class="co"># View the model</span></span>
<span id="cb42-380"><a href="#cb42-380" aria-hidden="true" tabindex="-1"></a>lda_model</span>
<span id="cb42-381"><a href="#cb42-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-382"><a href="#cb42-382" aria-hidden="true" tabindex="-1"></a><span class="co"># Discriminant scores</span></span>
<span id="cb42-383"><a href="#cb42-383" aria-hidden="true" tabindex="-1"></a>lda_scores <span class="ot">&lt;-</span> <span class="fu">predict</span>(lda_model)<span class="sc">$</span>x</span>
<span id="cb42-384"><a href="#cb42-384" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-385"><a href="#cb42-385" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb42-386"><a href="#cb42-386" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(lda_scores,</span>
<span id="cb42-387"><a href="#cb42-387" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>)[iris<span class="sc">$</span>Species],</span>
<span id="cb42-388"><a href="#cb42-388" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb42-389"><a href="#cb42-389" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Discriminant Function Scores"</span>,</span>
<span id="cb42-390"><a href="#cb42-390" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"LD1"</span>, <span class="at">ylab =</span> <span class="st">"LD2"</span>)</span>
<span id="cb42-391"><a href="#cb42-391" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="fu">levels</span>(iris<span class="sc">$</span>Species),</span>
<span id="cb42-392"><a href="#cb42-392" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="st">"red"</span>, <span class="st">"green"</span>, <span class="st">"blue"</span>), <span class="at">pch =</span> <span class="dv">19</span>)</span>
<span id="cb42-393"><a href="#cb42-393" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-394"><a href="#cb42-394" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-395"><a href="#cb42-395" aria-hidden="true" tabindex="-1"></a><span class="fu">### Interpreting DFA Output</span></span>
<span id="cb42-396"><a href="#cb42-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-397"><a href="#cb42-397" aria-hidden="true" tabindex="-1"></a>Key components:</span>
<span id="cb42-398"><a href="#cb42-398" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-399"><a href="#cb42-399" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Coefficients of linear discriminants**: Weights for creating discriminant scores</span>
<span id="cb42-400"><a href="#cb42-400" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Proportion of trace**: Variance explained by each discriminant function</span>
<span id="cb42-401"><a href="#cb42-401" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Group means**: Average score on each discriminant function for each group</span>
<span id="cb42-402"><a href="#cb42-402" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-405"><a href="#cb42-405" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-406"><a href="#cb42-406" aria-hidden="true" tabindex="-1"></a><span class="co"># Coefficients (loadings)</span></span>
<span id="cb42-407"><a href="#cb42-407" aria-hidden="true" tabindex="-1"></a>lda_model<span class="sc">$</span>scaling</span>
<span id="cb42-408"><a href="#cb42-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-409"><a href="#cb42-409" aria-hidden="true" tabindex="-1"></a><span class="co"># Proportion of separation explained</span></span>
<span id="cb42-410"><a href="#cb42-410" aria-hidden="true" tabindex="-1"></a>lda_model<span class="sc">$</span>svd<span class="sc">^</span><span class="dv">2</span> <span class="sc">/</span> <span class="fu">sum</span>(lda_model<span class="sc">$</span>svd<span class="sc">^</span><span class="dv">2</span>)</span>
<span id="cb42-411"><a href="#cb42-411" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-412"><a href="#cb42-412" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-413"><a href="#cb42-413" aria-hidden="true" tabindex="-1"></a><span class="fu">### Using DFA for Prediction</span></span>
<span id="cb42-414"><a href="#cb42-414" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-415"><a href="#cb42-415" aria-hidden="true" tabindex="-1"></a>DFA can classify new observations into groups based on their discriminant scores:</span>
<span id="cb42-416"><a href="#cb42-416" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-419"><a href="#cb42-419" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-420"><a href="#cb42-420" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification accuracy</span></span>
<span id="cb42-421"><a href="#cb42-421" aria-hidden="true" tabindex="-1"></a>predictions <span class="ot">&lt;-</span> <span class="fu">predict</span>(lda_model)<span class="sc">$</span>class</span>
<span id="cb42-422"><a href="#cb42-422" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Predicted =</span> predictions, <span class="at">Actual =</span> iris<span class="sc">$</span>Species)</span>
<span id="cb42-423"><a href="#cb42-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-424"><a href="#cb42-424" aria-hidden="true" tabindex="-1"></a><span class="co"># Classification accuracy</span></span>
<span id="cb42-425"><a href="#cb42-425" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(predictions <span class="sc">==</span> iris<span class="sc">$</span>Species)</span>
<span id="cb42-426"><a href="#cb42-426" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-427"><a href="#cb42-427" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-428"><a href="#cb42-428" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cross-Validated Classification</span></span>
<span id="cb42-429"><a href="#cb42-429" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-430"><a href="#cb42-430" aria-hidden="true" tabindex="-1"></a>For honest estimates of classification accuracy, use leave-one-out cross-validation:</span>
<span id="cb42-431"><a href="#cb42-431" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-434"><a href="#cb42-434" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-435"><a href="#cb42-435" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated LDA</span></span>
<span id="cb42-436"><a href="#cb42-436" aria-hidden="true" tabindex="-1"></a>lda_cv <span class="ot">&lt;-</span> <span class="fu">lda</span>(Species <span class="sc">~</span> ., <span class="at">data =</span> iris, <span class="at">CV =</span> <span class="cn">TRUE</span>)</span>
<span id="cb42-437"><a href="#cb42-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-438"><a href="#cb42-438" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated classification table</span></span>
<span id="cb42-439"><a href="#cb42-439" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Predicted =</span> lda_cv<span class="sc">$</span>class, <span class="at">Actual =</span> iris<span class="sc">$</span>Species)</span>
<span id="cb42-440"><a href="#cb42-440" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-441"><a href="#cb42-441" aria-hidden="true" tabindex="-1"></a><span class="co"># Cross-validated accuracy</span></span>
<span id="cb42-442"><a href="#cb42-442" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(lda_cv<span class="sc">$</span>class <span class="sc">==</span> iris<span class="sc">$</span>Species)</span>
<span id="cb42-443"><a href="#cb42-443" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-444"><a href="#cb42-444" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-445"><a href="#cb42-445" aria-hidden="true" tabindex="-1"></a><span class="fu">### DFA for Biomarker Discovery</span></span>
<span id="cb42-446"><a href="#cb42-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-447"><a href="#cb42-447" aria-hidden="true" tabindex="-1"></a>DFA is valuable for identifying which variables best distinguish groups—useful in biomarker discovery:</span>
<span id="cb42-448"><a href="#cb42-448" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-451"><a href="#cb42-451" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-452"><a href="#cb42-452" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb42-453"><a href="#cb42-453" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb42-454"><a href="#cb42-454" aria-hidden="true" tabindex="-1"></a><span class="co"># Which variables contribute most to separation?</span></span>
<span id="cb42-455"><a href="#cb42-455" aria-hidden="true" tabindex="-1"></a>scaling_df <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb42-456"><a href="#cb42-456" aria-hidden="true" tabindex="-1"></a>  <span class="at">Variable =</span> <span class="fu">rownames</span>(lda_model<span class="sc">$</span>scaling),</span>
<span id="cb42-457"><a href="#cb42-457" aria-hidden="true" tabindex="-1"></a>  <span class="at">LD1 =</span> <span class="fu">abs</span>(lda_model<span class="sc">$</span>scaling[, <span class="dv">1</span>]),</span>
<span id="cb42-458"><a href="#cb42-458" aria-hidden="true" tabindex="-1"></a>  <span class="at">LD2 =</span> <span class="fu">abs</span>(lda_model<span class="sc">$</span>scaling[, <span class="dv">2</span>])</span>
<span id="cb42-459"><a href="#cb42-459" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-460"><a href="#cb42-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-461"><a href="#cb42-461" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(scaling_df<span class="sc">$</span>LD1, <span class="at">names.arg =</span> scaling_df<span class="sc">$</span>Variable,</span>
<span id="cb42-462"><a href="#cb42-462" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Variable Contributions to LD1"</span>,</span>
<span id="cb42-463"><a href="#cb42-463" aria-hidden="true" tabindex="-1"></a>        <span class="at">ylab =</span> <span class="st">"Absolute Coefficient"</span>,</span>
<span id="cb42-464"><a href="#cb42-464" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="st">"steelblue"</span>)</span>
<span id="cb42-465"><a href="#cb42-465" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-466"><a href="#cb42-466" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-467"><a href="#cb42-467" aria-hidden="true" tabindex="-1"></a><span class="fu">## Comparing Methods</span></span>
<span id="cb42-468"><a href="#cb42-468" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-469"><a href="#cb42-469" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Method <span class="pp">|</span> Input <span class="pp">|</span> Output <span class="pp">|</span> Supervision <span class="pp">|</span> Best For <span class="pp">|</span></span>
<span id="cb42-470"><a href="#cb42-470" aria-hidden="true" tabindex="-1"></a><span class="pp">|:-------|:------|:-------|:------------|:---------|</span></span>
<span id="cb42-471"><a href="#cb42-471" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> PCA <span class="pp">|</span> Variables <span class="pp">|</span> Continuous scores <span class="pp">|</span> None <span class="pp">|</span> Reducing correlated variables <span class="pp">|</span></span>
<span id="cb42-472"><a href="#cb42-472" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> PCoA <span class="pp">|</span> Distance matrix <span class="pp">|</span> Continuous scores <span class="pp">|</span> None <span class="pp">|</span> Preserving sample distances <span class="pp">|</span></span>
<span id="cb42-473"><a href="#cb42-473" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> NMDS <span class="pp">|</span> Distance matrix <span class="pp">|</span> Ordinal scores <span class="pp">|</span> None <span class="pp">|</span> Ecological community data <span class="pp">|</span></span>
<span id="cb42-474"><a href="#cb42-474" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Cluster Analysis <span class="pp">|</span> Variables or distances <span class="pp">|</span> Group assignments <span class="pp">|</span> None <span class="pp">|</span> Finding natural groupings <span class="pp">|</span></span>
<span id="cb42-475"><a href="#cb42-475" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> MANOVA <span class="pp">|</span> Variables + groups <span class="pp">|</span> Test statistics <span class="pp">|</span> Groups known <span class="pp">|</span> Testing group differences <span class="pp">|</span></span>
<span id="cb42-476"><a href="#cb42-476" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> DFA <span class="pp">|</span> Variables + groups <span class="pp">|</span> Discriminant scores <span class="pp">|</span> Groups known <span class="pp">|</span> Classifying observations <span class="pp">|</span></span>
<span id="cb42-477"><a href="#cb42-477" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-478"><a href="#cb42-478" aria-hidden="true" tabindex="-1"></a><span class="fu">## Using Ordination Scores in Further Analyses</span></span>
<span id="cb42-479"><a href="#cb42-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-480"><a href="#cb42-480" aria-hidden="true" tabindex="-1"></a>PC scores and discriminant scores are legitimate new variables that can be used in downstream analysis:</span>
<span id="cb42-481"><a href="#cb42-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-482"><a href="#cb42-482" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Regression of scores on other continuous variables</span>
<span id="cb42-483"><a href="#cb42-483" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>ANOVA comparing groups on ordination scores</span>
<span id="cb42-484"><a href="#cb42-484" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Correlation of scores with environmental gradients</span>
<span id="cb42-485"><a href="#cb42-485" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-486"><a href="#cb42-486" aria-hidden="true" tabindex="-1"></a>This is valuable when you have many correlated variables and want to reduce dimensionality before hypothesis testing.</span>
<span id="cb42-487"><a href="#cb42-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-490"><a href="#cb42-490" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb42-491"><a href="#cb42-491" aria-hidden="true" tabindex="-1"></a><span class="co"># Use PC scores in ANOVA</span></span>
<span id="cb42-492"><a href="#cb42-492" aria-hidden="true" tabindex="-1"></a>pc_scores <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb42-493"><a href="#cb42-493" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC1 =</span> iris_pca<span class="sc">$</span>x[, <span class="dv">1</span>],</span>
<span id="cb42-494"><a href="#cb42-494" aria-hidden="true" tabindex="-1"></a>  <span class="at">PC2 =</span> iris_pca<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb42-495"><a href="#cb42-495" aria-hidden="true" tabindex="-1"></a>  <span class="at">Species =</span> iris<span class="sc">$</span>Species</span>
<span id="cb42-496"><a href="#cb42-496" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb42-497"><a href="#cb42-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-498"><a href="#cb42-498" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(<span class="fu">aov</span>(PC1 <span class="sc">~</span> Species, <span class="at">data =</span> pc_scores))</span>
<span id="cb42-499"><a href="#cb42-499" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb42-500"><a href="#cb42-500" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-501"><a href="#cb42-501" aria-hidden="true" tabindex="-1"></a><span class="fu">## Practical Workflow</span></span>
<span id="cb42-502"><a href="#cb42-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-503"><a href="#cb42-503" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Explore data**: Check for outliers, missing values, scaling issues</span>
<span id="cb42-504"><a href="#cb42-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-505"><a href="#cb42-505" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Standardize if needed**: Especially important when variables are on different scales</span>
<span id="cb42-506"><a href="#cb42-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-507"><a href="#cb42-507" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Choose appropriate method**: Based on your data type and question</span>
<span id="cb42-508"><a href="#cb42-508" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-509"><a href="#cb42-509" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Examine output**: Scree plots, loadings, clustering diagnostics</span>
<span id="cb42-510"><a href="#cb42-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-511"><a href="#cb42-511" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>**Validate**: Cross-validation for classification; permutation tests for significance</span>
<span id="cb42-512"><a href="#cb42-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-513"><a href="#cb42-513" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>**Interpret biologically**: What do the patterns mean in your system?</span>
<span id="cb42-514"><a href="#cb42-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-515"><a href="#cb42-515" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb42-516"><a href="#cb42-516" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-517"><a href="#cb42-517" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Dimensionality reduction creates fewer variables that capture most information</span>
<span id="cb42-518"><a href="#cb42-518" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>PCA finds linear combinations maximizing variance; useful for correlated variables</span>
<span id="cb42-519"><a href="#cb42-519" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>PCoA works from distance matrices; useful for ecological and genetic data</span>
<span id="cb42-520"><a href="#cb42-520" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Cluster analysis groups similar observations together</span>
<span id="cb42-521"><a href="#cb42-521" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>MANOVA tests whether groups differ on multiple response variables simultaneously</span>
<span id="cb42-522"><a href="#cb42-522" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>DFA finds combinations that best discriminate known groups</span>
<span id="cb42-523"><a href="#cb42-523" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>These methods can be combined: use PCA to reduce dimensions, then cluster or classify</span>
<span id="cb42-524"><a href="#cb42-524" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-525"><a href="#cb42-525" aria-hidden="true" tabindex="-1"></a><span class="fu">## Additional Resources</span></span>
<span id="cb42-526"><a href="#cb42-526" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-527"><a href="#cb42-527" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>@james2023islr - Modern treatment of dimensionality reduction and clustering</span>
<span id="cb42-528"><a href="#cb42-528" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>@logan2010biostatistical - MANOVA and DFA in biological research contexts</span>
<span id="cb42-529"><a href="#cb42-529" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Borcard, D., Gillet, F., &amp; Legendre, P. (2018). *Numerical Ecology with R* - Comprehensive ordination methods</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Statistics for Bioengineering by William Cresko</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Built with <a href="https://quarto.org/">Quarto</a></p>
</div>
  </div>
</footer>




</body></html>