<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>36&nbsp; Clustering – Statistics for the Biosciences and Bioengineering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/36-dimensionality-reduction.html" rel="next">
<link href="../chapters/34-svm.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-485d01fc63b59abcd3ee1bf1e8e2748d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/28-intro-statistical-learning.html">Statistical Learning</a></li><li class="breadcrumb-item"><a href="../chapters/35-clustering.html"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Clustering</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Statistics for the Biosciences and Bioengineering</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Why This Book?</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Core Data Science Tools</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/01-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/02-installing-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Installing Core Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/03-unix-command-line.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Unix and the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/04-r-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/05-markdown-latex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Markdown and LaTeX</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Data Exploration</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/06-tidy-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Tidy Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/07-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data Wrangling with dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/08-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Exploratory Data Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/09-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Data Visualization</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Probability and Distributions</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/10-probability-foundations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Foundations of Probability</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/11-discrete-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Discrete Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/12-continuous-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Continuous Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/13-sampling-estimation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Sampling and Parameter Estimation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Inference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/14-experimental-design.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Experimental Design Principles</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/15-hypothesis-testing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/16-t-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">T-Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/17-nonparametric-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Nonparametric Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/18-bootstrapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Resampling Methods</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/19-what-are-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">What are Models?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/20-correlation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">Correlation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/21-simple-linear-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Simple Linear Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/22-residual-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Residual Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/23-statistical-power.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Statistical Power</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/24-multiple-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Multiple Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/25-single-factor-anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Single Factor ANOVA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/26-multifactor-anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Multi-Factor ANOVA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/27-glm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Generalized Linear Models</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/28-intro-statistical-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Introduction to Statistical Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/29-model-validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Model Validation and the Bias-Variance Tradeoff</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/30-regularization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Regularization Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/31-smoothing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Smoothing and Non-Parametric Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/32-classification-methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Classification and Performance Metrics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/33-trees-forests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Decision Trees and Random Forests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/34-svm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Support Vector Machines</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/35-clustering.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/36-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/37-tsne-umap.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">Non-Linear Dimensionality Reduction: t-SNE and UMAP</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/38-bayesian-statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Bayesian Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/39-deep-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">40</span>&nbsp; <span class="chapter-title">Introduction to Deep Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/40-high-performance-computing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">41</span>&nbsp; <span class="chapter-title">High Performance Computing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Scientific Communication</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/41-presenting-results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">42</span>&nbsp; <span class="chapter-title">Presenting Statistical Results</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true">
 <span class="menu-text">Historical Context</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-8" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-8" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A1-eugenics-history.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">43</span>&nbsp; <span class="chapter-title">The Eugenics History of Statistics</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-9" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-9" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A2-keyboard-shortcuts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">44</span>&nbsp; <span class="chapter-title">Keyboard Shortcuts Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A3-unix-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">45</span>&nbsp; <span class="chapter-title">Unix Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A4-r-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">46</span>&nbsp; <span class="chapter-title">R Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A5-quarto-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">47</span>&nbsp; <span class="chapter-title">Quarto Markdown Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A6-latex-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">48</span>&nbsp; <span class="chapter-title">LaTeX Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A7-greek-letters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">49</span>&nbsp; <span class="chapter-title">Greek Letters in Mathematics and Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A8-probability-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">50</span>&nbsp; <span class="chapter-title">Common Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A9-sampling-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">51</span>&nbsp; <span class="chapter-title">Sampling Distributions in Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A10-matrix-algebra.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">52</span>&nbsp; <span class="chapter-title">Matrix Algebra Fundamentals</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction-to-unsupervised-learning" id="toc-introduction-to-unsupervised-learning" class="nav-link active" data-scroll-target="#introduction-to-unsupervised-learning"><span class="header-section-number">36.1</span> Introduction to Unsupervised Learning</a>
  <ul class="collapse">
  <li><a href="#when-clustering-is-useful" id="toc-when-clustering-is-useful" class="nav-link" data-scroll-target="#when-clustering-is-useful">When Clustering is Useful</a></li>
  <li><a href="#clustering-requires-distance" id="toc-clustering-requires-distance" class="nav-link" data-scroll-target="#clustering-requires-distance">Clustering Requires Distance</a></li>
  </ul></li>
  <li><a href="#a-motivating-example-movie-ratings" id="toc-a-motivating-example-movie-ratings" class="nav-link" data-scroll-target="#a-motivating-example-movie-ratings"><span class="header-section-number">36.2</span> A Motivating Example: Movie Ratings</a></li>
  <li><a href="#hierarchical-clustering" id="toc-hierarchical-clustering" class="nav-link" data-scroll-target="#hierarchical-clustering"><span class="header-section-number">36.3</span> Hierarchical Clustering</a>
  <ul class="collapse">
  <li><a href="#the-agglomerative-algorithm" id="toc-the-agglomerative-algorithm" class="nav-link" data-scroll-target="#the-agglomerative-algorithm">The Agglomerative Algorithm</a></li>
  <li><a href="#visualizing-with-dendrograms" id="toc-visualizing-with-dendrograms" class="nav-link" data-scroll-target="#visualizing-with-dendrograms">Visualizing with Dendrograms</a></li>
  <li><a href="#interpreting-dendrograms" id="toc-interpreting-dendrograms" class="nav-link" data-scroll-target="#interpreting-dendrograms">Interpreting Dendrograms</a></li>
  <li><a href="#cutting-the-tree-to-form-clusters" id="toc-cutting-the-tree-to-form-clusters" class="nav-link" data-scroll-target="#cutting-the-tree-to-form-clusters">Cutting the Tree to Form Clusters</a></li>
  <li><a href="#linkage-methods" id="toc-linkage-methods" class="nav-link" data-scroll-target="#linkage-methods">Linkage Methods</a></li>
  <li><a href="#mathematical-properties" id="toc-mathematical-properties" class="nav-link" data-scroll-target="#mathematical-properties">Mathematical Properties</a></li>
  </ul></li>
  <li><a href="#k-means-clustering" id="toc-k-means-clustering" class="nav-link" data-scroll-target="#k-means-clustering"><span class="header-section-number">36.4</span> K-Means Clustering</a>
  <ul class="collapse">
  <li><a href="#the-k-means-algorithm" id="toc-the-k-means-algorithm" class="nav-link" data-scroll-target="#the-k-means-algorithm">The K-Means Algorithm</a></li>
  <li><a href="#k-means-in-r" id="toc-k-means-in-r" class="nav-link" data-scroll-target="#k-means-in-r">K-Means in R</a></li>
  <li><a href="#comparing-to-hierarchical-clustering" id="toc-comparing-to-hierarchical-clustering" class="nav-link" data-scroll-target="#comparing-to-hierarchical-clustering">Comparing to Hierarchical Clustering</a></li>
  <li><a href="#visualizing-k-means-results" id="toc-visualizing-k-means-results" class="nav-link" data-scroll-target="#visualizing-k-means-results">Visualizing K-Means Results</a></li>
  </ul></li>
  <li><a href="#choosing-the-number-of-clusters" id="toc-choosing-the-number-of-clusters" class="nav-link" data-scroll-target="#choosing-the-number-of-clusters"><span class="header-section-number">36.5</span> Choosing the Number of Clusters</a>
  <ul class="collapse">
  <li><a href="#the-elbow-method" id="toc-the-elbow-method" class="nav-link" data-scroll-target="#the-elbow-method">The Elbow Method</a></li>
  <li><a href="#silhouette-analysis" id="toc-silhouette-analysis" class="nav-link" data-scroll-target="#silhouette-analysis">Silhouette Analysis</a></li>
  <li><a href="#gap-statistic" id="toc-gap-statistic" class="nav-link" data-scroll-target="#gap-statistic">Gap Statistic</a></li>
  </ul></li>
  <li><a href="#heatmaps" id="toc-heatmaps" class="nav-link" data-scroll-target="#heatmaps"><span class="header-section-number">36.6</span> Heatmaps</a>
  <ul class="collapse">
  <li><a href="#basic-heatmap-construction" id="toc-basic-heatmap-construction" class="nav-link" data-scroll-target="#basic-heatmap-construction">Basic Heatmap Construction</a></li>
  <li><a href="#heatmaps-in-r" id="toc-heatmaps-in-r" class="nav-link" data-scroll-target="#heatmaps-in-r">Heatmaps in R</a></li>
  <li><a href="#customizing-heatmaps-with-color-bars" id="toc-customizing-heatmaps-with-color-bars" class="nav-link" data-scroll-target="#customizing-heatmaps-with-color-bars">Customizing Heatmaps with Color Bars</a></li>
  <li><a href="#filtering-features-for-better-visualization" id="toc-filtering-features-for-better-visualization" class="nav-link" data-scroll-target="#filtering-features-for-better-visualization">Filtering Features for Better Visualization</a></li>
  <li><a href="#advanced-heatmaps-with-pheatmap" id="toc-advanced-heatmaps-with-pheatmap" class="nav-link" data-scroll-target="#advanced-heatmaps-with-pheatmap">Advanced Heatmaps with pheatmap</a></li>
  </ul></li>
  <li><a href="#practical-considerations" id="toc-practical-considerations" class="nav-link" data-scroll-target="#practical-considerations"><span class="header-section-number">36.7</span> Practical Considerations</a>
  <ul class="collapse">
  <li><a href="#standardization" id="toc-standardization" class="nav-link" data-scroll-target="#standardization">Standardization</a></li>
  <li><a href="#handling-missing-data" id="toc-handling-missing-data" class="nav-link" data-scroll-target="#handling-missing-data">Handling Missing Data</a></li>
  <li><a href="#cluster-validation" id="toc-cluster-validation" class="nav-link" data-scroll-target="#cluster-validation">Cluster Validation</a></li>
  <li><a href="#reproducibility" id="toc-reproducibility" class="nav-link" data-scroll-target="#reproducibility">Reproducibility</a></li>
  </ul></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">36.8</span> Exercises</a></li>
  <li><a href="#k-means-vs-hierarchical-clustering" id="toc-k-means-vs-hierarchical-clustering" class="nav-link" data-scroll-target="#k-means-vs-hierarchical-clustering"><span class="header-section-number">36.9</span> K-Means vs Hierarchical Clustering</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">36.10</span> Summary</a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="header-section-number">36.11</span> Additional Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/28-intro-statistical-learning.html">Statistical Learning</a></li><li class="breadcrumb-item"><a href="../chapters/35-clustering.html"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Clustering</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-clustering" class="quarto-section-identifier"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Clustering</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction-to-unsupervised-learning" class="level2" data-number="36.1">
<h2 data-number="36.1" class="anchored" data-anchor-id="introduction-to-unsupervised-learning"><span class="header-section-number">36.1</span> Introduction to Unsupervised Learning</h2>
<p>The machine learning algorithms we’ve discussed so far—regression, classification—are examples of <strong>supervised learning</strong>. The name comes from the fact that we use known outcomes (labels) in a training set to “supervise” the creation of our prediction algorithm.</p>
<p><strong>Unsupervised learning</strong> is fundamentally different: we do not necessarily know the outcomes and instead are interested in discovering hidden structure in the data. Clustering algorithms are the most common unsupervised methods—they use features to group observations into clusters without predefined labels.</p>
<section id="when-clustering-is-useful" class="level3">
<h3 class="anchored" data-anchor-id="when-clustering-is-useful">When Clustering is Useful</h3>
<p>Clustering excels when:</p>
<ul>
<li>You want to discover natural groupings in your data</li>
<li>Labels are unavailable or expensive to obtain</li>
<li>You’re performing exploratory data analysis</li>
<li>You need to segment populations for further study</li>
</ul>
<p>Examples in biology:</p>
<ul>
<li>Identifying cell types from single-cell RNA-seq data</li>
<li>Grouping patients by disease subtype</li>
<li>Finding functional modules in protein interaction networks</li>
<li>Classifying microbiome communities</li>
</ul>
</section>
<section id="clustering-requires-distance" class="level3">
<h3 class="anchored" data-anchor-id="clustering-requires-distance">Clustering Requires Distance</h3>
<p>A first step in any clustering algorithm is defining a <strong>distance</strong> between observations. How we measure similarity fundamentally affects what clusters we find. We typically use Euclidean distance, but other metrics (Manhattan, correlation-based) are sometimes more appropriate.</p>
</section>
</section>
<section id="a-motivating-example-movie-ratings" class="level2" data-number="36.2">
<h2 data-number="36.2" class="anchored" data-anchor-id="a-motivating-example-movie-ratings"><span class="header-section-number">36.2</span> A Motivating Example: Movie Ratings</h2>
<p>Let’s construct a clustering example using movie ratings. We’ll identify whether there are natural groupings among popular movies based on how users rate them.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dslabs)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"movielens"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 50 most-rated movies</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>top <span class="ot">&lt;-</span> movielens <span class="sc">%&gt;%</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(movieId) <span class="sc">%&gt;%</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">n =</span> <span class="fu">n</span>(), <span class="at">title =</span> <span class="fu">first</span>(title)) <span class="sc">%&gt;%</span></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="dv">50</span>, n) <span class="sc">%&gt;%</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(movieId)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Create rating matrix for users with at least 25 ratings</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> movielens <span class="sc">%&gt;%</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(movieId <span class="sc">%in%</span> top) <span class="sc">%&gt;%</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(userId) <span class="sc">%&gt;%</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">n</span>() <span class="sc">&gt;=</span> <span class="dv">25</span>) <span class="sc">%&gt;%</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(title, userId, rating) <span class="sc">%&gt;%</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> userId, <span class="at">values_from =</span> rating)</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a>row_names <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(x<span class="sc">$</span>title, <span class="st">": Episode"</span>) <span class="sc">%&gt;%</span> <span class="fu">str_trunc</span>(<span class="dv">20</span>)</span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> x[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Center by user and movie means</span></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x, <span class="dv">2</span>, <span class="fu">colMeans</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x, <span class="dv">1</span>, <span class="fu">rowMeans</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(x) <span class="ot">&lt;-</span> row_names</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>We now have a matrix of centered ratings. Can we find groups of similar movies based on how users rate them?</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="hierarchical-clustering" class="level2" data-number="36.3">
<h2 data-number="36.3" class="anchored" data-anchor-id="hierarchical-clustering"><span class="header-section-number">36.3</span> Hierarchical Clustering</h2>
<p><strong>Hierarchical clustering</strong> builds a tree-like structure (dendrogram) showing nested relationships between observations or clusters.</p>
<section id="the-agglomerative-algorithm" class="level3">
<h3 class="anchored" data-anchor-id="the-agglomerative-algorithm">The Agglomerative Algorithm</h3>
<p>The most common approach is <strong>agglomerative</strong> (bottom-up) clustering:</p>
<ol type="1">
<li>Start with each observation as its own cluster</li>
<li>Find the two closest clusters and merge them</li>
<li>Repeat until all observations are in a single cluster</li>
<li>The result is a tree showing the sequence of merges</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="visualizing-with-dendrograms" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-with-dendrograms">Visualizing with Dendrograms</h3>
<p>A <strong>dendrogram</strong> displays the hierarchical structure:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>rafalib<span class="sc">::</span><span class="fu">mypar</span>()</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(h, <span class="at">cex =</span> <span class="fl">0.65</span>, <span class="at">main =</span> <span class="st">"Hierarchical Clustering of Movies"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-dendrogram" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-dendrogram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-dendrogram-1.png" class="img-fluid figure-img" width="960">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-dendrogram-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.1: Hierarchical clustering dendrogram of movies. Height indicates distance at which clusters merge. Movies that split at lower heights are more similar.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="interpreting-dendrograms" class="level3">
<h3 class="anchored" data-anchor-id="interpreting-dendrograms">Interpreting Dendrograms</h3>
<p>To find the distance between any two movies:</p>
<ol type="1">
<li>Start at the movies’ leaves</li>
<li>Move up until you reach the node where they first join</li>
<li>The height of that node is the distance between them</li>
</ol>
<p>For example, the Star Wars movies merge at low heights (they’re similar), while the distance between Raiders of the Lost Ark and Silence of the Lambs is much greater.</p>
</section>
<section id="cutting-the-tree-to-form-clusters" class="level3">
<h3 class="anchored" data-anchor-id="cutting-the-tree-to-form-clusters">Cutting the Tree to Form Clusters</h3>
<p>To create discrete clusters, we “cut” the dendrogram either:</p>
<ol type="1">
<li>At a specified height (minimum distance for cluster membership)</li>
<li>To obtain a specified number of clusters</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">cutree</span>(h, <span class="at">k =</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
<p>Let’s examine some of the resulting clusters:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Blockbusters cluster</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Cluster 4 (Blockbusters):</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>Cluster 4 (Blockbusters):</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groups)[groups <span class="sc">==</span> <span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Fugitive, The"        "Speed"                "Mission: Impossible" 
[4] "Independence Day ..."</code></pre>
</div>
</div>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Another cluster</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Cluster 9:</span><span class="sc">\n</span><span class="st">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>
Cluster 9:</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groups)[groups <span class="sc">==</span> <span class="dv">9</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Dumb &amp; Dumber (Du..." "Ace Ventura: Pet ..." "Mask, The"           </code></pre>
</div>
</div>
<p>The clustering reveals meaningful groupings—similar genres and franchises cluster together.</p>
</section>
<section id="linkage-methods" class="level3">
<h3 class="anchored" data-anchor-id="linkage-methods">Linkage Methods</h3>
<p>Different <strong>linkage methods</strong> define how cluster-to-cluster distance is computed:</p>
<table class="caption-top table">
<colgroup>
<col style="width: 21%">
<col style="width: 34%">
<col style="width: 44%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Method</th>
<th style="text-align: left;">Description</th>
<th style="text-align: left;">Characteristics</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>Complete</strong></td>
<td style="text-align: left;">Maximum distance between any two points</td>
<td style="text-align: left;">Produces compact, spherical clusters</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Single</strong></td>
<td style="text-align: left;">Minimum distance between any two points</td>
<td style="text-align: left;">Tends to create elongated chains</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Average</strong></td>
<td style="text-align: left;">Mean distance between all pairs</td>
<td style="text-align: left;">Balanced approach</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Ward’s</strong></td>
<td style="text-align: left;">Minimizes within-cluster variance</td>
<td style="text-align: left;">Often produces most interpretable clusters</td>
</tr>
</tbody>
</table>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create sample data for comparison</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">4</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">2</span>, <span class="dv">3</span>)),</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">4</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">8</span>, <span class="dv">1</span>))</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="ot">&lt;-</span> <span class="fu">dist</span>(sample_data)</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"single"</span>), <span class="at">main =</span> <span class="st">"Single Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"complete"</span>), <span class="at">main =</span> <span class="st">"Complete Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"average"</span>), <span class="at">main =</span> <span class="st">"Average Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"ward.D2"</span>), <span class="at">main =</span> <span class="st">"Ward's Method"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-linkage-comparison" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-linkage-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-linkage-comparison-1.png" class="img-fluid figure-img" width="960">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-linkage-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.2: Comparison of different linkage methods on the same data. Each method produces different cluster structures.
</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Choosing a Linkage Method
</div>
</div>
<div class="callout-body-container callout-body">
<ul>
<li><strong>Ward’s method</strong> often produces the most interpretable clusters and is a good default</li>
<li><strong>Complete linkage</strong> is useful when you want compact clusters</li>
<li><strong>Single linkage</strong> can detect elongated structures but is sensitive to noise</li>
<li><strong>Average linkage</strong> is a compromise between single and complete</li>
</ul>
</div>
</div>
</section>
<section id="mathematical-properties" class="level3">
<h3 class="anchored" data-anchor-id="mathematical-properties">Mathematical Properties</h3>
<p><strong>Cophenetic correlation</strong> measures how well the dendrogram preserves original distances:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Cophenetic correlation for Ward's method</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>hc_ward <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"ward.D2"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>cophenetic_dist <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(hc_ward)</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(dist_matrix, cophenetic_dist)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.814684</code></pre>
</div>
</div>
<p>Higher values (closer to 1) indicate better preservation of the original distance structure.</p>
</section>
</section>
<section id="k-means-clustering" class="level2" data-number="36.4">
<h2 data-number="36.4" class="anchored" data-anchor-id="k-means-clustering"><span class="header-section-number">36.4</span> K-Means Clustering</h2>
<p><strong>K-means clustering</strong> partitions observations into exactly <span class="math inline">\(K\)</span> groups by minimizing within-cluster variance. Unlike hierarchical clustering, you must specify <span class="math inline">\(K\)</span> in advance.</p>
<section id="the-k-means-algorithm" class="level3">
<h3 class="anchored" data-anchor-id="the-k-means-algorithm">The K-Means Algorithm</h3>
<ol type="1">
<li><strong>Initialize</strong>: Randomly select <span class="math inline">\(K\)</span> cluster centers</li>
<li><strong>Assign</strong>: Assign each observation to the nearest center</li>
<li><strong>Update</strong>: Recompute centers as the mean of assigned observations</li>
<li><strong>Repeat</strong>: Continue until assignments no longer change</li>
</ol>
<p>The algorithm minimizes the <strong>within-cluster sum of squares (WSS)</strong>: <span class="math display">\[
\text{WSS} = \sum_{k=1}^{K} \sum_{x_i \in C_k} ||x_i - \mu_k||^2
\]</span></p>
<p>where <span class="math inline">\(\mu_k\)</span> is the centroid of cluster <span class="math inline">\(k\)</span>.</p>
</section>
<section id="k-means-in-r" class="level3">
<h3 class="anchored" data-anchor-id="k-means-in-r">K-Means in R</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># K-means requires no missing values</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>x_0 <span class="ot">&lt;-</span> x</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>x_0[<span class="fu">is.na</span>(x_0)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Run k-means with 10 clusters</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(x_0, <span class="at">centers =</span> <span class="dv">10</span>, <span class="at">nstart =</span> <span class="dv">25</span>)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster assignments</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>groups_km <span class="ot">&lt;-</span> k<span class="sc">$</span>cluster</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(groups_km)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>groups_km
 1  2  3  4  5  6  7  8  9 10 
 9  6  3  6  8  1  3  3  3  8 </code></pre>
</div>
</div>
<p>The <code>nstart</code> parameter runs the algorithm multiple times with different random starts, selecting the best result. This is important because k-means can converge to local optima.</p>
</section>
<section id="comparing-to-hierarchical-clustering" class="level3">
<h3 class="anchored" data-anchor-id="comparing-to-hierarchical-clustering">Comparing to Hierarchical Clustering</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare k-means to hierarchical clustering</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">KMeans =</span> groups_km, <span class="at">Hierarchical =</span> groups)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>      Hierarchical
KMeans 1 2 3 4 5 6 7 8 9 10
    1  1 2 0 1 2 1 0 2 0  0
    2  0 0 0 2 3 1 0 0 0  0
    3  0 0 0 0 0 0 0 0 3  0
    4  1 5 0 0 0 0 0 0 0  0
    5  0 1 0 0 0 2 3 2 0  0
    6  0 0 0 0 0 0 0 0 0  1
    7  0 0 3 0 0 0 0 0 0  0
    8  0 0 3 0 0 0 0 0 0  0
    9  0 0 0 1 0 0 0 2 0  0
    10 8 0 0 0 0 0 0 0 0  0</code></pre>
</div>
</div>
</section>
<section id="visualizing-k-means-results" class="level3">
<h3 class="anchored" data-anchor-id="visualizing-k-means-results">Visualizing K-Means Results</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Demonstrate k-means on 2D data</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>km_example <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">25</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sample_data<span class="sc">$</span>x, sample_data<span class="sc">$</span>y,</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> km_example<span class="sc">$</span>cluster,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">1.5</span>,</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"X"</span>, <span class="at">ylab =</span> <span class="st">"Y"</span>,</span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"K-Means Clustering (K=3)"</span>)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(km_example<span class="sc">$</span>centers, <span class="at">pch =</span> <span class="dv">4</span>, <span class="at">cex =</span> <span class="dv">3</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="fu">c</span>(<span class="st">"Cluster 1"</span>, <span class="st">"Cluster 2"</span>, <span class="st">"Cluster 3"</span>, <span class="st">"Centers"</span>),</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>), <span class="at">pch =</span> <span class="fu">c</span>(<span class="dv">19</span>, <span class="dv">19</span>, <span class="dv">19</span>, <span class="dv">4</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-kmeans-example" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-kmeans-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-kmeans-example-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-kmeans-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.3: K-means clustering on simulated data with three natural clusters
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="choosing-the-number-of-clusters" class="level2" data-number="36.5">
<h2 data-number="36.5" class="anchored" data-anchor-id="choosing-the-number-of-clusters"><span class="header-section-number">36.5</span> Choosing the Number of Clusters</h2>
<p>A fundamental challenge in clustering is determining the appropriate number of clusters. Several methods can help:</p>
<section id="the-elbow-method" class="level3">
<h3 class="anchored" data-anchor-id="the-elbow-method">The Elbow Method</h3>
<p>Plot within-cluster sum of squares against <span class="math inline">\(K\)</span> and look for an “elbow” where the rate of decrease slows:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>wss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)<span class="sc">$</span>tot.withinss</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, wss, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters (K)"</span>,</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Total Within-Cluster Sum of Squares"</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Elbow Method"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-elbow-method" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-elbow-method-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-elbow-method-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-elbow-method-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.4: Elbow method for selecting number of clusters. Look for the ‘bend’ in the curve where adding more clusters provides diminishing returns.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="silhouette-analysis" class="level3">
<h3 class="anchored" data-anchor-id="silhouette-analysis">Silhouette Analysis</h3>
<p>The <strong>silhouette score</strong> measures how similar an observation is to its own cluster compared to other clusters:</p>
<p><span class="math display">\[
s(i) = \frac{b(i) - a(i)}{\max(a(i), b(i))}
\]</span></p>
<p>where: - <span class="math inline">\(a(i)\)</span> = average distance to other points in the same cluster - <span class="math inline">\(b(i)\)</span> = average distance to points in the nearest other cluster</p>
<p>Values range from -1 to 1: - Close to 1: well-clustered - Close to 0: on the boundary - Negative: possibly misclassified</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate silhouette scores for different K</span></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>sil_width <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>  km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a>  ss <span class="ot">&lt;-</span> <span class="fu">silhouette</span>(km<span class="sc">$</span>cluster, <span class="fu">dist</span>(sample_data))</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(ss[, <span class="st">"sil_width"</span>])</span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, sil_width, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>,</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Average Silhouette Width"</span>,</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Silhouette Method"</span>)</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a><span class="co"># Detailed silhouette plot for K=3</span></span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>km_3 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>sil_3 <span class="ot">&lt;-</span> <span class="fu">silhouette</span>(km_3<span class="sc">$</span>cluster, <span class="fu">dist</span>(sample_data))</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sil_3, <span class="at">col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">main =</span> <span class="st">"Silhouette Plot (K=3)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-silhouette" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-silhouette-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-silhouette-1.png" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-silhouette-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.5: Silhouette analysis for different numbers of clusters. Higher average silhouette width indicates better-defined clusters.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="gap-statistic" class="level3">
<h3 class="anchored" data-anchor-id="gap-statistic">Gap Statistic</h3>
<p>The gap statistic compares the within-cluster dispersion to that expected under a null reference distribution:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>gap_stat <span class="ot">&lt;-</span> <span class="fu">clusGap</span>(sample_data, <span class="at">FUN =</span> kmeans, <span class="at">nstart =</span> <span class="dv">25</span>, <span class="at">K.max =</span> <span class="dv">10</span>, <span class="at">B =</span> <span class="dv">50</span>)</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gap_stat, <span class="at">main =</span> <span class="st">"Gap Statistic"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
</section>
<section id="heatmaps" class="level2" data-number="36.6">
<h2 data-number="36.6" class="anchored" data-anchor-id="heatmaps"><span class="header-section-number">36.6</span> Heatmaps</h2>
<p><strong>Heatmaps</strong> are powerful visualization tools that combine clustering with color-coded data display. They’re particularly useful for high-dimensional data like gene expression matrices.</p>
<section id="basic-heatmap-construction" class="level3">
<h3 class="anchored" data-anchor-id="basic-heatmap-construction">Basic Heatmap Construction</h3>
<p>The idea:</p>
<ol type="1">
<li>Plot the data matrix as an image (colors represent values)</li>
<li>Cluster rows (observations) and reorder by similarity</li>
<li>Cluster columns (features) and reorder by similarity</li>
<li>Display dendrograms showing the hierarchical structure</li>
</ol>
</section>
<section id="heatmaps-in-r" class="level3">
<h3 class="anchored" data-anchor-id="heatmaps-in-r">Heatmaps in R</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"tissue_gene_expression"</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Center the data</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>x_tissue <span class="ot">&lt;-</span> <span class="fu">sweep</span>(tissue_gene_expression<span class="sc">$</span>x, <span class="dv">2</span>,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">colMeans</span>(tissue_gene_expression<span class="sc">$</span>x))</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic heatmap (subsample for visibility)</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>gene_sample <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(x_tissue), <span class="dv">50</span>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, gene_sample],</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">8</span>),</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>,</span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Tissue Gene Expression"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap-basic" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-heatmap-basic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-heatmap-basic-1.png" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heatmap-basic-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.6: Heatmap of tissue gene expression data with hierarchical clustering on both rows and columns
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="customizing-heatmaps-with-color-bars" class="level3">
<h3 class="anchored" data-anchor-id="customizing-heatmaps-with-color-bars">Customizing Heatmaps with Color Bars</h3>
<p>Adding color bars shows group membership:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Color palette for tissues</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>tissue_colors <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">7</span>, <span class="st">"Set2"</span>)[<span class="fu">as.numeric</span>(tissue_gene_expression<span class="sc">$</span>y)]</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap with row colors</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, gene_sample],</span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>        <span class="at">RowSideColors =</span> tissue_colors,</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">10</span>),</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>,</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Tissue Gene Expression"</span>)</span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a><span class="co"># Add legend</span></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">levels</span>(tissue_gene_expression<span class="sc">$</span>y),</span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">brewer.pal</span>(<span class="dv">7</span>, <span class="st">"Set2"</span>), <span class="at">cex =</span> <span class="fl">0.7</span>, <span class="at">bty =</span> <span class="st">"n"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap-colorbar" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-heatmap-colorbar-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-heatmap-colorbar-1.png" class="img-fluid figure-img" width="864">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heatmap-colorbar-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.7: Heatmap with color sidebar indicating tissue type. Hierarchical clustering reveals that samples group by tissue.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="filtering-features-for-better-visualization" class="level3">
<h3 class="anchored" data-anchor-id="filtering-features-for-better-visualization">Filtering Features for Better Visualization</h3>
<p>Not all features contribute meaningful information to clustering. Filtering to high-variance features often improves visualization:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(matrixStats)</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 50 most variable genes</span></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>sds <span class="ot">&lt;-</span> <span class="fu">colSds</span>(x_tissue, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">order</span>(sds, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, top_genes],</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>        <span class="at">RowSideColors =</span> tissue_colors,</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">10</span>),</span>
<span id="cb27-11"><a href="#cb27-11" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>, <span class="at">cexCol =</span> <span class="fl">0.6</span>,</span>
<span id="cb27-12"><a href="#cb27-12" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Top 50 Variable Genes"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-heatmap-filtered" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-heatmap-filtered-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-heatmap-filtered-1.png" class="img-fluid figure-img" width="864">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-heatmap-filtered-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.8: Heatmap of the 50 most variable genes. High-variance features often reveal the most interesting biological patterns.
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="advanced-heatmaps-with-pheatmap" class="level3">
<h3 class="anchored" data-anchor-id="advanced-heatmaps-with-pheatmap">Advanced Heatmaps with pheatmap</h3>
<p>The <code>pheatmap</code> package provides more control and better aesthetics:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare annotation</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>annotation_row <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">Tissue =</span> tissue_gene_expression<span class="sc">$</span>y,</span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="fu">rownames</span>(x_tissue)</span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(x_tissue[, top_genes],</span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>))(<span class="dv">100</span>),</span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> annotation_row,</span>
<span id="cb28-12"><a href="#cb28-12" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb28-13"><a href="#cb28-13" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb28-14"><a href="#cb28-14" aria-hidden="true" tabindex="-1"></a>         <span class="at">main =</span> <span class="st">"Tissue Gene Expression Heatmap"</span>,</span>
<span id="cb28-15"><a href="#cb28-15" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize =</span> <span class="dv">8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-pheatmap" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pheatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="35-clustering_files/figure-html/fig-pheatmap-1.png" class="img-fluid figure-img" width="864">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pheatmap-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;36.9: Enhanced heatmap using pheatmap with annotation tracks and customized appearance
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="practical-considerations" class="level2" data-number="36.7">
<h2 data-number="36.7" class="anchored" data-anchor-id="practical-considerations"><span class="header-section-number">36.7</span> Practical Considerations</h2>
<section id="standardization" class="level3">
<h3 class="anchored" data-anchor-id="standardization">Standardization</h3>
<p>Features on different scales should be standardized before clustering:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Standardize features</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>x_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(x_tissue)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a><span class="co"># This ensures all features contribute equally to distance calculations</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="handling-missing-data" class="level3">
<h3 class="anchored" data-anchor-id="handling-missing-data">Handling Missing Data</h3>
<p>K-means cannot handle missing values. Options include:</p>
<ol type="1">
<li>Remove observations or features with missing data</li>
<li>Impute missing values</li>
<li>Use algorithms that handle missingness (some hierarchical methods)</li>
</ol>
</section>
<section id="cluster-validation" class="level3">
<h3 class="anchored" data-anchor-id="cluster-validation">Cluster Validation</h3>
<p>How do we know if clusters are “real”? Several approaches:</p>
<ol type="1">
<li><strong>Stability</strong>: Do clusters persist across different subsamples?</li>
<li><strong>External validation</strong>: Do clusters correspond to known groups?</li>
<li><strong>Internal metrics</strong>: Silhouette scores, within-cluster variance</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare clusters to known tissue labels</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Cluster =</span> km_3<span class="sc">$</span>cluster, <span class="at">True =</span> sample_data<span class="sc">$</span>x <span class="sc">&gt;</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code>       True
Cluster FALSE TRUE
      1    20    0
      2     0   26
      3    11    3</code></pre>
</div>
</div>
</section>
<section id="reproducibility" class="level3">
<h3 class="anchored" data-anchor-id="reproducibility">Reproducibility</h3>
<p>K-means depends on random initialization. Always:</p>
<ol type="1">
<li>Set a seed for reproducibility</li>
<li>Use <code>nstart</code> to run multiple initializations</li>
<li>Report the random seed used</li>
</ol>
</section>
</section>
<section id="exercises" class="level2" data-number="36.8">
<h2 data-number="36.8" class="anchored" data-anchor-id="exercises"><span class="header-section-number">36.8</span> Exercises</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise Clu.1: Distance Computation
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Load the <code>tissue_gene_expression</code> dataset. Remove the row means and compute the distance between each observation. Store the result in <code>d</code>.</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise Clu.2: Hierarchical Clustering
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="2" type="1">
<li>Make a hierarchical clustering plot and add the tissue types as labels. Which tissues cluster together?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise Clu.3: K-Means Clustering
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="3" type="1">
<li>Run k-means clustering on the data with <span class="math inline">\(K = 7\)</span>. Make a table comparing the identified clusters to the actual tissue types. Run the algorithm several times to see how the answer changes.</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise Clu.4: Heatmap Visualization
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="4" type="1">
<li>Select the 50 most variable genes. Create a heatmap with:
<ul>
<li>Observations in columns</li>
<li>Centered predictors</li>
<li>A color bar showing tissue types</li>
<li>Use <code>ColSideColors</code> argument and <code>col = brewer.pal(11, "RdBu")</code></li>
</ul></li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise Clu.5: Comparing Linkage Methods
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="5" type="1">
<li>Compare different linkage methods on the movie data. Which produces the most interpretable clusters?</li>
</ol>
</div>
</div>
</section>
<section id="k-means-vs-hierarchical-clustering" class="level2" data-number="36.9">
<h2 data-number="36.9" class="anchored" data-anchor-id="k-means-vs-hierarchical-clustering"><span class="header-section-number">36.9</span> K-Means vs Hierarchical Clustering</h2>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Aspect</th>
<th style="text-align: left;">K-Means</th>
<th style="text-align: left;">Hierarchical</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Clusters</td>
<td style="text-align: left;">Must specify K</td>
<td style="text-align: left;">Can cut at any level</td>
</tr>
<tr class="even">
<td style="text-align: left;">Structure</td>
<td style="text-align: left;">Flat partition</td>
<td style="text-align: left;">Nested hierarchy</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Scalability</td>
<td style="text-align: left;">O(nKt) - efficient</td>
<td style="text-align: left;">O(n³) - slower</td>
</tr>
<tr class="even">
<td style="text-align: left;">Shape</td>
<td style="text-align: left;">Spherical clusters</td>
<td style="text-align: left;">Arbitrary shapes</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Missing data</td>
<td style="text-align: left;">Cannot handle</td>
<td style="text-align: left;">Some methods can</td>
</tr>
<tr class="even">
<td style="text-align: left;">Reproducibility</td>
<td style="text-align: left;">Depends on initialization</td>
<td style="text-align: left;">Deterministic</td>
</tr>
</tbody>
</table>
</section>
<section id="summary" class="level2" data-number="36.10">
<h2 data-number="36.10" class="anchored" data-anchor-id="summary"><span class="header-section-number">36.10</span> Summary</h2>
<ul>
<li><strong>Unsupervised learning</strong> discovers structure without predefined labels</li>
<li><strong>Hierarchical clustering</strong> builds a tree showing nested relationships
<ul>
<li>Dendrograms visualize the hierarchy</li>
<li>Linkage methods define cluster-to-cluster distance</li>
<li>Cut the tree to form discrete clusters</li>
</ul></li>
<li><strong>K-means clustering</strong> partitions data into K groups
<ul>
<li>Minimizes within-cluster variance</li>
<li>Requires specifying K in advance</li>
<li>Use multiple starts to avoid local optima</li>
</ul></li>
<li><strong>Heatmaps</strong> combine clustering with data visualization
<ul>
<li>Color encodes values</li>
<li>Rows and columns reordered by similarity</li>
<li>Filter to high-variance features for clarity</li>
</ul></li>
<li><strong>Choosing K</strong> requires judgment and multiple criteria
<ul>
<li>Elbow method: look for bend in WSS curve</li>
<li>Silhouette analysis: measure cluster coherence</li>
<li>Domain knowledge should guide interpretation</li>
</ul></li>
</ul>
</section>
<section id="additional-resources" class="level2" data-number="36.11">
<h2 data-number="36.11" class="anchored" data-anchor-id="additional-resources"><span class="header-section-number">36.11</span> Additional Resources</h2>
<ul>
<li><span class="citation" data-cites="james2023islr">James et al. (<a href="../references.html#ref-james2023islr" role="doc-biblioref">2023</a>)</span> - Comprehensive treatment of clustering methods</li>
<li><span class="citation" data-cites="logan2010biostatistical">Logan (<a href="../references.html#ref-logan2010biostatistical" role="doc-biblioref">2010</a>)</span> - Applications in biological research</li>
<li>Hahsler, M., Piekenbrock, M., &amp; Doran, D. (2019). dbscan: Fast Density-Based Clustering with R</li>
<li><code>pheatmap</code> and <code>ComplexHeatmap</code> package documentation for advanced heatmaps</li>
</ul>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-james2023islr" class="csl-entry" role="listitem">
James, Gareth, Daniela Witten, Trevor Hastie, and Robert Tibshirani. 2023. <em>An Introduction to Statistical Learning with Applications in r</em>. 2nd ed. Springer. <a href="https://www.statlearning.com">https://www.statlearning.com</a>.
</div>
<div id="ref-logan2010biostatistical" class="csl-entry" role="listitem">
Logan, Murray. 2010. <em>Biostatistical Design and Analysis Using r</em>. Wiley-Blackwell.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/34-svm.html" class="pagination-link" aria-label="Support Vector Machines">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Support Vector Machines</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/36-dimensionality-reduction.html" class="pagination-link" aria-label="Dimensionality Reduction and Multivariate Methods">
        <span class="nav-page-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb32" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Clustering {#sec-clustering}</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-6"><a href="#cb32-6" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb32-7"><a href="#cb32-7" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb32-8"><a href="#cb32-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb32-9"><a href="#cb32-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb32-10"><a href="#cb32-10" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span>
<span id="cb32-11"><a href="#cb32-11" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-12"><a href="#cb32-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-13"><a href="#cb32-13" aria-hidden="true" tabindex="-1"></a><span class="fu">## Introduction to Unsupervised Learning</span></span>
<span id="cb32-14"><a href="#cb32-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-15"><a href="#cb32-15" aria-hidden="true" tabindex="-1"></a>The machine learning algorithms we've discussed so far—regression, classification—are examples of **supervised learning**. The name comes from the fact that we use known outcomes (labels) in a training set to "supervise" the creation of our prediction algorithm.</span>
<span id="cb32-16"><a href="#cb32-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-17"><a href="#cb32-17" aria-hidden="true" tabindex="-1"></a>**Unsupervised learning** is fundamentally different: we do not necessarily know the outcomes and instead are interested in discovering hidden structure in the data. Clustering algorithms are the most common unsupervised methods—they use features to group observations into clusters without predefined labels.</span>
<span id="cb32-18"><a href="#cb32-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-19"><a href="#cb32-19" aria-hidden="true" tabindex="-1"></a><span class="fu">### When Clustering is Useful</span></span>
<span id="cb32-20"><a href="#cb32-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-21"><a href="#cb32-21" aria-hidden="true" tabindex="-1"></a>Clustering excels when:</span>
<span id="cb32-22"><a href="#cb32-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-23"><a href="#cb32-23" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You want to discover natural groupings in your data</span>
<span id="cb32-24"><a href="#cb32-24" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Labels are unavailable or expensive to obtain</span>
<span id="cb32-25"><a href="#cb32-25" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You're performing exploratory data analysis</span>
<span id="cb32-26"><a href="#cb32-26" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You need to segment populations for further study</span>
<span id="cb32-27"><a href="#cb32-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-28"><a href="#cb32-28" aria-hidden="true" tabindex="-1"></a>Examples in biology:</span>
<span id="cb32-29"><a href="#cb32-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-30"><a href="#cb32-30" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Identifying cell types from single-cell RNA-seq data</span>
<span id="cb32-31"><a href="#cb32-31" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Grouping patients by disease subtype</span>
<span id="cb32-32"><a href="#cb32-32" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Finding functional modules in protein interaction networks</span>
<span id="cb32-33"><a href="#cb32-33" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Classifying microbiome communities</span>
<span id="cb32-34"><a href="#cb32-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-35"><a href="#cb32-35" aria-hidden="true" tabindex="-1"></a><span class="fu">### Clustering Requires Distance</span></span>
<span id="cb32-36"><a href="#cb32-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-37"><a href="#cb32-37" aria-hidden="true" tabindex="-1"></a>A first step in any clustering algorithm is defining a **distance** between observations. How we measure similarity fundamentally affects what clusters we find. We typically use Euclidean distance, but other metrics (Manhattan, correlation-based) are sometimes more appropriate.</span>
<span id="cb32-38"><a href="#cb32-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-39"><a href="#cb32-39" aria-hidden="true" tabindex="-1"></a><span class="fu">## A Motivating Example: Movie Ratings</span></span>
<span id="cb32-40"><a href="#cb32-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-41"><a href="#cb32-41" aria-hidden="true" tabindex="-1"></a>Let's construct a clustering example using movie ratings. We'll identify whether there are natural groupings among popular movies based on how users rate them.</span>
<span id="cb32-42"><a href="#cb32-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-45"><a href="#cb32-45" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-46"><a href="#cb32-46" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dslabs)</span>
<span id="cb32-47"><a href="#cb32-47" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"movielens"</span>)</span>
<span id="cb32-48"><a href="#cb32-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-49"><a href="#cb32-49" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 50 most-rated movies</span></span>
<span id="cb32-50"><a href="#cb32-50" aria-hidden="true" tabindex="-1"></a>top <span class="ot">&lt;-</span> movielens <span class="sc">%&gt;%</span></span>
<span id="cb32-51"><a href="#cb32-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(movieId) <span class="sc">%&gt;%</span></span>
<span id="cb32-52"><a href="#cb32-52" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarize</span>(<span class="at">n =</span> <span class="fu">n</span>(), <span class="at">title =</span> <span class="fu">first</span>(title)) <span class="sc">%&gt;%</span></span>
<span id="cb32-53"><a href="#cb32-53" aria-hidden="true" tabindex="-1"></a>  <span class="fu">top_n</span>(<span class="dv">50</span>, n) <span class="sc">%&gt;%</span></span>
<span id="cb32-54"><a href="#cb32-54" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(movieId)</span>
<span id="cb32-55"><a href="#cb32-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-56"><a href="#cb32-56" aria-hidden="true" tabindex="-1"></a><span class="co"># Create rating matrix for users with at least 25 ratings</span></span>
<span id="cb32-57"><a href="#cb32-57" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> movielens <span class="sc">%&gt;%</span></span>
<span id="cb32-58"><a href="#cb32-58" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(movieId <span class="sc">%in%</span> top) <span class="sc">%&gt;%</span></span>
<span id="cb32-59"><a href="#cb32-59" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(userId) <span class="sc">%&gt;%</span></span>
<span id="cb32-60"><a href="#cb32-60" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="fu">n</span>() <span class="sc">&gt;=</span> <span class="dv">25</span>) <span class="sc">%&gt;%</span></span>
<span id="cb32-61"><a href="#cb32-61" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ungroup</span>() <span class="sc">%&gt;%</span></span>
<span id="cb32-62"><a href="#cb32-62" aria-hidden="true" tabindex="-1"></a>  <span class="fu">select</span>(title, userId, rating) <span class="sc">%&gt;%</span></span>
<span id="cb32-63"><a href="#cb32-63" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_wider</span>(<span class="at">names_from =</span> userId, <span class="at">values_from =</span> rating)</span>
<span id="cb32-64"><a href="#cb32-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-65"><a href="#cb32-65" aria-hidden="true" tabindex="-1"></a>row_names <span class="ot">&lt;-</span> <span class="fu">str_remove</span>(x<span class="sc">$</span>title, <span class="st">": Episode"</span>) <span class="sc">%&gt;%</span> <span class="fu">str_trunc</span>(<span class="dv">20</span>)</span>
<span id="cb32-66"><a href="#cb32-66" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> x[, <span class="sc">-</span><span class="dv">1</span>] <span class="sc">%&gt;%</span> <span class="fu">as.matrix</span>()</span>
<span id="cb32-67"><a href="#cb32-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-68"><a href="#cb32-68" aria-hidden="true" tabindex="-1"></a><span class="co"># Center by user and movie means</span></span>
<span id="cb32-69"><a href="#cb32-69" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x, <span class="dv">2</span>, <span class="fu">colMeans</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb32-70"><a href="#cb32-70" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">sweep</span>(x, <span class="dv">1</span>, <span class="fu">rowMeans</span>(x, <span class="at">na.rm =</span> <span class="cn">TRUE</span>))</span>
<span id="cb32-71"><a href="#cb32-71" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(x) <span class="ot">&lt;-</span> row_names</span>
<span id="cb32-72"><a href="#cb32-72" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-73"><a href="#cb32-73" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-74"><a href="#cb32-74" aria-hidden="true" tabindex="-1"></a>We now have a matrix of centered ratings. Can we find groups of similar movies based on how users rate them?</span>
<span id="cb32-75"><a href="#cb32-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-78"><a href="#cb32-78" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-79"><a href="#cb32-79" aria-hidden="true" tabindex="-1"></a>d <span class="ot">&lt;-</span> <span class="fu">dist</span>(x)</span>
<span id="cb32-80"><a href="#cb32-80" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-81"><a href="#cb32-81" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-82"><a href="#cb32-82" aria-hidden="true" tabindex="-1"></a><span class="fu">## Hierarchical Clustering</span></span>
<span id="cb32-83"><a href="#cb32-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-84"><a href="#cb32-84" aria-hidden="true" tabindex="-1"></a>**Hierarchical clustering** builds a tree-like structure (dendrogram) showing nested relationships between observations or clusters.</span>
<span id="cb32-85"><a href="#cb32-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-86"><a href="#cb32-86" aria-hidden="true" tabindex="-1"></a><span class="fu">### The Agglomerative Algorithm</span></span>
<span id="cb32-87"><a href="#cb32-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-88"><a href="#cb32-88" aria-hidden="true" tabindex="-1"></a>The most common approach is **agglomerative** (bottom-up) clustering:</span>
<span id="cb32-89"><a href="#cb32-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-90"><a href="#cb32-90" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Start with each observation as its own cluster</span>
<span id="cb32-91"><a href="#cb32-91" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Find the two closest clusters and merge them</span>
<span id="cb32-92"><a href="#cb32-92" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Repeat until all observations are in a single cluster</span>
<span id="cb32-93"><a href="#cb32-93" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>The result is a tree showing the sequence of merges</span>
<span id="cb32-94"><a href="#cb32-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-97"><a href="#cb32-97" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-98"><a href="#cb32-98" aria-hidden="true" tabindex="-1"></a>h <span class="ot">&lt;-</span> <span class="fu">hclust</span>(d)</span>
<span id="cb32-99"><a href="#cb32-99" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-100"><a href="#cb32-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-101"><a href="#cb32-101" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing with Dendrograms</span></span>
<span id="cb32-102"><a href="#cb32-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-103"><a href="#cb32-103" aria-hidden="true" tabindex="-1"></a>A **dendrogram** displays the hierarchical structure:</span>
<span id="cb32-104"><a href="#cb32-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-107"><a href="#cb32-107" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-108"><a href="#cb32-108" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-dendrogram</span></span>
<span id="cb32-109"><a href="#cb32-109" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Hierarchical clustering dendrogram of movies. Height indicates distance at which clusters merge. Movies that split at lower heights are more similar."</span></span>
<span id="cb32-110"><a href="#cb32-110" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb32-111"><a href="#cb32-111" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb32-112"><a href="#cb32-112" aria-hidden="true" tabindex="-1"></a>rafalib<span class="sc">::</span><span class="fu">mypar</span>()</span>
<span id="cb32-113"><a href="#cb32-113" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(h, <span class="at">cex =</span> <span class="fl">0.65</span>, <span class="at">main =</span> <span class="st">"Hierarchical Clustering of Movies"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb32-114"><a href="#cb32-114" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-115"><a href="#cb32-115" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-116"><a href="#cb32-116" aria-hidden="true" tabindex="-1"></a><span class="fu">### Interpreting Dendrograms</span></span>
<span id="cb32-117"><a href="#cb32-117" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-118"><a href="#cb32-118" aria-hidden="true" tabindex="-1"></a>To find the distance between any two movies:</span>
<span id="cb32-119"><a href="#cb32-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-120"><a href="#cb32-120" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Start at the movies' leaves</span>
<span id="cb32-121"><a href="#cb32-121" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Move up until you reach the node where they first join</span>
<span id="cb32-122"><a href="#cb32-122" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>The height of that node is the distance between them</span>
<span id="cb32-123"><a href="#cb32-123" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-124"><a href="#cb32-124" aria-hidden="true" tabindex="-1"></a>For example, the Star Wars movies merge at low heights (they're similar), while the distance between Raiders of the Lost Ark and Silence of the Lambs is much greater.</span>
<span id="cb32-125"><a href="#cb32-125" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-126"><a href="#cb32-126" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cutting the Tree to Form Clusters</span></span>
<span id="cb32-127"><a href="#cb32-127" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-128"><a href="#cb32-128" aria-hidden="true" tabindex="-1"></a>To create discrete clusters, we "cut" the dendrogram either:</span>
<span id="cb32-129"><a href="#cb32-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-130"><a href="#cb32-130" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>At a specified height (minimum distance for cluster membership)</span>
<span id="cb32-131"><a href="#cb32-131" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>To obtain a specified number of clusters</span>
<span id="cb32-132"><a href="#cb32-132" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-135"><a href="#cb32-135" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-136"><a href="#cb32-136" aria-hidden="true" tabindex="-1"></a>groups <span class="ot">&lt;-</span> <span class="fu">cutree</span>(h, <span class="at">k =</span> <span class="dv">10</span>)</span>
<span id="cb32-137"><a href="#cb32-137" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-138"><a href="#cb32-138" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-139"><a href="#cb32-139" aria-hidden="true" tabindex="-1"></a>Let's examine some of the resulting clusters:</span>
<span id="cb32-140"><a href="#cb32-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-143"><a href="#cb32-143" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-144"><a href="#cb32-144" aria-hidden="true" tabindex="-1"></a><span class="co"># Blockbusters cluster</span></span>
<span id="cb32-145"><a href="#cb32-145" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"Cluster 4 (Blockbusters):</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb32-146"><a href="#cb32-146" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groups)[groups <span class="sc">==</span> <span class="dv">4</span>]</span>
<span id="cb32-147"><a href="#cb32-147" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-148"><a href="#cb32-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-151"><a href="#cb32-151" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-152"><a href="#cb32-152" aria-hidden="true" tabindex="-1"></a><span class="co"># Another cluster</span></span>
<span id="cb32-153"><a href="#cb32-153" aria-hidden="true" tabindex="-1"></a><span class="fu">cat</span>(<span class="st">"</span><span class="sc">\n</span><span class="st">Cluster 9:</span><span class="sc">\n</span><span class="st">"</span>)</span>
<span id="cb32-154"><a href="#cb32-154" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(groups)[groups <span class="sc">==</span> <span class="dv">9</span>]</span>
<span id="cb32-155"><a href="#cb32-155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-156"><a href="#cb32-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-157"><a href="#cb32-157" aria-hidden="true" tabindex="-1"></a>The clustering reveals meaningful groupings—similar genres and franchises cluster together.</span>
<span id="cb32-158"><a href="#cb32-158" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-159"><a href="#cb32-159" aria-hidden="true" tabindex="-1"></a><span class="fu">### Linkage Methods</span></span>
<span id="cb32-160"><a href="#cb32-160" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-161"><a href="#cb32-161" aria-hidden="true" tabindex="-1"></a>Different **linkage methods** define how cluster-to-cluster distance is computed:</span>
<span id="cb32-162"><a href="#cb32-162" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-163"><a href="#cb32-163" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Method <span class="pp">|</span> Description <span class="pp">|</span> Characteristics <span class="pp">|</span></span>
<span id="cb32-164"><a href="#cb32-164" aria-hidden="true" tabindex="-1"></a><span class="pp">|:-------|:------------|:----------------|</span></span>
<span id="cb32-165"><a href="#cb32-165" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Complete** <span class="pp">|</span> Maximum distance between any two points <span class="pp">|</span> Produces compact, spherical clusters <span class="pp">|</span></span>
<span id="cb32-166"><a href="#cb32-166" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Single** <span class="pp">|</span> Minimum distance between any two points <span class="pp">|</span> Tends to create elongated chains <span class="pp">|</span></span>
<span id="cb32-167"><a href="#cb32-167" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Average** <span class="pp">|</span> Mean distance between all pairs <span class="pp">|</span> Balanced approach <span class="pp">|</span></span>
<span id="cb32-168"><a href="#cb32-168" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Ward's** <span class="pp">|</span> Minimizes within-cluster variance <span class="pp">|</span> Often produces most interpretable clusters <span class="pp">|</span></span>
<span id="cb32-169"><a href="#cb32-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-172"><a href="#cb32-172" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-173"><a href="#cb32-173" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-linkage-comparison</span></span>
<span id="cb32-174"><a href="#cb32-174" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Comparison of different linkage methods on the same data. Each method produces different cluster structures."</span></span>
<span id="cb32-175"><a href="#cb32-175" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb32-176"><a href="#cb32-176" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb32-177"><a href="#cb32-177" aria-hidden="true" tabindex="-1"></a><span class="co"># Create sample data for comparison</span></span>
<span id="cb32-178"><a href="#cb32-178" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb32-179"><a href="#cb32-179" aria-hidden="true" tabindex="-1"></a>sample_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb32-180"><a href="#cb32-180" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">4</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">2</span>, <span class="dv">3</span>)),</span>
<span id="cb32-181"><a href="#cb32-181" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> <span class="fu">c</span>(<span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">0</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">4</span>, <span class="dv">1</span>), <span class="fu">rnorm</span>(<span class="dv">20</span>, <span class="dv">8</span>, <span class="dv">1</span>))</span>
<span id="cb32-182"><a href="#cb32-182" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb32-183"><a href="#cb32-183" aria-hidden="true" tabindex="-1"></a>dist_matrix <span class="ot">&lt;-</span> <span class="fu">dist</span>(sample_data)</span>
<span id="cb32-184"><a href="#cb32-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-185"><a href="#cb32-185" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">2</span>))</span>
<span id="cb32-186"><a href="#cb32-186" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"single"</span>), <span class="at">main =</span> <span class="st">"Single Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb32-187"><a href="#cb32-187" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"complete"</span>), <span class="at">main =</span> <span class="st">"Complete Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb32-188"><a href="#cb32-188" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"average"</span>), <span class="at">main =</span> <span class="st">"Average Linkage"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb32-189"><a href="#cb32-189" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"ward.D2"</span>), <span class="at">main =</span> <span class="st">"Ward's Method"</span>, <span class="at">xlab =</span> <span class="st">""</span>)</span>
<span id="cb32-190"><a href="#cb32-190" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-191"><a href="#cb32-191" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-192"><a href="#cb32-192" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip}</span>
<span id="cb32-193"><a href="#cb32-193" aria-hidden="true" tabindex="-1"></a><span class="fu">## Choosing a Linkage Method</span></span>
<span id="cb32-194"><a href="#cb32-194" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-195"><a href="#cb32-195" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Ward's method** often produces the most interpretable clusters and is a good default</span>
<span id="cb32-196"><a href="#cb32-196" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Complete linkage** is useful when you want compact clusters</span>
<span id="cb32-197"><a href="#cb32-197" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Single linkage** can detect elongated structures but is sensitive to noise</span>
<span id="cb32-198"><a href="#cb32-198" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Average linkage** is a compromise between single and complete</span>
<span id="cb32-199"><a href="#cb32-199" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-200"><a href="#cb32-200" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-201"><a href="#cb32-201" aria-hidden="true" tabindex="-1"></a><span class="fu">### Mathematical Properties</span></span>
<span id="cb32-202"><a href="#cb32-202" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-203"><a href="#cb32-203" aria-hidden="true" tabindex="-1"></a>**Cophenetic correlation** measures how well the dendrogram preserves original distances:</span>
<span id="cb32-204"><a href="#cb32-204" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-207"><a href="#cb32-207" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-208"><a href="#cb32-208" aria-hidden="true" tabindex="-1"></a><span class="co"># Cophenetic correlation for Ward's method</span></span>
<span id="cb32-209"><a href="#cb32-209" aria-hidden="true" tabindex="-1"></a>hc_ward <span class="ot">&lt;-</span> <span class="fu">hclust</span>(dist_matrix, <span class="at">method =</span> <span class="st">"ward.D2"</span>)</span>
<span id="cb32-210"><a href="#cb32-210" aria-hidden="true" tabindex="-1"></a>cophenetic_dist <span class="ot">&lt;-</span> <span class="fu">cophenetic</span>(hc_ward)</span>
<span id="cb32-211"><a href="#cb32-211" aria-hidden="true" tabindex="-1"></a><span class="fu">cor</span>(dist_matrix, cophenetic_dist)</span>
<span id="cb32-212"><a href="#cb32-212" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-213"><a href="#cb32-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-214"><a href="#cb32-214" aria-hidden="true" tabindex="-1"></a>Higher values (closer to 1) indicate better preservation of the original distance structure.</span>
<span id="cb32-215"><a href="#cb32-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-216"><a href="#cb32-216" aria-hidden="true" tabindex="-1"></a><span class="fu">## K-Means Clustering</span></span>
<span id="cb32-217"><a href="#cb32-217" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-218"><a href="#cb32-218" aria-hidden="true" tabindex="-1"></a>**K-means clustering** partitions observations into exactly $K$ groups by minimizing within-cluster variance. Unlike hierarchical clustering, you must specify $K$ in advance.</span>
<span id="cb32-219"><a href="#cb32-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-220"><a href="#cb32-220" aria-hidden="true" tabindex="-1"></a><span class="fu">### The K-Means Algorithm</span></span>
<span id="cb32-221"><a href="#cb32-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-222"><a href="#cb32-222" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Initialize**: Randomly select $K$ cluster centers</span>
<span id="cb32-223"><a href="#cb32-223" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Assign**: Assign each observation to the nearest center</span>
<span id="cb32-224"><a href="#cb32-224" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Update**: Recompute centers as the mean of assigned observations</span>
<span id="cb32-225"><a href="#cb32-225" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Repeat**: Continue until assignments no longer change</span>
<span id="cb32-226"><a href="#cb32-226" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-227"><a href="#cb32-227" aria-hidden="true" tabindex="-1"></a>The algorithm minimizes the **within-cluster sum of squares (WSS)**:</span>
<span id="cb32-228"><a href="#cb32-228" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb32-229"><a href="#cb32-229" aria-hidden="true" tabindex="-1"></a>\text{WSS} = \sum_{k=1}^{K} \sum_{x_i \in C_k} ||x_i - \mu_k||^2</span>
<span id="cb32-230"><a href="#cb32-230" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb32-231"><a href="#cb32-231" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-232"><a href="#cb32-232" aria-hidden="true" tabindex="-1"></a>where $\mu_k$ is the centroid of cluster $k$.</span>
<span id="cb32-233"><a href="#cb32-233" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-234"><a href="#cb32-234" aria-hidden="true" tabindex="-1"></a><span class="fu">### K-Means in R</span></span>
<span id="cb32-235"><a href="#cb32-235" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-238"><a href="#cb32-238" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-239"><a href="#cb32-239" aria-hidden="true" tabindex="-1"></a><span class="co"># K-means requires no missing values</span></span>
<span id="cb32-240"><a href="#cb32-240" aria-hidden="true" tabindex="-1"></a>x_0 <span class="ot">&lt;-</span> x</span>
<span id="cb32-241"><a href="#cb32-241" aria-hidden="true" tabindex="-1"></a>x_0[<span class="fu">is.na</span>(x_0)] <span class="ot">&lt;-</span> <span class="dv">0</span></span>
<span id="cb32-242"><a href="#cb32-242" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-243"><a href="#cb32-243" aria-hidden="true" tabindex="-1"></a><span class="co"># Run k-means with 10 clusters</span></span>
<span id="cb32-244"><a href="#cb32-244" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb32-245"><a href="#cb32-245" aria-hidden="true" tabindex="-1"></a>k <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(x_0, <span class="at">centers =</span> <span class="dv">10</span>, <span class="at">nstart =</span> <span class="dv">25</span>)</span>
<span id="cb32-246"><a href="#cb32-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-247"><a href="#cb32-247" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster assignments</span></span>
<span id="cb32-248"><a href="#cb32-248" aria-hidden="true" tabindex="-1"></a>groups_km <span class="ot">&lt;-</span> k<span class="sc">$</span>cluster</span>
<span id="cb32-249"><a href="#cb32-249" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(groups_km)</span>
<span id="cb32-250"><a href="#cb32-250" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-251"><a href="#cb32-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-252"><a href="#cb32-252" aria-hidden="true" tabindex="-1"></a>The <span class="in">`nstart`</span> parameter runs the algorithm multiple times with different random starts, selecting the best result. This is important because k-means can converge to local optima.</span>
<span id="cb32-253"><a href="#cb32-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-254"><a href="#cb32-254" aria-hidden="true" tabindex="-1"></a><span class="fu">### Comparing to Hierarchical Clustering</span></span>
<span id="cb32-255"><a href="#cb32-255" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-258"><a href="#cb32-258" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-259"><a href="#cb32-259" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare k-means to hierarchical clustering</span></span>
<span id="cb32-260"><a href="#cb32-260" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">KMeans =</span> groups_km, <span class="at">Hierarchical =</span> groups)</span>
<span id="cb32-261"><a href="#cb32-261" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-262"><a href="#cb32-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-263"><a href="#cb32-263" aria-hidden="true" tabindex="-1"></a><span class="fu">### Visualizing K-Means Results</span></span>
<span id="cb32-264"><a href="#cb32-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-267"><a href="#cb32-267" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-268"><a href="#cb32-268" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-kmeans-example</span></span>
<span id="cb32-269"><a href="#cb32-269" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "K-means clustering on simulated data with three natural clusters"</span></span>
<span id="cb32-270"><a href="#cb32-270" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb32-271"><a href="#cb32-271" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb32-272"><a href="#cb32-272" aria-hidden="true" tabindex="-1"></a><span class="co"># Demonstrate k-means on 2D data</span></span>
<span id="cb32-273"><a href="#cb32-273" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb32-274"><a href="#cb32-274" aria-hidden="true" tabindex="-1"></a>km_example <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">25</span>)</span>
<span id="cb32-275"><a href="#cb32-275" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-276"><a href="#cb32-276" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sample_data<span class="sc">$</span>x, sample_data<span class="sc">$</span>y,</span>
<span id="cb32-277"><a href="#cb32-277" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> km_example<span class="sc">$</span>cluster,</span>
<span id="cb32-278"><a href="#cb32-278" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">1.5</span>,</span>
<span id="cb32-279"><a href="#cb32-279" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"X"</span>, <span class="at">ylab =</span> <span class="st">"Y"</span>,</span>
<span id="cb32-280"><a href="#cb32-280" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"K-Means Clustering (K=3)"</span>)</span>
<span id="cb32-281"><a href="#cb32-281" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(km_example<span class="sc">$</span>centers, <span class="at">pch =</span> <span class="dv">4</span>, <span class="at">cex =</span> <span class="dv">3</span>, <span class="at">lwd =</span> <span class="dv">3</span>)</span>
<span id="cb32-282"><a href="#cb32-282" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topleft"</span>, <span class="fu">c</span>(<span class="st">"Cluster 1"</span>, <span class="st">"Cluster 2"</span>, <span class="st">"Cluster 3"</span>, <span class="st">"Centers"</span>),</span>
<span id="cb32-283"><a href="#cb32-283" aria-hidden="true" tabindex="-1"></a>       <span class="at">col =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">1</span>), <span class="at">pch =</span> <span class="fu">c</span>(<span class="dv">19</span>, <span class="dv">19</span>, <span class="dv">19</span>, <span class="dv">4</span>))</span>
<span id="cb32-284"><a href="#cb32-284" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-285"><a href="#cb32-285" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-286"><a href="#cb32-286" aria-hidden="true" tabindex="-1"></a><span class="fu">## Choosing the Number of Clusters</span></span>
<span id="cb32-287"><a href="#cb32-287" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-288"><a href="#cb32-288" aria-hidden="true" tabindex="-1"></a>A fundamental challenge in clustering is determining the appropriate number of clusters. Several methods can help:</span>
<span id="cb32-289"><a href="#cb32-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-290"><a href="#cb32-290" aria-hidden="true" tabindex="-1"></a><span class="fu">### The Elbow Method</span></span>
<span id="cb32-291"><a href="#cb32-291" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-292"><a href="#cb32-292" aria-hidden="true" tabindex="-1"></a>Plot within-cluster sum of squares against $K$ and look for an "elbow" where the rate of decrease slows:</span>
<span id="cb32-293"><a href="#cb32-293" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-296"><a href="#cb32-296" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-297"><a href="#cb32-297" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-elbow-method</span></span>
<span id="cb32-298"><a href="#cb32-298" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Elbow method for selecting number of clusters. Look for the 'bend' in the curve where adding more clusters provides diminishing returns."</span></span>
<span id="cb32-299"><a href="#cb32-299" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb32-300"><a href="#cb32-300" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb32-301"><a href="#cb32-301" aria-hidden="true" tabindex="-1"></a>wss <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb32-302"><a href="#cb32-302" aria-hidden="true" tabindex="-1"></a>  <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)<span class="sc">$</span>tot.withinss</span>
<span id="cb32-303"><a href="#cb32-303" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-304"><a href="#cb32-304" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-305"><a href="#cb32-305" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>, wss, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb32-306"><a href="#cb32-306" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters (K)"</span>,</span>
<span id="cb32-307"><a href="#cb32-307" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Total Within-Cluster Sum of Squares"</span>,</span>
<span id="cb32-308"><a href="#cb32-308" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Elbow Method"</span>)</span>
<span id="cb32-309"><a href="#cb32-309" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-310"><a href="#cb32-310" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-311"><a href="#cb32-311" aria-hidden="true" tabindex="-1"></a><span class="fu">### Silhouette Analysis</span></span>
<span id="cb32-312"><a href="#cb32-312" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-313"><a href="#cb32-313" aria-hidden="true" tabindex="-1"></a>The **silhouette score** measures how similar an observation is to its own cluster compared to other clusters:</span>
<span id="cb32-314"><a href="#cb32-314" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-315"><a href="#cb32-315" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb32-316"><a href="#cb32-316" aria-hidden="true" tabindex="-1"></a>s(i) = \frac{b(i) - a(i)}{\max(a(i), b(i))}</span>
<span id="cb32-317"><a href="#cb32-317" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb32-318"><a href="#cb32-318" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-319"><a href="#cb32-319" aria-hidden="true" tabindex="-1"></a>where:</span>
<span id="cb32-320"><a href="#cb32-320" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>$a(i)$ = average distance to other points in the same cluster</span>
<span id="cb32-321"><a href="#cb32-321" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>$b(i)$ = average distance to points in the nearest other cluster</span>
<span id="cb32-322"><a href="#cb32-322" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-323"><a href="#cb32-323" aria-hidden="true" tabindex="-1"></a>Values range from -1 to 1:</span>
<span id="cb32-324"><a href="#cb32-324" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Close to 1: well-clustered</span>
<span id="cb32-325"><a href="#cb32-325" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Close to 0: on the boundary</span>
<span id="cb32-326"><a href="#cb32-326" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Negative: possibly misclassified</span>
<span id="cb32-327"><a href="#cb32-327" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-330"><a href="#cb32-330" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-331"><a href="#cb32-331" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-silhouette</span></span>
<span id="cb32-332"><a href="#cb32-332" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Silhouette analysis for different numbers of clusters. Higher average silhouette width indicates better-defined clusters."</span></span>
<span id="cb32-333"><a href="#cb32-333" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb32-334"><a href="#cb32-334" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 5</span></span>
<span id="cb32-335"><a href="#cb32-335" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb32-336"><a href="#cb32-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-337"><a href="#cb32-337" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate silhouette scores for different K</span></span>
<span id="cb32-338"><a href="#cb32-338" aria-hidden="true" tabindex="-1"></a>sil_width <span class="ot">&lt;-</span> <span class="fu">sapply</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, <span class="cf">function</span>(k) {</span>
<span id="cb32-339"><a href="#cb32-339" aria-hidden="true" tabindex="-1"></a>  km <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> k, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb32-340"><a href="#cb32-340" aria-hidden="true" tabindex="-1"></a>  ss <span class="ot">&lt;-</span> <span class="fu">silhouette</span>(km<span class="sc">$</span>cluster, <span class="fu">dist</span>(sample_data))</span>
<span id="cb32-341"><a href="#cb32-341" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mean</span>(ss[, <span class="st">"sil_width"</span>])</span>
<span id="cb32-342"><a href="#cb32-342" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb32-343"><a href="#cb32-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-344"><a href="#cb32-344" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb32-345"><a href="#cb32-345" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">10</span>, sil_width, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">pch =</span> <span class="dv">19</span>,</span>
<span id="cb32-346"><a href="#cb32-346" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Number of Clusters"</span>,</span>
<span id="cb32-347"><a href="#cb32-347" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Average Silhouette Width"</span>,</span>
<span id="cb32-348"><a href="#cb32-348" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Silhouette Method"</span>)</span>
<span id="cb32-349"><a href="#cb32-349" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-350"><a href="#cb32-350" aria-hidden="true" tabindex="-1"></a><span class="co"># Detailed silhouette plot for K=3</span></span>
<span id="cb32-351"><a href="#cb32-351" aria-hidden="true" tabindex="-1"></a>km_3 <span class="ot">&lt;-</span> <span class="fu">kmeans</span>(sample_data, <span class="at">centers =</span> <span class="dv">3</span>, <span class="at">nstart =</span> <span class="dv">20</span>)</span>
<span id="cb32-352"><a href="#cb32-352" aria-hidden="true" tabindex="-1"></a>sil_3 <span class="ot">&lt;-</span> <span class="fu">silhouette</span>(km_3<span class="sc">$</span>cluster, <span class="fu">dist</span>(sample_data))</span>
<span id="cb32-353"><a href="#cb32-353" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(sil_3, <span class="at">col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>, <span class="at">main =</span> <span class="st">"Silhouette Plot (K=3)"</span>)</span>
<span id="cb32-354"><a href="#cb32-354" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-355"><a href="#cb32-355" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-356"><a href="#cb32-356" aria-hidden="true" tabindex="-1"></a><span class="fu">### Gap Statistic</span></span>
<span id="cb32-357"><a href="#cb32-357" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-358"><a href="#cb32-358" aria-hidden="true" tabindex="-1"></a>The gap statistic compares the within-cluster dispersion to that expected under a null reference distribution:</span>
<span id="cb32-359"><a href="#cb32-359" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-362"><a href="#cb32-362" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-363"><a href="#cb32-363" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb32-364"><a href="#cb32-364" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cluster)</span>
<span id="cb32-365"><a href="#cb32-365" aria-hidden="true" tabindex="-1"></a>gap_stat <span class="ot">&lt;-</span> <span class="fu">clusGap</span>(sample_data, <span class="at">FUN =</span> kmeans, <span class="at">nstart =</span> <span class="dv">25</span>, <span class="at">K.max =</span> <span class="dv">10</span>, <span class="at">B =</span> <span class="dv">50</span>)</span>
<span id="cb32-366"><a href="#cb32-366" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(gap_stat, <span class="at">main =</span> <span class="st">"Gap Statistic"</span>)</span>
<span id="cb32-367"><a href="#cb32-367" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-368"><a href="#cb32-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-369"><a href="#cb32-369" aria-hidden="true" tabindex="-1"></a><span class="fu">## Heatmaps</span></span>
<span id="cb32-370"><a href="#cb32-370" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-371"><a href="#cb32-371" aria-hidden="true" tabindex="-1"></a>**Heatmaps** are powerful visualization tools that combine clustering with color-coded data display. They're particularly useful for high-dimensional data like gene expression matrices.</span>
<span id="cb32-372"><a href="#cb32-372" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-373"><a href="#cb32-373" aria-hidden="true" tabindex="-1"></a><span class="fu">### Basic Heatmap Construction</span></span>
<span id="cb32-374"><a href="#cb32-374" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-375"><a href="#cb32-375" aria-hidden="true" tabindex="-1"></a>The idea:</span>
<span id="cb32-376"><a href="#cb32-376" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-377"><a href="#cb32-377" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Plot the data matrix as an image (colors represent values)</span>
<span id="cb32-378"><a href="#cb32-378" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Cluster rows (observations) and reorder by similarity</span>
<span id="cb32-379"><a href="#cb32-379" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Cluster columns (features) and reorder by similarity</span>
<span id="cb32-380"><a href="#cb32-380" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Display dendrograms showing the hierarchical structure</span>
<span id="cb32-381"><a href="#cb32-381" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-382"><a href="#cb32-382" aria-hidden="true" tabindex="-1"></a><span class="fu">### Heatmaps in R</span></span>
<span id="cb32-383"><a href="#cb32-383" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-386"><a href="#cb32-386" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-387"><a href="#cb32-387" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-heatmap-basic</span></span>
<span id="cb32-388"><a href="#cb32-388" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Heatmap of tissue gene expression data with hierarchical clustering on both rows and columns"</span></span>
<span id="cb32-389"><a href="#cb32-389" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb32-390"><a href="#cb32-390" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb32-391"><a href="#cb32-391" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"tissue_gene_expression"</span>)</span>
<span id="cb32-392"><a href="#cb32-392" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-393"><a href="#cb32-393" aria-hidden="true" tabindex="-1"></a><span class="co"># Center the data</span></span>
<span id="cb32-394"><a href="#cb32-394" aria-hidden="true" tabindex="-1"></a>x_tissue <span class="ot">&lt;-</span> <span class="fu">sweep</span>(tissue_gene_expression<span class="sc">$</span>x, <span class="dv">2</span>,</span>
<span id="cb32-395"><a href="#cb32-395" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">colMeans</span>(tissue_gene_expression<span class="sc">$</span>x))</span>
<span id="cb32-396"><a href="#cb32-396" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-397"><a href="#cb32-397" aria-hidden="true" tabindex="-1"></a><span class="co"># Basic heatmap (subsample for visibility)</span></span>
<span id="cb32-398"><a href="#cb32-398" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb32-399"><a href="#cb32-399" aria-hidden="true" tabindex="-1"></a>gene_sample <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">ncol</span>(x_tissue), <span class="dv">50</span>)</span>
<span id="cb32-400"><a href="#cb32-400" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, gene_sample],</span>
<span id="cb32-401"><a href="#cb32-401" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb32-402"><a href="#cb32-402" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">8</span>),</span>
<span id="cb32-403"><a href="#cb32-403" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>,</span>
<span id="cb32-404"><a href="#cb32-404" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Tissue Gene Expression"</span>)</span>
<span id="cb32-405"><a href="#cb32-405" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-406"><a href="#cb32-406" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-407"><a href="#cb32-407" aria-hidden="true" tabindex="-1"></a><span class="fu">### Customizing Heatmaps with Color Bars</span></span>
<span id="cb32-408"><a href="#cb32-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-409"><a href="#cb32-409" aria-hidden="true" tabindex="-1"></a>Adding color bars shows group membership:</span>
<span id="cb32-410"><a href="#cb32-410" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-413"><a href="#cb32-413" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-414"><a href="#cb32-414" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-heatmap-colorbar</span></span>
<span id="cb32-415"><a href="#cb32-415" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Heatmap with color sidebar indicating tissue type. Hierarchical clustering reveals that samples group by tissue."</span></span>
<span id="cb32-416"><a href="#cb32-416" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb32-417"><a href="#cb32-417" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb32-418"><a href="#cb32-418" aria-hidden="true" tabindex="-1"></a><span class="co"># Color palette for tissues</span></span>
<span id="cb32-419"><a href="#cb32-419" aria-hidden="true" tabindex="-1"></a>tissue_colors <span class="ot">&lt;-</span> <span class="fu">brewer.pal</span>(<span class="dv">7</span>, <span class="st">"Set2"</span>)[<span class="fu">as.numeric</span>(tissue_gene_expression<span class="sc">$</span>y)]</span>
<span id="cb32-420"><a href="#cb32-420" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-421"><a href="#cb32-421" aria-hidden="true" tabindex="-1"></a><span class="co"># Heatmap with row colors</span></span>
<span id="cb32-422"><a href="#cb32-422" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, gene_sample],</span>
<span id="cb32-423"><a href="#cb32-423" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb32-424"><a href="#cb32-424" aria-hidden="true" tabindex="-1"></a>        <span class="at">RowSideColors =</span> tissue_colors,</span>
<span id="cb32-425"><a href="#cb32-425" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">8</span>, <span class="dv">10</span>),</span>
<span id="cb32-426"><a href="#cb32-426" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>,</span>
<span id="cb32-427"><a href="#cb32-427" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Tissue Gene Expression"</span>)</span>
<span id="cb32-428"><a href="#cb32-428" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-429"><a href="#cb32-429" aria-hidden="true" tabindex="-1"></a><span class="co"># Add legend</span></span>
<span id="cb32-430"><a href="#cb32-430" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">levels</span>(tissue_gene_expression<span class="sc">$</span>y),</span>
<span id="cb32-431"><a href="#cb32-431" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">brewer.pal</span>(<span class="dv">7</span>, <span class="st">"Set2"</span>), <span class="at">cex =</span> <span class="fl">0.7</span>, <span class="at">bty =</span> <span class="st">"n"</span>)</span>
<span id="cb32-432"><a href="#cb32-432" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-433"><a href="#cb32-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-434"><a href="#cb32-434" aria-hidden="true" tabindex="-1"></a><span class="fu">### Filtering Features for Better Visualization</span></span>
<span id="cb32-435"><a href="#cb32-435" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-436"><a href="#cb32-436" aria-hidden="true" tabindex="-1"></a>Not all features contribute meaningful information to clustering. Filtering to high-variance features often improves visualization:</span>
<span id="cb32-437"><a href="#cb32-437" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-440"><a href="#cb32-440" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-441"><a href="#cb32-441" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-heatmap-filtered</span></span>
<span id="cb32-442"><a href="#cb32-442" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Heatmap of the 50 most variable genes. High-variance features often reveal the most interesting biological patterns."</span></span>
<span id="cb32-443"><a href="#cb32-443" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb32-444"><a href="#cb32-444" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 8</span></span>
<span id="cb32-445"><a href="#cb32-445" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(matrixStats)</span>
<span id="cb32-446"><a href="#cb32-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-447"><a href="#cb32-447" aria-hidden="true" tabindex="-1"></a><span class="co"># Select 50 most variable genes</span></span>
<span id="cb32-448"><a href="#cb32-448" aria-hidden="true" tabindex="-1"></a>sds <span class="ot">&lt;-</span> <span class="fu">colSds</span>(x_tissue, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb32-449"><a href="#cb32-449" aria-hidden="true" tabindex="-1"></a>top_genes <span class="ot">&lt;-</span> <span class="fu">order</span>(sds, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb32-450"><a href="#cb32-450" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-451"><a href="#cb32-451" aria-hidden="true" tabindex="-1"></a><span class="fu">heatmap</span>(x_tissue[, top_genes],</span>
<span id="cb32-452"><a href="#cb32-452" aria-hidden="true" tabindex="-1"></a>        <span class="at">col =</span> <span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>),</span>
<span id="cb32-453"><a href="#cb32-453" aria-hidden="true" tabindex="-1"></a>        <span class="at">RowSideColors =</span> tissue_colors,</span>
<span id="cb32-454"><a href="#cb32-454" aria-hidden="true" tabindex="-1"></a>        <span class="at">margins =</span> <span class="fu">c</span>(<span class="dv">10</span>, <span class="dv">10</span>),</span>
<span id="cb32-455"><a href="#cb32-455" aria-hidden="true" tabindex="-1"></a>        <span class="at">cexRow =</span> <span class="fl">0.7</span>, <span class="at">cexCol =</span> <span class="fl">0.6</span>,</span>
<span id="cb32-456"><a href="#cb32-456" aria-hidden="true" tabindex="-1"></a>        <span class="at">main =</span> <span class="st">"Top 50 Variable Genes"</span>)</span>
<span id="cb32-457"><a href="#cb32-457" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-458"><a href="#cb32-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-459"><a href="#cb32-459" aria-hidden="true" tabindex="-1"></a><span class="fu">### Advanced Heatmaps with pheatmap</span></span>
<span id="cb32-460"><a href="#cb32-460" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-461"><a href="#cb32-461" aria-hidden="true" tabindex="-1"></a>The <span class="in">`pheatmap`</span> package provides more control and better aesthetics:</span>
<span id="cb32-462"><a href="#cb32-462" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-465"><a href="#cb32-465" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-466"><a href="#cb32-466" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-pheatmap</span></span>
<span id="cb32-467"><a href="#cb32-467" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Enhanced heatmap using pheatmap with annotation tracks and customized appearance"</span></span>
<span id="cb32-468"><a href="#cb32-468" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb32-469"><a href="#cb32-469" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 7</span></span>
<span id="cb32-470"><a href="#cb32-470" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb32-471"><a href="#cb32-471" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb32-472"><a href="#cb32-472" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-473"><a href="#cb32-473" aria-hidden="true" tabindex="-1"></a><span class="co"># Prepare annotation</span></span>
<span id="cb32-474"><a href="#cb32-474" aria-hidden="true" tabindex="-1"></a>annotation_row <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb32-475"><a href="#cb32-475" aria-hidden="true" tabindex="-1"></a>  <span class="at">Tissue =</span> tissue_gene_expression<span class="sc">$</span>y,</span>
<span id="cb32-476"><a href="#cb32-476" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="fu">rownames</span>(x_tissue)</span>
<span id="cb32-477"><a href="#cb32-477" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb32-478"><a href="#cb32-478" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-479"><a href="#cb32-479" aria-hidden="true" tabindex="-1"></a><span class="fu">pheatmap</span>(x_tissue[, top_genes],</span>
<span id="cb32-480"><a href="#cb32-480" aria-hidden="true" tabindex="-1"></a>         <span class="at">color =</span> <span class="fu">colorRampPalette</span>(<span class="fu">brewer.pal</span>(<span class="dv">11</span>, <span class="st">"RdBu"</span>))(<span class="dv">100</span>),</span>
<span id="cb32-481"><a href="#cb32-481" aria-hidden="true" tabindex="-1"></a>         <span class="at">annotation_row =</span> annotation_row,</span>
<span id="cb32-482"><a href="#cb32-482" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_rownames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb32-483"><a href="#cb32-483" aria-hidden="true" tabindex="-1"></a>         <span class="at">show_colnames =</span> <span class="cn">FALSE</span>,</span>
<span id="cb32-484"><a href="#cb32-484" aria-hidden="true" tabindex="-1"></a>         <span class="at">main =</span> <span class="st">"Tissue Gene Expression Heatmap"</span>,</span>
<span id="cb32-485"><a href="#cb32-485" aria-hidden="true" tabindex="-1"></a>         <span class="at">fontsize =</span> <span class="dv">8</span>)</span>
<span id="cb32-486"><a href="#cb32-486" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-487"><a href="#cb32-487" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-488"><a href="#cb32-488" aria-hidden="true" tabindex="-1"></a><span class="fu">## Practical Considerations</span></span>
<span id="cb32-489"><a href="#cb32-489" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-490"><a href="#cb32-490" aria-hidden="true" tabindex="-1"></a><span class="fu">### Standardization</span></span>
<span id="cb32-491"><a href="#cb32-491" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-492"><a href="#cb32-492" aria-hidden="true" tabindex="-1"></a>Features on different scales should be standardized before clustering:</span>
<span id="cb32-493"><a href="#cb32-493" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-496"><a href="#cb32-496" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-497"><a href="#cb32-497" aria-hidden="true" tabindex="-1"></a><span class="co"># Standardize features</span></span>
<span id="cb32-498"><a href="#cb32-498" aria-hidden="true" tabindex="-1"></a>x_scaled <span class="ot">&lt;-</span> <span class="fu">scale</span>(x_tissue)</span>
<span id="cb32-499"><a href="#cb32-499" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-500"><a href="#cb32-500" aria-hidden="true" tabindex="-1"></a><span class="co"># This ensures all features contribute equally to distance calculations</span></span>
<span id="cb32-501"><a href="#cb32-501" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-502"><a href="#cb32-502" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-503"><a href="#cb32-503" aria-hidden="true" tabindex="-1"></a><span class="fu">### Handling Missing Data</span></span>
<span id="cb32-504"><a href="#cb32-504" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-505"><a href="#cb32-505" aria-hidden="true" tabindex="-1"></a>K-means cannot handle missing values. Options include:</span>
<span id="cb32-506"><a href="#cb32-506" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-507"><a href="#cb32-507" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Remove observations or features with missing data</span>
<span id="cb32-508"><a href="#cb32-508" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Impute missing values</span>
<span id="cb32-509"><a href="#cb32-509" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Use algorithms that handle missingness (some hierarchical methods)</span>
<span id="cb32-510"><a href="#cb32-510" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-511"><a href="#cb32-511" aria-hidden="true" tabindex="-1"></a><span class="fu">### Cluster Validation</span></span>
<span id="cb32-512"><a href="#cb32-512" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-513"><a href="#cb32-513" aria-hidden="true" tabindex="-1"></a>How do we know if clusters are "real"? Several approaches:</span>
<span id="cb32-514"><a href="#cb32-514" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-515"><a href="#cb32-515" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Stability**: Do clusters persist across different subsamples?</span>
<span id="cb32-516"><a href="#cb32-516" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**External validation**: Do clusters correspond to known groups?</span>
<span id="cb32-517"><a href="#cb32-517" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Internal metrics**: Silhouette scores, within-cluster variance</span>
<span id="cb32-518"><a href="#cb32-518" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-521"><a href="#cb32-521" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb32-522"><a href="#cb32-522" aria-hidden="true" tabindex="-1"></a><span class="co"># Compare clusters to known tissue labels</span></span>
<span id="cb32-523"><a href="#cb32-523" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="at">Cluster =</span> km_3<span class="sc">$</span>cluster, <span class="at">True =</span> sample_data<span class="sc">$</span>x <span class="sc">&gt;</span> <span class="dv">2</span>)</span>
<span id="cb32-524"><a href="#cb32-524" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb32-525"><a href="#cb32-525" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-526"><a href="#cb32-526" aria-hidden="true" tabindex="-1"></a><span class="fu">### Reproducibility</span></span>
<span id="cb32-527"><a href="#cb32-527" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-528"><a href="#cb32-528" aria-hidden="true" tabindex="-1"></a>K-means depends on random initialization. Always:</span>
<span id="cb32-529"><a href="#cb32-529" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-530"><a href="#cb32-530" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Set a seed for reproducibility</span>
<span id="cb32-531"><a href="#cb32-531" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Use <span class="in">`nstart`</span> to run multiple initializations</span>
<span id="cb32-532"><a href="#cb32-532" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Report the random seed used</span>
<span id="cb32-533"><a href="#cb32-533" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-534"><a href="#cb32-534" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercises</span></span>
<span id="cb32-535"><a href="#cb32-535" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-536"><a href="#cb32-536" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb32-537"><a href="#cb32-537" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise Clu.1: Distance Computation</span></span>
<span id="cb32-538"><a href="#cb32-538" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-539"><a href="#cb32-539" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Load the <span class="in">`tissue_gene_expression`</span> dataset. Remove the row means and compute the distance between each observation. Store the result in <span class="in">`d`</span>.</span>
<span id="cb32-540"><a href="#cb32-540" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-541"><a href="#cb32-541" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-542"><a href="#cb32-542" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb32-543"><a href="#cb32-543" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise Clu.2: Hierarchical Clustering</span></span>
<span id="cb32-544"><a href="#cb32-544" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-545"><a href="#cb32-545" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Make a hierarchical clustering plot and add the tissue types as labels. Which tissues cluster together?</span>
<span id="cb32-546"><a href="#cb32-546" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-547"><a href="#cb32-547" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-548"><a href="#cb32-548" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb32-549"><a href="#cb32-549" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise Clu.3: K-Means Clustering</span></span>
<span id="cb32-550"><a href="#cb32-550" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-551"><a href="#cb32-551" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>Run k-means clustering on the data with $K = 7$. Make a table comparing the identified clusters to the actual tissue types. Run the algorithm several times to see how the answer changes.</span>
<span id="cb32-552"><a href="#cb32-552" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-553"><a href="#cb32-553" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-554"><a href="#cb32-554" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb32-555"><a href="#cb32-555" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise Clu.4: Heatmap Visualization</span></span>
<span id="cb32-556"><a href="#cb32-556" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-557"><a href="#cb32-557" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Select the 50 most variable genes. Create a heatmap with:</span>
<span id="cb32-558"><a href="#cb32-558" aria-hidden="true" tabindex="-1"></a><span class="ss">   - </span>Observations in columns</span>
<span id="cb32-559"><a href="#cb32-559" aria-hidden="true" tabindex="-1"></a><span class="ss">   - </span>Centered predictors</span>
<span id="cb32-560"><a href="#cb32-560" aria-hidden="true" tabindex="-1"></a><span class="ss">   - </span>A color bar showing tissue types</span>
<span id="cb32-561"><a href="#cb32-561" aria-hidden="true" tabindex="-1"></a><span class="ss">   - </span>Use <span class="in">`ColSideColors`</span> argument and <span class="in">`col = brewer.pal(11, "RdBu")`</span></span>
<span id="cb32-562"><a href="#cb32-562" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-563"><a href="#cb32-563" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-564"><a href="#cb32-564" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb32-565"><a href="#cb32-565" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise Clu.5: Comparing Linkage Methods</span></span>
<span id="cb32-566"><a href="#cb32-566" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-567"><a href="#cb32-567" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Compare different linkage methods on the movie data. Which produces the most interpretable clusters?</span>
<span id="cb32-568"><a href="#cb32-568" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb32-569"><a href="#cb32-569" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-570"><a href="#cb32-570" aria-hidden="true" tabindex="-1"></a><span class="fu">## K-Means vs Hierarchical Clustering</span></span>
<span id="cb32-571"><a href="#cb32-571" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-572"><a href="#cb32-572" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Aspect <span class="pp">|</span> K-Means <span class="pp">|</span> Hierarchical <span class="pp">|</span></span>
<span id="cb32-573"><a href="#cb32-573" aria-hidden="true" tabindex="-1"></a><span class="pp">|:-------|:--------|:-------------|</span></span>
<span id="cb32-574"><a href="#cb32-574" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Clusters <span class="pp">|</span> Must specify K <span class="pp">|</span> Can cut at any level <span class="pp">|</span></span>
<span id="cb32-575"><a href="#cb32-575" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Structure <span class="pp">|</span> Flat partition <span class="pp">|</span> Nested hierarchy <span class="pp">|</span></span>
<span id="cb32-576"><a href="#cb32-576" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Scalability <span class="pp">|</span> O(nKt) - efficient <span class="pp">|</span> O(n³) - slower <span class="pp">|</span></span>
<span id="cb32-577"><a href="#cb32-577" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Shape <span class="pp">|</span> Spherical clusters <span class="pp">|</span> Arbitrary shapes <span class="pp">|</span></span>
<span id="cb32-578"><a href="#cb32-578" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Missing data <span class="pp">|</span> Cannot handle <span class="pp">|</span> Some methods can <span class="pp">|</span></span>
<span id="cb32-579"><a href="#cb32-579" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Reproducibility <span class="pp">|</span> Depends on initialization <span class="pp">|</span> Deterministic <span class="pp">|</span></span>
<span id="cb32-580"><a href="#cb32-580" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-581"><a href="#cb32-581" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb32-582"><a href="#cb32-582" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-583"><a href="#cb32-583" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Unsupervised learning** discovers structure without predefined labels</span>
<span id="cb32-584"><a href="#cb32-584" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Hierarchical clustering** builds a tree showing nested relationships</span>
<span id="cb32-585"><a href="#cb32-585" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Dendrograms visualize the hierarchy</span>
<span id="cb32-586"><a href="#cb32-586" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Linkage methods define cluster-to-cluster distance</span>
<span id="cb32-587"><a href="#cb32-587" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Cut the tree to form discrete clusters</span>
<span id="cb32-588"><a href="#cb32-588" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**K-means clustering** partitions data into K groups</span>
<span id="cb32-589"><a href="#cb32-589" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Minimizes within-cluster variance</span>
<span id="cb32-590"><a href="#cb32-590" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Requires specifying K in advance</span>
<span id="cb32-591"><a href="#cb32-591" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Use multiple starts to avoid local optima</span>
<span id="cb32-592"><a href="#cb32-592" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Heatmaps** combine clustering with data visualization</span>
<span id="cb32-593"><a href="#cb32-593" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Color encodes values</span>
<span id="cb32-594"><a href="#cb32-594" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Rows and columns reordered by similarity</span>
<span id="cb32-595"><a href="#cb32-595" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Filter to high-variance features for clarity</span>
<span id="cb32-596"><a href="#cb32-596" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Choosing K** requires judgment and multiple criteria</span>
<span id="cb32-597"><a href="#cb32-597" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Elbow method: look for bend in WSS curve</span>
<span id="cb32-598"><a href="#cb32-598" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Silhouette analysis: measure cluster coherence</span>
<span id="cb32-599"><a href="#cb32-599" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Domain knowledge should guide interpretation</span>
<span id="cb32-600"><a href="#cb32-600" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-601"><a href="#cb32-601" aria-hidden="true" tabindex="-1"></a><span class="fu">## Additional Resources</span></span>
<span id="cb32-602"><a href="#cb32-602" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb32-603"><a href="#cb32-603" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>@james2023islr - Comprehensive treatment of clustering methods</span>
<span id="cb32-604"><a href="#cb32-604" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>@logan2010biostatistical - Applications in biological research</span>
<span id="cb32-605"><a href="#cb32-605" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Hahsler, M., Piekenbrock, M., &amp; Doran, D. (2019). dbscan: Fast Density-Based Clustering with R</span>
<span id="cb32-606"><a href="#cb32-606" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`pheatmap`</span> and <span class="in">`ComplexHeatmap`</span> package documentation for advanced heatmaps</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Statistics for Biosciences and Bioengineering</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Built with <a href="https://quarto.org/">Quarto</a></p>
</div>
  </div>
</footer>




</body></html>