<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.30">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>39&nbsp; Non-Linear Dimensionality Reduction: t-SNE and UMAP – Statistics for the Biosciences and Bioengineering</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
  margin-bottom: 0em;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<link href="../chapters/39-bayesian-statistics.html" rel="next">
<link href="../chapters/37-dimensionality-reduction.html" rel="prev">
<script src="../site_libs/quarto-html/quarto.js" type="module"></script>
<script src="../site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting-de070a7b0ab54f8780927367ac907214.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap-485d01fc63b59abcd3ee1bf1e8e2748d.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://cdnjs.cloudflare.com/polyfill/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

<script type="text/javascript">
const typesetMath = (el) => {
  if (window.MathJax) {
    // MathJax Typeset
    window.MathJax.typeset([el]);
  } else if (window.katex) {
    // KaTeX Render
    var mathElements = el.getElementsByClassName("math");
    var macros = [];
    for (var i = 0; i < mathElements.length; i++) {
      var texText = mathElements[i].firstChild;
      if (mathElements[i].tagName == "SPAN") {
        window.katex.render(texText.data, mathElements[i], {
          displayMode: mathElements[i].classList.contains('display'),
          throwOnError: false,
          macros: macros,
          fleqn: false
        });
      }
    }
  }
}
window.Quarto = {
  typesetMath
};
</script>

</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/29-intro-statistical-learning.html">Statistical Learning</a></li><li class="breadcrumb-item"><a href="../chapters/38-tsne-umap.html"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Non-Linear Dimensionality Reduction: t-SNE and UMAP</span></a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="../">Statistics for the Biosciences and Bioengineering</a> 
        <div class="sidebar-tools-main">
  <a href="" class="quarto-reader-toggle quarto-navigation-tool px-1" onclick="window.quartoToggleReader(); return false;" title="Toggle reader mode">
  <div class="quarto-reader-toggle-btn">
  <i class="bi"></i>
  </div>
</a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Why This Book?</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true">
 <span class="menu-text">Core Data Science Tools</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/01-introduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/02-installing-tools.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Installing Core Tools</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/03-unix-command-line.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Unix and the Command Line</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/04-r-rstudio.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">R and RStudio</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/05-markdown-latex.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Markdown and LaTeX</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true">
 <span class="menu-text">Data Exploration</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/06-tidy-data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Tidy Data</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/07-data-wrangling.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Data Wrangling with dplyr</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/08-exploratory-data-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Exploratory Data Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/09-data-visualization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Data Visualization</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true">
 <span class="menu-text">Probability and Distributions</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/10-probability-foundations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Foundations of Probability</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/11-discrete-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Discrete Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/12-continuous-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Continuous Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/13-sampling-estimation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Sampling and Parameter Estimation</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Inference</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/14-experimental-design.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Experimental Design Principles</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/15-hypothesis-testing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/16-t-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">T-Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/17-nonparametric-tests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Nonparametric Tests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/18-bootstrapping.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Resampling Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/19-presenting-results.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">20</span>&nbsp; <span class="chapter-title">Presenting Statistical Results</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true">
 <span class="menu-text">Linear Models</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/20-what-are-models.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">21</span>&nbsp; <span class="chapter-title">What are Models?</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/21-correlation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">22</span>&nbsp; <span class="chapter-title">Correlation</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/22-simple-linear-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">23</span>&nbsp; <span class="chapter-title">Simple Linear Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/23-residual-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">24</span>&nbsp; <span class="chapter-title">Residual Analysis</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/24-statistical-power.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">25</span>&nbsp; <span class="chapter-title">Statistical Power</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/25-multiple-regression.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">26</span>&nbsp; <span class="chapter-title">Multiple Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/26-single-factor-anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">27</span>&nbsp; <span class="chapter-title">Single Factor ANOVA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/27-multifactor-anova.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">28</span>&nbsp; <span class="chapter-title">Multi-Factor ANOVA</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/28-glm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">29</span>&nbsp; <span class="chapter-title">Generalized Linear Models</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true">
 <span class="menu-text">Statistical Learning</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-6" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-6" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/29-intro-statistical-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">30</span>&nbsp; <span class="chapter-title">Introduction to Statistical Learning</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/30-model-validation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">31</span>&nbsp; <span class="chapter-title">Model Validation and the Bias-Variance Tradeoff</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/31-regularization.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">32</span>&nbsp; <span class="chapter-title">Regularization Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/32-smoothing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">33</span>&nbsp; <span class="chapter-title">Smoothing and Non-Parametric Regression</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/33-classification-methods.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">34</span>&nbsp; <span class="chapter-title">Classification and Performance Metrics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/34-trees-forests.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">35</span>&nbsp; <span class="chapter-title">Decision Trees and Random Forests</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/35-svm.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">36</span>&nbsp; <span class="chapter-title">Support Vector Machines</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/36-clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">37</span>&nbsp; <span class="chapter-title">Clustering</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/37-dimensionality-reduction.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/38-tsne-umap.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Non-Linear Dimensionality Reduction: t-SNE and UMAP</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/39-bayesian-statistics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">40</span>&nbsp; <span class="chapter-title">Bayesian Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/40-deep-learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">41</span>&nbsp; <span class="chapter-title">Introduction to Deep Learning</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-7" role="navigation" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-7" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A1-eugenics-history.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">42</span>&nbsp; <span class="chapter-title">The Eugenics History of Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A2-keyboard-shortcuts.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">43</span>&nbsp; <span class="chapter-title">Keyboard Shortcuts Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A3-unix-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">44</span>&nbsp; <span class="chapter-title">Unix Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A4-r-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">45</span>&nbsp; <span class="chapter-title">R Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A5-quarto-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">46</span>&nbsp; <span class="chapter-title">Quarto Markdown Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A6-latex-reference.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">47</span>&nbsp; <span class="chapter-title">LaTeX Command Reference</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A7-greek-letters.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">48</span>&nbsp; <span class="chapter-title">Greek Letters in Mathematics and Statistics</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A8-probability-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">49</span>&nbsp; <span class="chapter-title">Common Probability Distributions</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A9-sampling-distributions.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">50</span>&nbsp; <span class="chapter-title">Sampling Distributions in Hypothesis Testing</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A10-matrix-algebra.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">51</span>&nbsp; <span class="chapter-title">Matrix Algebra Fundamentals</span></span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../chapters/A11-high-performance-computing.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">52</span>&nbsp; <span class="chapter-title">High Performance Computing</span></span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#beyond-linear-methods" id="toc-beyond-linear-methods" class="nav-link active" data-scroll-target="#beyond-linear-methods"><span class="header-section-number">39.1</span> Beyond Linear Methods</a></li>
  <li><a href="#limitations-of-pca" id="toc-limitations-of-pca" class="nav-link" data-scroll-target="#limitations-of-pca"><span class="header-section-number">39.2</span> Limitations of PCA</a></li>
  <li><a href="#t-sne-t-distributed-stochastic-neighbor-embedding" id="toc-t-sne-t-distributed-stochastic-neighbor-embedding" class="nav-link" data-scroll-target="#t-sne-t-distributed-stochastic-neighbor-embedding"><span class="header-section-number">39.3</span> t-SNE: t-distributed Stochastic Neighbor Embedding</a>
  <ul class="collapse">
  <li><a href="#how-t-sne-works" id="toc-how-t-sne-works" class="nav-link" data-scroll-target="#how-t-sne-works">How t-SNE Works</a></li>
  <li><a href="#the-perplexity-parameter" id="toc-the-perplexity-parameter" class="nav-link" data-scroll-target="#the-perplexity-parameter">The Perplexity Parameter</a></li>
  <li><a href="#t-sne-in-r" id="toc-t-sne-in-r" class="nav-link" data-scroll-target="#t-sne-in-r">t-SNE in R</a></li>
  <li><a href="#important-considerations-for-t-sne" id="toc-important-considerations-for-t-sne" class="nav-link" data-scroll-target="#important-considerations-for-t-sne">Important Considerations for t-SNE</a></li>
  </ul></li>
  <li><a href="#umap-uniform-manifold-approximation-and-projection" id="toc-umap-uniform-manifold-approximation-and-projection" class="nav-link" data-scroll-target="#umap-uniform-manifold-approximation-and-projection"><span class="header-section-number">39.4</span> UMAP: Uniform Manifold Approximation and Projection</a>
  <ul class="collapse">
  <li><a href="#how-umap-works" id="toc-how-umap-works" class="nav-link" data-scroll-target="#how-umap-works">How UMAP Works</a></li>
  <li><a href="#key-umap-parameters" id="toc-key-umap-parameters" class="nav-link" data-scroll-target="#key-umap-parameters">Key UMAP Parameters</a></li>
  <li><a href="#umap-in-r" id="toc-umap-in-r" class="nav-link" data-scroll-target="#umap-in-r">UMAP in R</a></li>
  <li><a href="#effect-of-umap-parameters" id="toc-effect-of-umap-parameters" class="nav-link" data-scroll-target="#effect-of-umap-parameters">Effect of UMAP Parameters</a></li>
  </ul></li>
  <li><a href="#comparing-pca-t-sne-and-umap" id="toc-comparing-pca-t-sne-and-umap" class="nav-link" data-scroll-target="#comparing-pca-t-sne-and-umap"><span class="header-section-number">39.5</span> Comparing PCA, t-SNE, and UMAP</a>
  <ul class="collapse">
  <li><a href="#key-differences" id="toc-key-differences" class="nav-link" data-scroll-target="#key-differences">Key Differences</a></li>
  <li><a href="#when-to-use-each-method" id="toc-when-to-use-each-method" class="nav-link" data-scroll-target="#when-to-use-each-method">When to Use Each Method</a></li>
  </ul></li>
  <li><a href="#biological-applications" id="toc-biological-applications" class="nav-link" data-scroll-target="#biological-applications"><span class="header-section-number">39.6</span> Biological Applications</a>
  <ul class="collapse">
  <li><a href="#single-cell-rna-seq" id="toc-single-cell-rna-seq" class="nav-link" data-scroll-target="#single-cell-rna-seq">Single-Cell RNA-seq</a></li>
  </ul></li>
  <li><a href="#practical-workflow" id="toc-practical-workflow" class="nav-link" data-scroll-target="#practical-workflow"><span class="header-section-number">39.7</span> Practical Workflow</a></li>
  <li><a href="#exercises" id="toc-exercises" class="nav-link" data-scroll-target="#exercises"><span class="header-section-number">39.8</span> Exercises</a></li>
  <li><a href="#summary" id="toc-summary" class="nav-link" data-scroll-target="#summary"><span class="header-section-number">39.9</span> Summary</a></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="header-section-number">39.10</span> Additional Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default"><nav class="quarto-page-breadcrumbs quarto-title-breadcrumbs d-none d-lg-block" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../chapters/29-intro-statistical-learning.html">Statistical Learning</a></li><li class="breadcrumb-item"><a href="../chapters/38-tsne-umap.html"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Non-Linear Dimensionality Reduction: t-SNE and UMAP</span></a></li></ol></nav>
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title"><span id="sec-tsne-umap" class="quarto-section-identifier"><span class="chapter-number">39</span>&nbsp; <span class="chapter-title">Non-Linear Dimensionality Reduction: t-SNE and UMAP</span></span></h1><button type="button" class="btn code-tools-button dropdown-toggle" id="quarto-code-tools-menu" data-bs-toggle="dropdown" aria-expanded="false"><i class="bi"></i> Code</button><ul class="dropdown-menu dropdown-menu-end" aria-labelelledby="quarto-code-tools-menu"><li><a id="quarto-show-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Show All Code</a></li><li><a id="quarto-hide-all-code" class="dropdown-item" href="javascript:void(0)" role="button">Hide All Code</a></li><li><hr class="dropdown-divider"></li><li><a id="quarto-view-source" class="dropdown-item" href="javascript:void(0)" role="button">View Source</a></li></ul></div></div>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<div class="cell">
<div class="cell-output cell-output-error">
<pre><code>Error in `library()`:
! there is no package called 'umap'</code></pre>
</div>
</div>
<section id="beyond-linear-methods" class="level2" data-number="39.1">
<h2 data-number="39.1" class="anchored" data-anchor-id="beyond-linear-methods"><span class="header-section-number">39.1</span> Beyond Linear Methods</h2>
<p>In <a href="37-dimensionality-reduction.html" class="quarto-xref"><span>Chapter 38</span></a>, we introduced Principal Component Analysis (PCA) as a powerful tool for dimensionality reduction. PCA works by finding linear combinations of variables that maximize variance. However, PCA has a fundamental limitation: it can only capture <strong>linear</strong> relationships.</p>
<p>Real biological data often contains complex, non-linear structure:</p>
<ul>
<li>Cell populations in single-cell RNA-seq form distinct clusters with non-linear boundaries</li>
<li>Developmental trajectories follow curved paths through gene expression space</li>
<li>Protein structures exhibit complex geometric relationships</li>
</ul>
<p>When data lies on a curved <strong>manifold</strong> in high-dimensional space, linear methods like PCA may fail to reveal the true structure. This chapter introduces two powerful non-linear methods: <strong>t-SNE</strong> and <strong>UMAP</strong>.</p>
</section>
<section id="limitations-of-pca" class="level2" data-number="39.2">
<h2 data-number="39.2" class="anchored" data-anchor-id="limitations-of-pca"><span class="header-section-number">39.2</span> Limitations of PCA</h2>
<p>To see where PCA falls short, consider a simple example: data arranged in a curved pattern.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate data on a curved manifold (Swiss roll-like)</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">0</span>, <span class="dv">4</span><span class="sc">*</span>pi)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> t <span class="sc">*</span> <span class="fu">cos</span>(t)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> t <span class="sc">*</span> <span class="fu">sin</span>(t)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">0</span>, <span class="dv">10</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>color <span class="ot">&lt;-</span> t  <span class="co"># Color by position along curve</span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>curved_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">z =</span> z, <span class="at">t =</span> t)</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>pca_result <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(curved_data[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a><span class="co"># Original 3D projection</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(curved_data<span class="sc">$</span>x, curved_data<span class="sc">$</span>y, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"X"</span>, <span class="at">ylab =</span> <span class="st">"Y"</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Original Data (X-Y projection)"</span>)</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA result</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_result<span class="sc">$</span>x[, <span class="dv">1</span>], pca_result<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>,</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a><span class="co"># What we want: unfolded</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(t, z, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"Position along curve"</span>, <span class="at">ylab =</span> <span class="st">"Z"</span>,</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Ideal: Unfolded Manifold"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-pca-limitation" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pca-limitation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-pca-limitation-1.png" class="img-fluid figure-img" width="864">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pca-limitation-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.1: PCA captures the direction of maximum variance but fails to ‘unfold’ the curved structure. The first PC spreads points along the curve but mixes points from different parts.
</figcaption>
</figure>
</div>
</div>
</div>
<p>PCA projects the data onto the directions of maximum variance, but it cannot “unfold” the curved structure. Points that are far apart along the manifold (different colors) may end up close together in the PCA projection.</p>
</section>
<section id="t-sne-t-distributed-stochastic-neighbor-embedding" class="level2" data-number="39.3">
<h2 data-number="39.3" class="anchored" data-anchor-id="t-sne-t-distributed-stochastic-neighbor-embedding"><span class="header-section-number">39.3</span> t-SNE: t-distributed Stochastic Neighbor Embedding</h2>
<p><strong>t-SNE</strong> (t-distributed Stochastic Neighbor Embedding) <span class="citation" data-cites="vandermaaten2008visualizing">(<a href="../references.html#ref-vandermaaten2008visualizing" role="doc-biblioref">Maaten and Hinton 2008</a>)</span> is a non-linear dimensionality reduction technique designed specifically for visualization. It was introduced by van der Maaten and Hinton in 2008 and has become one of the most popular methods for visualizing high-dimensional data.</p>
<section id="how-t-sne-works" class="level3">
<h3 class="anchored" data-anchor-id="how-t-sne-works">How t-SNE Works</h3>
<p>The core idea of t-SNE is to preserve <strong>local neighborhood structure</strong>. Points that are close together in high-dimensional space should remain close together in the low-dimensional embedding.</p>
<p>The algorithm works in two steps:</p>
<p><strong>Step 1: Define similarities in high-dimensional space</strong></p>
<p>For each pair of points <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span>, t-SNE computes a probability <span class="math inline">\(p_{ij}\)</span> that represents how likely point <span class="math inline">\(j\)</span> would be picked as a neighbor of point <span class="math inline">\(i\)</span> if neighbors were chosen in proportion to their probability density under a Gaussian centered at <span class="math inline">\(i\)</span>:</p>
<p><span class="math display">\[
p_{j|i} = \frac{\exp(-\|x_i - x_j\|^2 / 2\sigma_i^2)}{\sum_{k \neq i} \exp(-\|x_i - x_k\|^2 / 2\sigma_i^2)}
\]</span></p>
<p>The variance <span class="math inline">\(\sigma_i\)</span> is chosen so that each point has a fixed number of effective neighbors (controlled by the <strong>perplexity</strong> parameter).</p>
<p><strong>Step 2: Define similarities in low-dimensional space</strong></p>
<p>In the low-dimensional embedding, t-SNE uses a <strong>t-distribution</strong> (with one degree of freedom, i.e., a Cauchy distribution) to compute similarities:</p>
<p><span class="math display">\[
q_{ij} = \frac{(1 + \|y_i - y_j\|^2)^{-1}}{\sum_{k \neq l} (1 + \|y_k - y_l\|^2)^{-1}}
\]</span></p>
<p>The t-distribution has heavier tails than a Gaussian, which allows dissimilar points to be placed far apart without incurring a large penalty.</p>
<p><strong>Step 3: Minimize the divergence</strong></p>
<p>t-SNE then minimizes the Kullback-Leibler divergence between the two distributions using gradient descent:</p>
<p><span class="math display">\[
KL(P \| Q) = \sum_{i \neq j} p_{ij} \log \frac{p_{ij}}{q_{ij}}
\]</span></p>
<p>This pushes the low-dimensional embedding to match the neighborhood structure of the high-dimensional data.</p>
</section>
<section id="the-perplexity-parameter" class="level3">
<h3 class="anchored" data-anchor-id="the-perplexity-parameter">The Perplexity Parameter</h3>
<p><strong>Perplexity</strong> is the most important parameter in t-SNE. It loosely corresponds to the number of effective nearest neighbors considered for each point. Typical values range from 5 to 50.</p>
<ul>
<li><strong>Low perplexity</strong>: Focuses on very local structure; may create many small, disconnected clusters</li>
<li><strong>High perplexity</strong>: Considers more global structure; clusters may merge together</li>
</ul>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Use MNIST digits for demonstration</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"mnist"</span>)) mnist <span class="ot">&lt;-</span> <span class="fu">read_mnist</span>()</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample for speed</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mnist<span class="sc">$</span>train<span class="sc">$</span>images), <span class="dv">2000</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>x_sample <span class="ot">&lt;-</span> mnist<span class="sc">$</span>train<span class="sc">$</span>images[idx, ]</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> mnist<span class="sc">$</span>train<span class="sc">$</span>labels[idx]</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a><span class="co"># t-SNE with different perplexities</span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>tsne_5 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">5</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>tsne_30 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>tsne_100 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_5<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 5"</span>)</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_30<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 30"</span>)</span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_100<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 100"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-tsne-perplexity" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-tsne-perplexity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-tsne-perplexity-1.png" class="img-fluid figure-img" width="864">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-tsne-perplexity-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.2: Effect of perplexity on t-SNE embeddings. Low perplexity emphasizes local structure (may fragment clusters), high perplexity emphasizes global structure (may merge clusters).
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="t-sne-in-r" class="level3">
<h3 class="anchored" data-anchor-id="t-sne-in-r">t-SNE in R</h3>
<p>The <code>Rtsne</code> package provides an efficient implementation:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Full t-SNE with good parameters</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>tsne_result <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">1000</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with digit labels</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>df_tsne <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">tSNE1 =</span> tsne_result<span class="sc">$</span>Y[, <span class="dv">1</span>],</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">tSNE2 =</span> tsne_result<span class="sc">$</span>Y[, <span class="dv">2</span>],</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">digit =</span> <span class="fu">factor</span>(labels)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_tsne, <span class="fu">aes</span>(tSNE1, tSNE2, <span class="at">color =</span> digit)) <span class="sc">+</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Spectral"</span>) <span class="sc">+</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"t-SNE of MNIST Digits"</span>,</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"t-SNE Dimension 1"</span>, <span class="at">y =</span> <span class="st">"t-SNE Dimension 2"</span>) <span class="sc">+</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-tsne-mnist" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-tsne-mnist-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-tsne-mnist-1.png" class="img-fluid figure-img" width="672">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-tsne-mnist-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.3: t-SNE visualization of MNIST digits reveals clear cluster structure corresponding to digit classes
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="important-considerations-for-t-sne" class="level3">
<h3 class="anchored" data-anchor-id="important-considerations-for-t-sne">Important Considerations for t-SNE</h3>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
t-SNE Caveats
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><p><strong>Cluster sizes are meaningless</strong>: t-SNE does not preserve density. A large cluster in t-SNE does not mean more points.</p></li>
<li><p><strong>Distances between clusters are meaningless</strong>: Only local structure is preserved. The distance between two clusters tells you nothing about their true similarity.</p></li>
<li><p><strong>Results depend on random initialization</strong>: Always set a seed for reproducibility. Run multiple times to check stability.</p></li>
<li><p><strong>Perplexity must be tuned</strong>: There’s no universally best value. Try multiple perplexities to understand your data.</p></li>
<li><p><strong>Computationally expensive</strong>: O(n²) complexity; can be slow for large datasets. Use PCA preprocessing to speed up.</p></li>
</ol>
</div>
</div>
</section>
</section>
<section id="umap-uniform-manifold-approximation-and-projection" class="level2" data-number="39.4">
<h2 data-number="39.4" class="anchored" data-anchor-id="umap-uniform-manifold-approximation-and-projection"><span class="header-section-number">39.4</span> UMAP: Uniform Manifold Approximation and Projection</h2>
<p><strong>UMAP</strong> (Uniform Manifold Approximation and Projection) <span class="citation" data-cites="mcinnes2018umap">(<a href="../references.html#ref-mcinnes2018umap" role="doc-biblioref">McInnes, Healy, and Melville 2018</a>)</span> is a more recent algorithm that has become increasingly popular, especially in single-cell genomics. UMAP is often faster than t-SNE and may better preserve global structure.</p>
<section id="how-umap-works" class="level3">
<h3 class="anchored" data-anchor-id="how-umap-works">How UMAP Works</h3>
<p>UMAP is grounded in Riemannian geometry and algebraic topology, but the practical intuition is similar to t-SNE: preserve neighborhood relationships.</p>
<p>Key differences from t-SNE:</p>
<ol type="1">
<li><p><strong>Mathematical foundation</strong>: UMAP assumes the data lies on a uniformly distributed manifold and uses concepts from topological data analysis.</p></li>
<li><p><strong>Graph-based approach</strong>: UMAP constructs a weighted graph representing the high-dimensional data, then optimizes a low-dimensional graph to match it.</p></li>
<li><p><strong>Different cost function</strong>: UMAP uses cross-entropy rather than KL divergence, which may help preserve more global structure.</p></li>
<li><p><strong>Faster</strong>: UMAP typically runs faster than t-SNE, especially on large datasets.</p></li>
</ol>
</section>
<section id="key-umap-parameters" class="level3">
<h3 class="anchored" data-anchor-id="key-umap-parameters">Key UMAP Parameters</h3>
<ul>
<li><strong>n_neighbors</strong>: Similar to perplexity in t-SNE. Controls the balance between local and global structure. Typical values: 5-50.</li>
<li><strong>min_dist</strong>: Controls how tightly UMAP packs points together. Small values create tighter clusters. Typical values: 0.0-0.5.</li>
<li><strong>metric</strong>: Distance metric used. Default is Euclidean, but many options available.</li>
</ul>
</section>
<section id="umap-in-r" class="level3">
<h3 class="anchored" data-anchor-id="umap-in-r">UMAP in R</h3>
<p>The <code>umap</code> package provides a convenient interface:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>umap_result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> <span class="dv">15</span>, <span class="at">min_dist =</span> <span class="fl">0.1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error in `umap()`:
! could not find function "umap"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>df_umap <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">UMAP1 =</span> umap_result<span class="sc">$</span>layout[, <span class="dv">1</span>],</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">UMAP2 =</span> umap_result<span class="sc">$</span>layout[, <span class="dv">2</span>],</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">digit =</span> <span class="fu">factor</span>(labels)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error:
! object 'umap_result' not found</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_umap, <span class="fu">aes</span>(UMAP1, UMAP2, <span class="at">color =</span> digit)) <span class="sc">+</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Spectral"</span>) <span class="sc">+</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"UMAP of MNIST Digits"</span>,</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"UMAP Dimension 1"</span>, <span class="at">y =</span> <span class="st">"UMAP Dimension 2"</span>) <span class="sc">+</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error:
! object 'df_umap' not found</code></pre>
</div>
</div>
</section>
<section id="effect-of-umap-parameters" class="level3">
<h3 class="anchored" data-anchor-id="effect-of-umap-parameters">Effect of UMAP Parameters</h3>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Different parameter combinations</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Vary n_neighbors</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (nn <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">15</span>, <span class="dv">50</span>)) {</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> nn, <span class="at">min_dist =</span> <span class="fl">0.1</span>)</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(result<span class="sc">$</span>layout, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>,</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"n_neighbors ="</span>, nn, <span class="st">", min_dist = 0.1"</span>))</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error in `umap()`:
! could not find function "umap"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Vary min_dist</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (md <span class="cf">in</span> <span class="fu">c</span>(<span class="fl">0.0</span>, <span class="fl">0.1</span>, <span class="fl">0.5</span>)) {</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> <span class="dv">15</span>, <span class="at">min_dist =</span> md)</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(result<span class="sc">$</span>layout, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"n_neighbors = 15, min_dist ="</span>, md))</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error in `umap()`:
! could not find function "umap"</code></pre>
</div>
</div>
</section>
</section>
<section id="comparing-pca-t-sne-and-umap" class="level2" data-number="39.5">
<h2 data-number="39.5" class="anchored" data-anchor-id="comparing-pca-t-sne-and-umap"><span class="header-section-number">39.5</span> Comparing PCA, t-SNE, and UMAP</h2>
<p>Let’s directly compare these three methods on the same dataset:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>pca_result <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x_sample)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_result<span class="sc">$</span>x[, <span class="dv">1</span>], pca_result<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>), <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a><span class="co"># t-SNE</span></span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_result<span class="sc">$</span>Y[, <span class="dv">1</span>], tsne_result<span class="sc">$</span>Y[, <span class="dv">2</span>],</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE"</span>)</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP</span></span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umap_result<span class="sc">$</span>layout[, <span class="dv">1</span>], umap_result<span class="sc">$</span>layout[, <span class="dv">2</span>],</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>, <span class="at">main =</span> <span class="st">"UMAP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error:
! object 'umap_result' not found</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-method-comparison" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-method-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-method-comparison-1.png" class="img-fluid figure-img" width="960">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-method-comparison-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.4: Comparison of PCA, t-SNE, and UMAP on MNIST digits. PCA captures global variance but mixes clusters; t-SNE and UMAP reveal clear cluster structure.
</figcaption>
</figure>
</div>
</div>
</div>
<section id="key-differences" class="level3">
<h3 class="anchored" data-anchor-id="key-differences">Key Differences</h3>
<table class="caption-top table">
<colgroup>
<col style="width: 30%">
<col style="width: 19%">
<col style="width: 26%">
<col style="width: 23%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Aspect</th>
<th style="text-align: left;">PCA</th>
<th style="text-align: left;">t-SNE</th>
<th style="text-align: left;">UMAP</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;"><strong>Type</strong></td>
<td style="text-align: left;">Linear</td>
<td style="text-align: left;">Non-linear</td>
<td style="text-align: left;">Non-linear</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Preserves</strong></td>
<td style="text-align: left;">Global variance</td>
<td style="text-align: left;">Local structure</td>
<td style="text-align: left;">Local + some global</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Speed</strong></td>
<td style="text-align: left;">Very fast</td>
<td style="text-align: left;">Slow (O(n²))</td>
<td style="text-align: left;">Fast</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Reproducibility</strong></td>
<td style="text-align: left;">Deterministic</td>
<td style="text-align: left;">Stochastic</td>
<td style="text-align: left;">Stochastic</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>New data</strong></td>
<td style="text-align: left;">Can project</td>
<td style="text-align: left;">Must rerun</td>
<td style="text-align: left;">Can project (with care)</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Interpretable axes</strong></td>
<td style="text-align: left;">Yes (loadings)</td>
<td style="text-align: left;">No</td>
<td style="text-align: left;">No</td>
</tr>
<tr class="odd">
<td style="text-align: left;"><strong>Cluster distances</strong></td>
<td style="text-align: left;">Meaningful</td>
<td style="text-align: left;">Not meaningful</td>
<td style="text-align: left;">Somewhat meaningful</td>
</tr>
<tr class="even">
<td style="text-align: left;"><strong>Best for</strong></td>
<td style="text-align: left;">Preprocessing, linear patterns</td>
<td style="text-align: left;">Visualization</td>
<td style="text-align: left;">Visualization, larger data</td>
</tr>
</tbody>
</table>
</section>
<section id="when-to-use-each-method" class="level3">
<h3 class="anchored" data-anchor-id="when-to-use-each-method">When to Use Each Method</h3>
<p><strong>Use PCA when:</strong></p>
<ul>
<li>You need interpretable dimensions (loadings)</li>
<li>You want to preprocess data before other analyses</li>
<li>The data structure is approximately linear</li>
<li>You need to project new data points</li>
<li>Speed is critical</li>
</ul>
<p><strong>Use t-SNE when:</strong></p>
<ul>
<li>Visualization is the primary goal</li>
<li>You want to explore local cluster structure</li>
<li>Dataset size is moderate (&lt; 50,000 points)</li>
<li>You’ll examine multiple perplexity values</li>
</ul>
<p><strong>Use UMAP when:</strong></p>
<ul>
<li>You have large datasets</li>
<li>You want faster computation than t-SNE</li>
<li>You care about preserving some global structure</li>
<li>You need to embed new points (though with caveats)</li>
</ul>
</section>
</section>
<section id="biological-applications" class="level2" data-number="39.6">
<h2 data-number="39.6" class="anchored" data-anchor-id="biological-applications"><span class="header-section-number">39.6</span> Biological Applications</h2>
<section id="single-cell-rna-seq" class="level3">
<h3 class="anchored" data-anchor-id="single-cell-rna-seq">Single-Cell RNA-seq</h3>
<p>One of the most common applications of t-SNE and UMAP is visualizing single-cell transcriptomic data:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulated single-cell-like data with distinct populations</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>n_genes <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 4 cell types</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>cell_types <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, n_cells, <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="at">prob =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>))</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate expression data with cell-type-specific patterns</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>expression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n_cells <span class="sc">*</span> n_genes), <span class="at">nrow =</span> n_cells)</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> cell_types <span class="sc">==</span> i</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add cell-type specific signal to first few genes</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>  expression[idx, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="ot">&lt;-</span> expression[idx, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="sc">+</span> i <span class="sc">*</span> <span class="dv">2</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>  expression[idx, (i<span class="sc">*</span><span class="dv">10</span>)<span class="sc">:</span>(i<span class="sc">*</span><span class="dv">10</span><span class="sc">+</span><span class="dv">5</span>)] <span class="ot">&lt;-</span> expression[idx, (i<span class="sc">*</span><span class="dv">10</span>)<span class="sc">:</span>(i<span class="sc">*</span><span class="dv">10</span><span class="sc">+</span><span class="dv">5</span>)] <span class="sc">+</span> <span class="dv">3</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply methods</span></span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>pca_cells <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(expression, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>tsne_cells <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(expression, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>umap_cells <span class="ot">&lt;-</span> <span class="fu">umap</span>(expression, <span class="at">n_neighbors =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error in `umap()`:
! could not find function "umap"</code></pre>
</div>
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_cells<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">paste</span>(<span class="st">"Type"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.7</span>)</span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_cells<span class="sc">$</span>Y, <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE"</span>)</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umap_cells<span class="sc">$</span>layout, <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>, <span class="at">main =</span> <span class="st">"UMAP"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-error">
<pre><code>Error:
! object 'umap_cells' not found</code></pre>
</div>
<div class="cell-output-display">
<div id="fig-biology-example" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-biology-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-biology-example-1.png" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-biology-example-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.5: t-SNE and UMAP are commonly used to visualize cell populations in single-cell genomics data
</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="practical-workflow" class="level2" data-number="39.7">
<h2 data-number="39.7" class="anchored" data-anchor-id="practical-workflow"><span class="header-section-number">39.7</span> Practical Workflow</h2>
<p>A typical workflow for non-linear dimensionality reduction:</p>
<ol type="1">
<li><strong>Preprocess data</strong>: Normalize, scale, and filter features</li>
<li><strong>Apply PCA first</strong>: Reduce to 50-100 dimensions to speed up t-SNE/UMAP</li>
<li><strong>Run t-SNE or UMAP</strong>: With appropriate parameters</li>
<li><strong>Explore parameters</strong>: Try multiple perplexity/n_neighbors values</li>
<li><strong>Validate</strong>: Compare with known labels or biological markers</li>
<li><strong>Iterate</strong>: Adjust preprocessing and parameters as needed</li>
</ol>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA preprocessing example</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="co"># First reduce to 50 PCs, then run t-SNE</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>pca_50 <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x_sample)<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>tsne_after_pca <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(pca_50, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>,</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">pca =</span> <span class="cn">FALSE</span>)  <span class="co"># Already did PCA</span></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_result<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE (direct)"</span>)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_after_pca<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE (after PCA to 50D)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-pca-preprocessing" class="quarto-float quarto-figure quarto-figure-center anchored">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-pca-preprocessing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="38-tsne-umap_files/figure-html/fig-pca-preprocessing-1.png" class="img-fluid figure-img" width="768">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig" id="fig-pca-preprocessing-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;39.6: Using PCA as a preprocessing step before t-SNE speeds computation and can improve results by removing noise
</figcaption>
</figure>
</div>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Best Practices
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li><strong>Always set a random seed</strong> for reproducibility</li>
<li><strong>Try multiple parameter values</strong> and compare results</li>
<li><strong>Use PCA preprocessing</strong> for large, high-dimensional datasets</li>
<li><strong>Don’t over-interpret</strong> cluster sizes or inter-cluster distances</li>
<li><strong>Run multiple times</strong> to assess stability of results</li>
<li><strong>Validate findings</strong> with independent methods or known biology</li>
</ol>
</div>
</div>
</section>
<section id="exercises" class="level2" data-number="39.8">
<h2 data-number="39.8" class="anchored" data-anchor-id="exercises"><span class="header-section-number">39.8</span> Exercises</h2>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.1: Comparing Methods
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Load the <code>tissue_gene_expression</code> dataset and apply PCA, t-SNE, and UMAP. Compare how well each method separates the different tissue types.</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.2: t-SNE Perplexity
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="2" type="1">
<li>For the t-SNE analysis, try perplexity values of 5, 15, 30, and 50. Which value produces the clearest separation of tissue types?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.3: UMAP Parameters
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="3" type="1">
<li>For UMAP, experiment with <code>n_neighbors</code> values of 5, 15, and 50, and <code>min_dist</code> values of 0.0, 0.1, and 0.5. How do these parameters affect the visualization?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.4: Stability Analysis
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="4" type="1">
<li>Run t-SNE on the MNIST digit dataset multiple times with different random seeds. How stable are the cluster positions? Do the same digits always cluster together?</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.5: Computation Time
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="5" type="1">
<li>Time how long PCA, t-SNE, and UMAP take to run on increasing subsets of data (500, 1000, 2000, 5000 points). Plot computation time versus sample size.</li>
</ol>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Exercise TU.6: PCA Preprocessing
</div>
</div>
<div class="callout-body-container callout-body">
<ol start="6" type="1">
<li>Apply PCA preprocessing (reducing to 50 dimensions) before running t-SNE. Compare the results and computation time to running t-SNE directly on the full data.</li>
</ol>
</div>
</div>
</section>
<section id="summary" class="level2" data-number="39.9">
<h2 data-number="39.9" class="anchored" data-anchor-id="summary"><span class="header-section-number">39.9</span> Summary</h2>
<ul>
<li><strong>Non-linear dimensionality reduction</strong> methods can reveal structure that linear methods like PCA miss</li>
<li><strong>t-SNE</strong> is designed for visualization and excels at preserving local neighborhood structure
<ul>
<li>The <strong>perplexity</strong> parameter controls the balance between local and global structure</li>
<li>Cluster sizes and inter-cluster distances are not meaningful</li>
<li>Results depend on random initialization—set a seed for reproducibility</li>
</ul></li>
<li><strong>UMAP</strong> is a newer alternative that is often faster and may preserve more global structure
<ul>
<li>Key parameters are <strong>n_neighbors</strong> and <strong>min_dist</strong></li>
<li>Generally faster than t-SNE, especially on large datasets</li>
</ul></li>
<li><strong>PCA vs t-SNE vs UMAP</strong>:
<ul>
<li>PCA: Fast, interpretable, preserves global variance, works well for linear structure</li>
<li>t-SNE: Best for visualizing local cluster structure, computationally expensive</li>
<li>UMAP: Good balance of speed and quality, preserves some global structure</li>
</ul></li>
<li><strong>Best practices</strong>: Use PCA preprocessing, try multiple parameter values, validate results, don’t over-interpret</li>
</ul>
</section>
<section id="additional-resources" class="level2" data-number="39.10">
<h2 data-number="39.10" class="anchored" data-anchor-id="additional-resources"><span class="header-section-number">39.10</span> Additional Resources</h2>
<ul>
<li>van der Maaten &amp; Hinton (2008). Visualizing Data using t-SNE. <em>JMLR</em></li>
<li>McInnes et al.&nbsp;(2018). UMAP: Uniform Manifold Approximation and Projection. <em>arXiv</em></li>
<li>Kobak &amp; Berens (2019). The art of using t-SNE for single-cell transcriptomics. <em>Nature Communications</em></li>
<li><span class="citation" data-cites="james2023islr">James et al. (<a href="../references.html#ref-james2023islr" role="doc-biblioref">2023</a>)</span> - Additional perspectives on dimensionality reduction</li>
</ul>


<!-- -->

<div id="refs" class="references csl-bib-body hanging-indent" data-entry-spacing="0" role="list" style="display: none">
<div id="ref-james2023islr" class="csl-entry" role="listitem">
James, Gareth, Daniela Witten, Trevor Hastie, and Robert Tibshirani. 2023. <em>An Introduction to Statistical Learning with Applications in r</em>. 2nd ed. Springer. <a href="https://www.statlearning.com">https://www.statlearning.com</a>.
</div>
<div id="ref-vandermaaten2008visualizing" class="csl-entry" role="listitem">
Maaten, Laurens van der, and Geoffrey Hinton. 2008. <span>“Visualizing Data Using t-SNE.”</span> <em>Journal of Machine Learning Research</em> 9: 2579–2605.
</div>
<div id="ref-mcinnes2018umap" class="csl-entry" role="listitem">
McInnes, Leland, John Healy, and James Melville. 2018. <span>“UMAP: Uniform Manifold Approximation and Projection for Dimension Reduction.”</span> <em>arXiv Preprint arXiv:1802.03426</em>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
    const viewSource = window.document.getElementById('quarto-view-source') ||
                       window.document.getElementById('quarto-code-tools-source');
    if (viewSource) {
      const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
      viewSource.addEventListener("click", function(e) {
        if (sourceUrl) {
          // rstudio viewer pane
          if (/\bcapabilities=\b/.test(window.location)) {
            window.open(sourceUrl);
          } else {
            window.location.href = sourceUrl;
          }
        } else {
          const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
          modal.show();
        }
        return false;
      });
    }
    function toggleCodeHandler(show) {
      return function(e) {
        const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
        for (let i=0; i<detailsSrc.length; i++) {
          const details = detailsSrc[i].parentElement;
          if (show) {
            details.open = true;
          } else {
            details.removeAttribute("open");
          }
        }
        const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
        const fromCls = show ? "hidden" : "unhidden";
        const toCls = show ? "unhidden" : "hidden";
        for (let i=0; i<cellCodeDivs.length; i++) {
          const codeDiv = cellCodeDivs[i];
          if (codeDiv.classList.contains(fromCls)) {
            codeDiv.classList.remove(fromCls);
            codeDiv.classList.add(toCls);
          } 
        }
        return false;
      }
    }
    const hideAllCode = window.document.getElementById("quarto-hide-all-code");
    if (hideAllCode) {
      hideAllCode.addEventListener("click", toggleCodeHandler(false));
    }
    const showAllCode = window.document.getElementById("quarto-show-all-code");
    if (showAllCode) {
      showAllCode.addEventListener("click", toggleCodeHandler(true));
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="../chapters/37-dimensionality-reduction.html" class="pagination-link" aria-label="Dimensionality Reduction and Multivariate Methods">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">38</span>&nbsp; <span class="chapter-title">Dimensionality Reduction and Multivariate Methods</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="../chapters/39-bayesian-statistics.html" class="pagination-link" aria-label="Bayesian Statistics">
        <span class="nav-page-text"><span class="chapter-number">40</span>&nbsp; <span class="chapter-title">Bayesian Statistics</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb22" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu"># Non-Linear Dimensionality Reduction: t-SNE and UMAP {#sec-tsne-umap}</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a><span class="co">#| message: false</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Rtsne)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(umap)</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dslabs)</span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a><span class="fu">theme_set</span>(<span class="fu">theme_minimal</span>())</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a><span class="fu">## Beyond Linear Methods</span></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>In @sec-dimensionality-reduction, we introduced Principal Component Analysis (PCA) as a powerful tool for dimensionality reduction. PCA works by finding linear combinations of variables that maximize variance. However, PCA has a fundamental limitation: it can only capture **linear** relationships.</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>Real biological data often contains complex, non-linear structure:</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Cell populations in single-cell RNA-seq form distinct clusters with non-linear boundaries</span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Developmental trajectories follow curved paths through gene expression space</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Protein structures exhibit complex geometric relationships</span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a>When data lies on a curved **manifold** in high-dimensional space, linear methods like PCA may fail to reveal the true structure. This chapter introduces two powerful non-linear methods: **t-SNE** and **UMAP**.</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a><span class="fu">## Limitations of PCA</span></span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a>To see where PCA falls short, consider a simple example: data arranged in a curved pattern.</span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-34"><a href="#cb22-34" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-pca-limitation</span></span>
<span id="cb22-35"><a href="#cb22-35" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "PCA captures the direction of maximum variance but fails to 'unfold' the curved structure. The first PC spreads points along the curve but mixes points from different parts."</span></span>
<span id="cb22-36"><a href="#cb22-36" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb22-37"><a href="#cb22-37" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb22-38"><a href="#cb22-38" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate data on a curved manifold (Swiss roll-like)</span></span>
<span id="cb22-39"><a href="#cb22-39" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-40"><a href="#cb22-40" aria-hidden="true" tabindex="-1"></a>n <span class="ot">&lt;-</span> <span class="dv">500</span></span>
<span id="cb22-41"><a href="#cb22-41" aria-hidden="true" tabindex="-1"></a>t <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">0</span>, <span class="dv">4</span><span class="sc">*</span>pi)</span>
<span id="cb22-42"><a href="#cb22-42" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> t <span class="sc">*</span> <span class="fu">cos</span>(t)</span>
<span id="cb22-43"><a href="#cb22-43" aria-hidden="true" tabindex="-1"></a>y <span class="ot">&lt;-</span> t <span class="sc">*</span> <span class="fu">sin</span>(t)</span>
<span id="cb22-44"><a href="#cb22-44" aria-hidden="true" tabindex="-1"></a>z <span class="ot">&lt;-</span> <span class="fu">runif</span>(n, <span class="dv">0</span>, <span class="dv">10</span>)</span>
<span id="cb22-45"><a href="#cb22-45" aria-hidden="true" tabindex="-1"></a>color <span class="ot">&lt;-</span> t  <span class="co"># Color by position along curve</span></span>
<span id="cb22-46"><a href="#cb22-46" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-47"><a href="#cb22-47" aria-hidden="true" tabindex="-1"></a>curved_data <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">x =</span> x, <span class="at">y =</span> y, <span class="at">z =</span> z, <span class="at">t =</span> t)</span>
<span id="cb22-48"><a href="#cb22-48" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-49"><a href="#cb22-49" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb22-50"><a href="#cb22-50" aria-hidden="true" tabindex="-1"></a>pca_result <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(curved_data[, <span class="dv">1</span><span class="sc">:</span><span class="dv">3</span>], <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-51"><a href="#cb22-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-52"><a href="#cb22-52" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb22-53"><a href="#cb22-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-54"><a href="#cb22-54" aria-hidden="true" tabindex="-1"></a><span class="co"># Original 3D projection</span></span>
<span id="cb22-55"><a href="#cb22-55" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(curved_data<span class="sc">$</span>x, curved_data<span class="sc">$</span>y, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb22-56"><a href="#cb22-56" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"X"</span>, <span class="at">ylab =</span> <span class="st">"Y"</span>,</span>
<span id="cb22-57"><a href="#cb22-57" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Original Data (X-Y projection)"</span>)</span>
<span id="cb22-58"><a href="#cb22-58" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-59"><a href="#cb22-59" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA result</span></span>
<span id="cb22-60"><a href="#cb22-60" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_result<span class="sc">$</span>x[, <span class="dv">1</span>], pca_result<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb22-61"><a href="#cb22-61" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb22-62"><a href="#cb22-62" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>,</span>
<span id="cb22-63"><a href="#cb22-63" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb22-64"><a href="#cb22-64" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-65"><a href="#cb22-65" aria-hidden="true" tabindex="-1"></a><span class="co"># What we want: unfolded</span></span>
<span id="cb22-66"><a href="#cb22-66" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(t, z, <span class="at">col =</span> <span class="fu">heat.colors</span>(<span class="dv">100</span>)[<span class="fu">cut</span>(color, <span class="dv">100</span>)],</span>
<span id="cb22-67"><a href="#cb22-67" aria-hidden="true" tabindex="-1"></a>     <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>, <span class="at">xlab =</span> <span class="st">"Position along curve"</span>, <span class="at">ylab =</span> <span class="st">"Z"</span>,</span>
<span id="cb22-68"><a href="#cb22-68" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Ideal: Unfolded Manifold"</span>)</span>
<span id="cb22-69"><a href="#cb22-69" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-70"><a href="#cb22-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-71"><a href="#cb22-71" aria-hidden="true" tabindex="-1"></a>PCA projects the data onto the directions of maximum variance, but it cannot "unfold" the curved structure. Points that are far apart along the manifold (different colors) may end up close together in the PCA projection.</span>
<span id="cb22-72"><a href="#cb22-72" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-73"><a href="#cb22-73" aria-hidden="true" tabindex="-1"></a><span class="fu">## t-SNE: t-distributed Stochastic Neighbor Embedding</span></span>
<span id="cb22-74"><a href="#cb22-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-75"><a href="#cb22-75" aria-hidden="true" tabindex="-1"></a>**t-SNE** (t-distributed Stochastic Neighbor Embedding) <span class="co">[</span><span class="ot">@vandermaaten2008visualizing</span><span class="co">]</span> is a non-linear dimensionality reduction technique designed specifically for visualization. It was introduced by van der Maaten and Hinton in 2008 and has become one of the most popular methods for visualizing high-dimensional data.</span>
<span id="cb22-76"><a href="#cb22-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-77"><a href="#cb22-77" aria-hidden="true" tabindex="-1"></a><span class="fu">### How t-SNE Works</span></span>
<span id="cb22-78"><a href="#cb22-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-79"><a href="#cb22-79" aria-hidden="true" tabindex="-1"></a>The core idea of t-SNE is to preserve **local neighborhood structure**. Points that are close together in high-dimensional space should remain close together in the low-dimensional embedding.</span>
<span id="cb22-80"><a href="#cb22-80" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-81"><a href="#cb22-81" aria-hidden="true" tabindex="-1"></a>The algorithm works in two steps:</span>
<span id="cb22-82"><a href="#cb22-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-83"><a href="#cb22-83" aria-hidden="true" tabindex="-1"></a>**Step 1: Define similarities in high-dimensional space**</span>
<span id="cb22-84"><a href="#cb22-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-85"><a href="#cb22-85" aria-hidden="true" tabindex="-1"></a>For each pair of points $i$ and $j$, t-SNE computes a probability $p_{ij}$ that represents how likely point $j$ would be picked as a neighbor of point $i$ if neighbors were chosen in proportion to their probability density under a Gaussian centered at $i$:</span>
<span id="cb22-86"><a href="#cb22-86" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-87"><a href="#cb22-87" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-88"><a href="#cb22-88" aria-hidden="true" tabindex="-1"></a>p_{j|i} = \frac{\exp(-<span class="sc">\|</span>x_i - x_j<span class="sc">\|</span>^2 / 2\sigma_i^2)}{\sum_{k \neq i} \exp(-<span class="sc">\|</span>x_i - x_k<span class="sc">\|</span>^2 / 2\sigma_i^2)}</span>
<span id="cb22-89"><a href="#cb22-89" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-90"><a href="#cb22-90" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-91"><a href="#cb22-91" aria-hidden="true" tabindex="-1"></a>The variance $\sigma_i$ is chosen so that each point has a fixed number of effective neighbors (controlled by the **perplexity** parameter).</span>
<span id="cb22-92"><a href="#cb22-92" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-93"><a href="#cb22-93" aria-hidden="true" tabindex="-1"></a>**Step 2: Define similarities in low-dimensional space**</span>
<span id="cb22-94"><a href="#cb22-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-95"><a href="#cb22-95" aria-hidden="true" tabindex="-1"></a>In the low-dimensional embedding, t-SNE uses a **t-distribution** (with one degree of freedom, i.e., a Cauchy distribution) to compute similarities:</span>
<span id="cb22-96"><a href="#cb22-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-97"><a href="#cb22-97" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-98"><a href="#cb22-98" aria-hidden="true" tabindex="-1"></a>q_{ij} = \frac{(1 + <span class="sc">\|</span>y_i - y_j<span class="sc">\|</span>^2)^{-1}}{\sum_{k \neq l} (1 + <span class="sc">\|</span>y_k - y_l<span class="sc">\|</span>^2)^{-1}}</span>
<span id="cb22-99"><a href="#cb22-99" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-100"><a href="#cb22-100" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-101"><a href="#cb22-101" aria-hidden="true" tabindex="-1"></a>The t-distribution has heavier tails than a Gaussian, which allows dissimilar points to be placed far apart without incurring a large penalty.</span>
<span id="cb22-102"><a href="#cb22-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-103"><a href="#cb22-103" aria-hidden="true" tabindex="-1"></a>**Step 3: Minimize the divergence**</span>
<span id="cb22-104"><a href="#cb22-104" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-105"><a href="#cb22-105" aria-hidden="true" tabindex="-1"></a>t-SNE then minimizes the Kullback-Leibler divergence between the two distributions using gradient descent:</span>
<span id="cb22-106"><a href="#cb22-106" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-107"><a href="#cb22-107" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-108"><a href="#cb22-108" aria-hidden="true" tabindex="-1"></a>KL(P <span class="sc">\|</span> Q) = \sum_{i \neq j} p_{ij} \log \frac{p_{ij}}{q_{ij}}</span>
<span id="cb22-109"><a href="#cb22-109" aria-hidden="true" tabindex="-1"></a>$$</span>
<span id="cb22-110"><a href="#cb22-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-111"><a href="#cb22-111" aria-hidden="true" tabindex="-1"></a>This pushes the low-dimensional embedding to match the neighborhood structure of the high-dimensional data.</span>
<span id="cb22-112"><a href="#cb22-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-113"><a href="#cb22-113" aria-hidden="true" tabindex="-1"></a><span class="fu">### The Perplexity Parameter</span></span>
<span id="cb22-114"><a href="#cb22-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-115"><a href="#cb22-115" aria-hidden="true" tabindex="-1"></a>**Perplexity** is the most important parameter in t-SNE. It loosely corresponds to the number of effective nearest neighbors considered for each point. Typical values range from 5 to 50.</span>
<span id="cb22-116"><a href="#cb22-116" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-117"><a href="#cb22-117" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Low perplexity**: Focuses on very local structure; may create many small, disconnected clusters</span>
<span id="cb22-118"><a href="#cb22-118" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**High perplexity**: Considers more global structure; clusters may merge together</span>
<span id="cb22-119"><a href="#cb22-119" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-122"><a href="#cb22-122" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-123"><a href="#cb22-123" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-tsne-perplexity</span></span>
<span id="cb22-124"><a href="#cb22-124" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Effect of perplexity on t-SNE embeddings. Low perplexity emphasizes local structure (may fragment clusters), high perplexity emphasizes global structure (may merge clusters)."</span></span>
<span id="cb22-125"><a href="#cb22-125" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb22-126"><a href="#cb22-126" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb22-127"><a href="#cb22-127" aria-hidden="true" tabindex="-1"></a><span class="co"># Use MNIST digits for demonstration</span></span>
<span id="cb22-128"><a href="#cb22-128" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">exists</span>(<span class="st">"mnist"</span>)) mnist <span class="ot">&lt;-</span> <span class="fu">read_mnist</span>()</span>
<span id="cb22-129"><a href="#cb22-129" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-130"><a href="#cb22-130" aria-hidden="true" tabindex="-1"></a><span class="co"># Sample for speed</span></span>
<span id="cb22-131"><a href="#cb22-131" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-132"><a href="#cb22-132" aria-hidden="true" tabindex="-1"></a>idx <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(mnist<span class="sc">$</span>train<span class="sc">$</span>images), <span class="dv">2000</span>)</span>
<span id="cb22-133"><a href="#cb22-133" aria-hidden="true" tabindex="-1"></a>x_sample <span class="ot">&lt;-</span> mnist<span class="sc">$</span>train<span class="sc">$</span>images[idx, ]</span>
<span id="cb22-134"><a href="#cb22-134" aria-hidden="true" tabindex="-1"></a>labels <span class="ot">&lt;-</span> mnist<span class="sc">$</span>train<span class="sc">$</span>labels[idx]</span>
<span id="cb22-135"><a href="#cb22-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-136"><a href="#cb22-136" aria-hidden="true" tabindex="-1"></a><span class="co"># t-SNE with different perplexities</span></span>
<span id="cb22-137"><a href="#cb22-137" aria-hidden="true" tabindex="-1"></a>tsne_5 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">5</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb22-138"><a href="#cb22-138" aria-hidden="true" tabindex="-1"></a>tsne_30 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb22-139"><a href="#cb22-139" aria-hidden="true" tabindex="-1"></a>tsne_100 <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">100</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">500</span>)</span>
<span id="cb22-140"><a href="#cb22-140" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-141"><a href="#cb22-141" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb22-142"><a href="#cb22-142" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_5<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-143"><a href="#cb22-143" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 5"</span>)</span>
<span id="cb22-144"><a href="#cb22-144" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_30<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-145"><a href="#cb22-145" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 30"</span>)</span>
<span id="cb22-146"><a href="#cb22-146" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_100<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-147"><a href="#cb22-147" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"Perplexity = 100"</span>)</span>
<span id="cb22-148"><a href="#cb22-148" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-149"><a href="#cb22-149" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-150"><a href="#cb22-150" aria-hidden="true" tabindex="-1"></a><span class="fu">### t-SNE in R</span></span>
<span id="cb22-151"><a href="#cb22-151" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-152"><a href="#cb22-152" aria-hidden="true" tabindex="-1"></a>The <span class="in">`Rtsne`</span> package provides an efficient implementation:</span>
<span id="cb22-153"><a href="#cb22-153" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-156"><a href="#cb22-156" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-157"><a href="#cb22-157" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-tsne-mnist</span></span>
<span id="cb22-158"><a href="#cb22-158" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "t-SNE visualization of MNIST digits reveals clear cluster structure corresponding to digit classes"</span></span>
<span id="cb22-159"><a href="#cb22-159" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb22-160"><a href="#cb22-160" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb22-161"><a href="#cb22-161" aria-hidden="true" tabindex="-1"></a><span class="co"># Full t-SNE with good parameters</span></span>
<span id="cb22-162"><a href="#cb22-162" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-163"><a href="#cb22-163" aria-hidden="true" tabindex="-1"></a>tsne_result <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(x_sample, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>, <span class="at">max_iter =</span> <span class="dv">1000</span>)</span>
<span id="cb22-164"><a href="#cb22-164" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-165"><a href="#cb22-165" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot with digit labels</span></span>
<span id="cb22-166"><a href="#cb22-166" aria-hidden="true" tabindex="-1"></a>df_tsne <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb22-167"><a href="#cb22-167" aria-hidden="true" tabindex="-1"></a>  <span class="at">tSNE1 =</span> tsne_result<span class="sc">$</span>Y[, <span class="dv">1</span>],</span>
<span id="cb22-168"><a href="#cb22-168" aria-hidden="true" tabindex="-1"></a>  <span class="at">tSNE2 =</span> tsne_result<span class="sc">$</span>Y[, <span class="dv">2</span>],</span>
<span id="cb22-169"><a href="#cb22-169" aria-hidden="true" tabindex="-1"></a>  <span class="at">digit =</span> <span class="fu">factor</span>(labels)</span>
<span id="cb22-170"><a href="#cb22-170" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-171"><a href="#cb22-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-172"><a href="#cb22-172" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_tsne, <span class="fu">aes</span>(tSNE1, tSNE2, <span class="at">color =</span> digit)) <span class="sc">+</span></span>
<span id="cb22-173"><a href="#cb22-173" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb22-174"><a href="#cb22-174" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Spectral"</span>) <span class="sc">+</span></span>
<span id="cb22-175"><a href="#cb22-175" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"t-SNE of MNIST Digits"</span>,</span>
<span id="cb22-176"><a href="#cb22-176" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"t-SNE Dimension 1"</span>, <span class="at">y =</span> <span class="st">"t-SNE Dimension 2"</span>) <span class="sc">+</span></span>
<span id="cb22-177"><a href="#cb22-177" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb22-178"><a href="#cb22-178" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-179"><a href="#cb22-179" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-180"><a href="#cb22-180" aria-hidden="true" tabindex="-1"></a><span class="fu">### Important Considerations for t-SNE</span></span>
<span id="cb22-181"><a href="#cb22-181" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-182"><a href="#cb22-182" aria-hidden="true" tabindex="-1"></a>::: {.callout-warning}</span>
<span id="cb22-183"><a href="#cb22-183" aria-hidden="true" tabindex="-1"></a><span class="fu">## t-SNE Caveats</span></span>
<span id="cb22-184"><a href="#cb22-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-185"><a href="#cb22-185" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Cluster sizes are meaningless**: t-SNE does not preserve density. A large cluster in t-SNE does not mean more points.</span>
<span id="cb22-186"><a href="#cb22-186" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-187"><a href="#cb22-187" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Distances between clusters are meaningless**: Only local structure is preserved. The distance between two clusters tells you nothing about their true similarity.</span>
<span id="cb22-188"><a href="#cb22-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-189"><a href="#cb22-189" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Results depend on random initialization**: Always set a seed for reproducibility. Run multiple times to check stability.</span>
<span id="cb22-190"><a href="#cb22-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-191"><a href="#cb22-191" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Perplexity must be tuned**: There's no universally best value. Try multiple perplexities to understand your data.</span>
<span id="cb22-192"><a href="#cb22-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-193"><a href="#cb22-193" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>**Computationally expensive**: O(n²) complexity; can be slow for large datasets. Use PCA preprocessing to speed up.</span>
<span id="cb22-194"><a href="#cb22-194" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-195"><a href="#cb22-195" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-196"><a href="#cb22-196" aria-hidden="true" tabindex="-1"></a><span class="fu">## UMAP: Uniform Manifold Approximation and Projection</span></span>
<span id="cb22-197"><a href="#cb22-197" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-198"><a href="#cb22-198" aria-hidden="true" tabindex="-1"></a>**UMAP** (Uniform Manifold Approximation and Projection) <span class="co">[</span><span class="ot">@mcinnes2018umap</span><span class="co">]</span> is a more recent algorithm that has become increasingly popular, especially in single-cell genomics. UMAP is often faster than t-SNE and may better preserve global structure.</span>
<span id="cb22-199"><a href="#cb22-199" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-200"><a href="#cb22-200" aria-hidden="true" tabindex="-1"></a><span class="fu">### How UMAP Works</span></span>
<span id="cb22-201"><a href="#cb22-201" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-202"><a href="#cb22-202" aria-hidden="true" tabindex="-1"></a>UMAP is grounded in Riemannian geometry and algebraic topology, but the practical intuition is similar to t-SNE: preserve neighborhood relationships.</span>
<span id="cb22-203"><a href="#cb22-203" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-204"><a href="#cb22-204" aria-hidden="true" tabindex="-1"></a>Key differences from t-SNE:</span>
<span id="cb22-205"><a href="#cb22-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-206"><a href="#cb22-206" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Mathematical foundation**: UMAP assumes the data lies on a uniformly distributed manifold and uses concepts from topological data analysis.</span>
<span id="cb22-207"><a href="#cb22-207" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-208"><a href="#cb22-208" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Graph-based approach**: UMAP constructs a weighted graph representing the high-dimensional data, then optimizes a low-dimensional graph to match it.</span>
<span id="cb22-209"><a href="#cb22-209" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-210"><a href="#cb22-210" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Different cost function**: UMAP uses cross-entropy rather than KL divergence, which may help preserve more global structure.</span>
<span id="cb22-211"><a href="#cb22-211" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-212"><a href="#cb22-212" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Faster**: UMAP typically runs faster than t-SNE, especially on large datasets.</span>
<span id="cb22-213"><a href="#cb22-213" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-214"><a href="#cb22-214" aria-hidden="true" tabindex="-1"></a><span class="fu">### Key UMAP Parameters</span></span>
<span id="cb22-215"><a href="#cb22-215" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-216"><a href="#cb22-216" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**n_neighbors**: Similar to perplexity in t-SNE. Controls the balance between local and global structure. Typical values: 5-50.</span>
<span id="cb22-217"><a href="#cb22-217" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**min_dist**: Controls how tightly UMAP packs points together. Small values create tighter clusters. Typical values: 0.0-0.5.</span>
<span id="cb22-218"><a href="#cb22-218" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**metric**: Distance metric used. Default is Euclidean, but many options available.</span>
<span id="cb22-219"><a href="#cb22-219" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-220"><a href="#cb22-220" aria-hidden="true" tabindex="-1"></a><span class="fu">### UMAP in R</span></span>
<span id="cb22-221"><a href="#cb22-221" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-222"><a href="#cb22-222" aria-hidden="true" tabindex="-1"></a>The <span class="in">`umap`</span> package provides a convenient interface:</span>
<span id="cb22-223"><a href="#cb22-223" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-226"><a href="#cb22-226" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-227"><a href="#cb22-227" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-umap-mnist</span></span>
<span id="cb22-228"><a href="#cb22-228" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "UMAP visualization of MNIST digits. Note the clear separation and potentially better preservation of global structure compared to t-SNE."</span></span>
<span id="cb22-229"><a href="#cb22-229" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 7</span></span>
<span id="cb22-230"><a href="#cb22-230" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb22-231"><a href="#cb22-231" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP</span></span>
<span id="cb22-232"><a href="#cb22-232" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-233"><a href="#cb22-233" aria-hidden="true" tabindex="-1"></a>umap_result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> <span class="dv">15</span>, <span class="at">min_dist =</span> <span class="fl">0.1</span>)</span>
<span id="cb22-234"><a href="#cb22-234" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-235"><a href="#cb22-235" aria-hidden="true" tabindex="-1"></a><span class="co"># Plot</span></span>
<span id="cb22-236"><a href="#cb22-236" aria-hidden="true" tabindex="-1"></a>df_umap <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb22-237"><a href="#cb22-237" aria-hidden="true" tabindex="-1"></a>  <span class="at">UMAP1 =</span> umap_result<span class="sc">$</span>layout[, <span class="dv">1</span>],</span>
<span id="cb22-238"><a href="#cb22-238" aria-hidden="true" tabindex="-1"></a>  <span class="at">UMAP2 =</span> umap_result<span class="sc">$</span>layout[, <span class="dv">2</span>],</span>
<span id="cb22-239"><a href="#cb22-239" aria-hidden="true" tabindex="-1"></a>  <span class="at">digit =</span> <span class="fu">factor</span>(labels)</span>
<span id="cb22-240"><a href="#cb22-240" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb22-241"><a href="#cb22-241" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-242"><a href="#cb22-242" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(df_umap, <span class="fu">aes</span>(UMAP1, UMAP2, <span class="at">color =</span> digit)) <span class="sc">+</span></span>
<span id="cb22-243"><a href="#cb22-243" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">alpha =</span> <span class="fl">0.6</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb22-244"><a href="#cb22-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_brewer</span>(<span class="at">palette =</span> <span class="st">"Spectral"</span>) <span class="sc">+</span></span>
<span id="cb22-245"><a href="#cb22-245" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"UMAP of MNIST Digits"</span>,</span>
<span id="cb22-246"><a href="#cb22-246" aria-hidden="true" tabindex="-1"></a>       <span class="at">x =</span> <span class="st">"UMAP Dimension 1"</span>, <span class="at">y =</span> <span class="st">"UMAP Dimension 2"</span>) <span class="sc">+</span></span>
<span id="cb22-247"><a href="#cb22-247" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">legend.position =</span> <span class="st">"right"</span>)</span>
<span id="cb22-248"><a href="#cb22-248" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-249"><a href="#cb22-249" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-250"><a href="#cb22-250" aria-hidden="true" tabindex="-1"></a><span class="fu">### Effect of UMAP Parameters</span></span>
<span id="cb22-251"><a href="#cb22-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-254"><a href="#cb22-254" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-255"><a href="#cb22-255" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-umap-params</span></span>
<span id="cb22-256"><a href="#cb22-256" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Effect of n_neighbors and min_dist on UMAP embeddings. These parameters control the balance between local detail and global structure."</span></span>
<span id="cb22-257"><a href="#cb22-257" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 9</span></span>
<span id="cb22-258"><a href="#cb22-258" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 6</span></span>
<span id="cb22-259"><a href="#cb22-259" aria-hidden="true" tabindex="-1"></a><span class="co"># Different parameter combinations</span></span>
<span id="cb22-260"><a href="#cb22-260" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">2</span>, <span class="dv">3</span>))</span>
<span id="cb22-261"><a href="#cb22-261" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-262"><a href="#cb22-262" aria-hidden="true" tabindex="-1"></a><span class="co"># Vary n_neighbors</span></span>
<span id="cb22-263"><a href="#cb22-263" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (nn <span class="cf">in</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">15</span>, <span class="dv">50</span>)) {</span>
<span id="cb22-264"><a href="#cb22-264" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-265"><a href="#cb22-265" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> nn, <span class="at">min_dist =</span> <span class="fl">0.1</span>)</span>
<span id="cb22-266"><a href="#cb22-266" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(result<span class="sc">$</span>layout, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-267"><a href="#cb22-267" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>,</span>
<span id="cb22-268"><a href="#cb22-268" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"n_neighbors ="</span>, nn, <span class="st">", min_dist = 0.1"</span>))</span>
<span id="cb22-269"><a href="#cb22-269" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-270"><a href="#cb22-270" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-271"><a href="#cb22-271" aria-hidden="true" tabindex="-1"></a><span class="co"># Vary min_dist</span></span>
<span id="cb22-272"><a href="#cb22-272" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (md <span class="cf">in</span> <span class="fu">c</span>(<span class="fl">0.0</span>, <span class="fl">0.1</span>, <span class="fl">0.5</span>)) {</span>
<span id="cb22-273"><a href="#cb22-273" aria-hidden="true" tabindex="-1"></a>  <span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-274"><a href="#cb22-274" aria-hidden="true" tabindex="-1"></a>  result <span class="ot">&lt;-</span> <span class="fu">umap</span>(x_sample, <span class="at">n_neighbors =</span> <span class="dv">15</span>, <span class="at">min_dist =</span> md)</span>
<span id="cb22-275"><a href="#cb22-275" aria-hidden="true" tabindex="-1"></a>  <span class="fu">plot</span>(result<span class="sc">$</span>layout, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-276"><a href="#cb22-276" aria-hidden="true" tabindex="-1"></a>       <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>,</span>
<span id="cb22-277"><a href="#cb22-277" aria-hidden="true" tabindex="-1"></a>       <span class="at">main =</span> <span class="fu">paste</span>(<span class="st">"n_neighbors = 15, min_dist ="</span>, md))</span>
<span id="cb22-278"><a href="#cb22-278" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-279"><a href="#cb22-279" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-280"><a href="#cb22-280" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-281"><a href="#cb22-281" aria-hidden="true" tabindex="-1"></a><span class="fu">## Comparing PCA, t-SNE, and UMAP</span></span>
<span id="cb22-282"><a href="#cb22-282" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-283"><a href="#cb22-283" aria-hidden="true" tabindex="-1"></a>Let's directly compare these three methods on the same dataset:</span>
<span id="cb22-284"><a href="#cb22-284" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-287"><a href="#cb22-287" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-288"><a href="#cb22-288" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-method-comparison</span></span>
<span id="cb22-289"><a href="#cb22-289" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Comparison of PCA, t-SNE, and UMAP on MNIST digits. PCA captures global variance but mixes clusters; t-SNE and UMAP reveal clear cluster structure."</span></span>
<span id="cb22-290"><a href="#cb22-290" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 10</span></span>
<span id="cb22-291"><a href="#cb22-291" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb22-292"><a href="#cb22-292" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb22-293"><a href="#cb22-293" aria-hidden="true" tabindex="-1"></a>pca_result <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x_sample)</span>
<span id="cb22-294"><a href="#cb22-294" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-295"><a href="#cb22-295" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb22-296"><a href="#cb22-296" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-297"><a href="#cb22-297" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA</span></span>
<span id="cb22-298"><a href="#cb22-298" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_result<span class="sc">$</span>x[, <span class="dv">1</span>], pca_result<span class="sc">$</span>x[, <span class="dv">2</span>],</span>
<span id="cb22-299"><a href="#cb22-299" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-300"><a href="#cb22-300" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb22-301"><a href="#cb22-301" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="dv">0</span><span class="sc">:</span><span class="dv">9</span>, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>), <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.6</span>, <span class="at">ncol =</span> <span class="dv">2</span>)</span>
<span id="cb22-302"><a href="#cb22-302" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-303"><a href="#cb22-303" aria-hidden="true" tabindex="-1"></a><span class="co"># t-SNE</span></span>
<span id="cb22-304"><a href="#cb22-304" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_result<span class="sc">$</span>Y[, <span class="dv">1</span>], tsne_result<span class="sc">$</span>Y[, <span class="dv">2</span>],</span>
<span id="cb22-305"><a href="#cb22-305" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-306"><a href="#cb22-306" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE"</span>)</span>
<span id="cb22-307"><a href="#cb22-307" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-308"><a href="#cb22-308" aria-hidden="true" tabindex="-1"></a><span class="co"># UMAP</span></span>
<span id="cb22-309"><a href="#cb22-309" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umap_result<span class="sc">$</span>layout[, <span class="dv">1</span>], umap_result<span class="sc">$</span>layout[, <span class="dv">2</span>],</span>
<span id="cb22-310"><a href="#cb22-310" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-311"><a href="#cb22-311" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>, <span class="at">main =</span> <span class="st">"UMAP"</span>)</span>
<span id="cb22-312"><a href="#cb22-312" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-313"><a href="#cb22-313" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-314"><a href="#cb22-314" aria-hidden="true" tabindex="-1"></a><span class="fu">### Key Differences</span></span>
<span id="cb22-315"><a href="#cb22-315" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-316"><a href="#cb22-316" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> Aspect <span class="pp">|</span> PCA <span class="pp">|</span> t-SNE <span class="pp">|</span> UMAP <span class="pp">|</span></span>
<span id="cb22-317"><a href="#cb22-317" aria-hidden="true" tabindex="-1"></a><span class="pp">|:-------|:----|:------|:-----|</span></span>
<span id="cb22-318"><a href="#cb22-318" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Type** <span class="pp">|</span> Linear <span class="pp">|</span> Non-linear <span class="pp">|</span> Non-linear <span class="pp">|</span></span>
<span id="cb22-319"><a href="#cb22-319" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Preserves** <span class="pp">|</span> Global variance <span class="pp">|</span> Local structure <span class="pp">|</span> Local + some global <span class="pp">|</span></span>
<span id="cb22-320"><a href="#cb22-320" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Speed** <span class="pp">|</span> Very fast <span class="pp">|</span> Slow (O(n²)) <span class="pp">|</span> Fast <span class="pp">|</span></span>
<span id="cb22-321"><a href="#cb22-321" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Reproducibility** <span class="pp">|</span> Deterministic <span class="pp">|</span> Stochastic <span class="pp">|</span> Stochastic <span class="pp">|</span></span>
<span id="cb22-322"><a href="#cb22-322" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **New data** <span class="pp">|</span> Can project <span class="pp">|</span> Must rerun <span class="pp">|</span> Can project (with care) <span class="pp">|</span></span>
<span id="cb22-323"><a href="#cb22-323" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Interpretable axes** <span class="pp">|</span> Yes (loadings) <span class="pp">|</span> No <span class="pp">|</span> No <span class="pp">|</span></span>
<span id="cb22-324"><a href="#cb22-324" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Cluster distances** <span class="pp">|</span> Meaningful <span class="pp">|</span> Not meaningful <span class="pp">|</span> Somewhat meaningful <span class="pp">|</span></span>
<span id="cb22-325"><a href="#cb22-325" aria-hidden="true" tabindex="-1"></a><span class="pp">|</span> **Best for** <span class="pp">|</span> Preprocessing, linear patterns <span class="pp">|</span> Visualization <span class="pp">|</span> Visualization, larger data <span class="pp">|</span></span>
<span id="cb22-326"><a href="#cb22-326" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-327"><a href="#cb22-327" aria-hidden="true" tabindex="-1"></a><span class="fu">### When to Use Each Method</span></span>
<span id="cb22-328"><a href="#cb22-328" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-329"><a href="#cb22-329" aria-hidden="true" tabindex="-1"></a>**Use PCA when:**</span>
<span id="cb22-330"><a href="#cb22-330" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-331"><a href="#cb22-331" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You need interpretable dimensions (loadings)</span>
<span id="cb22-332"><a href="#cb22-332" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You want to preprocess data before other analyses</span>
<span id="cb22-333"><a href="#cb22-333" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>The data structure is approximately linear</span>
<span id="cb22-334"><a href="#cb22-334" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You need to project new data points</span>
<span id="cb22-335"><a href="#cb22-335" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Speed is critical</span>
<span id="cb22-336"><a href="#cb22-336" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-337"><a href="#cb22-337" aria-hidden="true" tabindex="-1"></a>**Use t-SNE when:**</span>
<span id="cb22-338"><a href="#cb22-338" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-339"><a href="#cb22-339" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Visualization is the primary goal</span>
<span id="cb22-340"><a href="#cb22-340" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You want to explore local cluster structure</span>
<span id="cb22-341"><a href="#cb22-341" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Dataset size is moderate (&lt; 50,000 points)</span>
<span id="cb22-342"><a href="#cb22-342" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You'll examine multiple perplexity values</span>
<span id="cb22-343"><a href="#cb22-343" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-344"><a href="#cb22-344" aria-hidden="true" tabindex="-1"></a>**Use UMAP when:**</span>
<span id="cb22-345"><a href="#cb22-345" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-346"><a href="#cb22-346" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You have large datasets</span>
<span id="cb22-347"><a href="#cb22-347" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You want faster computation than t-SNE</span>
<span id="cb22-348"><a href="#cb22-348" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You care about preserving some global structure</span>
<span id="cb22-349"><a href="#cb22-349" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>You need to embed new points (though with caveats)</span>
<span id="cb22-350"><a href="#cb22-350" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-351"><a href="#cb22-351" aria-hidden="true" tabindex="-1"></a><span class="fu">## Biological Applications</span></span>
<span id="cb22-352"><a href="#cb22-352" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-353"><a href="#cb22-353" aria-hidden="true" tabindex="-1"></a><span class="fu">### Single-Cell RNA-seq</span></span>
<span id="cb22-354"><a href="#cb22-354" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-355"><a href="#cb22-355" aria-hidden="true" tabindex="-1"></a>One of the most common applications of t-SNE and UMAP is visualizing single-cell transcriptomic data:</span>
<span id="cb22-356"><a href="#cb22-356" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-359"><a href="#cb22-359" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-360"><a href="#cb22-360" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-biology-example</span></span>
<span id="cb22-361"><a href="#cb22-361" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "t-SNE and UMAP are commonly used to visualize cell populations in single-cell genomics data"</span></span>
<span id="cb22-362"><a href="#cb22-362" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb22-363"><a href="#cb22-363" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb22-364"><a href="#cb22-364" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulated single-cell-like data with distinct populations</span></span>
<span id="cb22-365"><a href="#cb22-365" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb22-366"><a href="#cb22-366" aria-hidden="true" tabindex="-1"></a>n_cells <span class="ot">&lt;-</span> <span class="dv">1000</span></span>
<span id="cb22-367"><a href="#cb22-367" aria-hidden="true" tabindex="-1"></a>n_genes <span class="ot">&lt;-</span> <span class="dv">50</span></span>
<span id="cb22-368"><a href="#cb22-368" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-369"><a href="#cb22-369" aria-hidden="true" tabindex="-1"></a><span class="co"># Simulate 4 cell types</span></span>
<span id="cb22-370"><a href="#cb22-370" aria-hidden="true" tabindex="-1"></a>cell_types <span class="ot">&lt;-</span> <span class="fu">sample</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, n_cells, <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="at">prob =</span> <span class="fu">c</span>(<span class="fl">0.4</span>, <span class="fl">0.3</span>, <span class="fl">0.2</span>, <span class="fl">0.1</span>))</span>
<span id="cb22-371"><a href="#cb22-371" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-372"><a href="#cb22-372" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate expression data with cell-type-specific patterns</span></span>
<span id="cb22-373"><a href="#cb22-373" aria-hidden="true" tabindex="-1"></a>expression <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="fu">rnorm</span>(n_cells <span class="sc">*</span> n_genes), <span class="at">nrow =</span> n_cells)</span>
<span id="cb22-374"><a href="#cb22-374" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>) {</span>
<span id="cb22-375"><a href="#cb22-375" aria-hidden="true" tabindex="-1"></a>  idx <span class="ot">&lt;-</span> cell_types <span class="sc">==</span> i</span>
<span id="cb22-376"><a href="#cb22-376" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Add cell-type specific signal to first few genes</span></span>
<span id="cb22-377"><a href="#cb22-377" aria-hidden="true" tabindex="-1"></a>  expression[idx, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="ot">&lt;-</span> expression[idx, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>] <span class="sc">+</span> i <span class="sc">*</span> <span class="dv">2</span></span>
<span id="cb22-378"><a href="#cb22-378" aria-hidden="true" tabindex="-1"></a>  expression[idx, (i<span class="sc">*</span><span class="dv">10</span>)<span class="sc">:</span>(i<span class="sc">*</span><span class="dv">10</span><span class="sc">+</span><span class="dv">5</span>)] <span class="ot">&lt;-</span> expression[idx, (i<span class="sc">*</span><span class="dv">10</span>)<span class="sc">:</span>(i<span class="sc">*</span><span class="dv">10</span><span class="sc">+</span><span class="dv">5</span>)] <span class="sc">+</span> <span class="dv">3</span></span>
<span id="cb22-379"><a href="#cb22-379" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-380"><a href="#cb22-380" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-381"><a href="#cb22-381" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply methods</span></span>
<span id="cb22-382"><a href="#cb22-382" aria-hidden="true" tabindex="-1"></a>pca_cells <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(expression, <span class="at">scale. =</span> <span class="cn">TRUE</span>)</span>
<span id="cb22-383"><a href="#cb22-383" aria-hidden="true" tabindex="-1"></a>tsne_cells <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(expression, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb22-384"><a href="#cb22-384" aria-hidden="true" tabindex="-1"></a>umap_cells <span class="ot">&lt;-</span> <span class="fu">umap</span>(expression, <span class="at">n_neighbors =</span> <span class="dv">15</span>)</span>
<span id="cb22-385"><a href="#cb22-385" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-386"><a href="#cb22-386" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>))</span>
<span id="cb22-387"><a href="#cb22-387" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(pca_cells<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>], <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-388"><a href="#cb22-388" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"PC1"</span>, <span class="at">ylab =</span> <span class="st">"PC2"</span>, <span class="at">main =</span> <span class="st">"PCA"</span>)</span>
<span id="cb22-389"><a href="#cb22-389" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">paste</span>(<span class="st">"Type"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>), <span class="at">col =</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.7</span>)</span>
<span id="cb22-390"><a href="#cb22-390" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-391"><a href="#cb22-391" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_cells<span class="sc">$</span>Y, <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-392"><a href="#cb22-392" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE"</span>)</span>
<span id="cb22-393"><a href="#cb22-393" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-394"><a href="#cb22-394" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(umap_cells<span class="sc">$</span>layout, <span class="at">col =</span> cell_types, <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.5</span>,</span>
<span id="cb22-395"><a href="#cb22-395" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"UMAP 1"</span>, <span class="at">ylab =</span> <span class="st">"UMAP 2"</span>, <span class="at">main =</span> <span class="st">"UMAP"</span>)</span>
<span id="cb22-396"><a href="#cb22-396" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-397"><a href="#cb22-397" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-398"><a href="#cb22-398" aria-hidden="true" tabindex="-1"></a><span class="fu">## Practical Workflow</span></span>
<span id="cb22-399"><a href="#cb22-399" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-400"><a href="#cb22-400" aria-hidden="true" tabindex="-1"></a>A typical workflow for non-linear dimensionality reduction:</span>
<span id="cb22-401"><a href="#cb22-401" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-402"><a href="#cb22-402" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Preprocess data**: Normalize, scale, and filter features</span>
<span id="cb22-403"><a href="#cb22-403" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Apply PCA first**: Reduce to 50-100 dimensions to speed up t-SNE/UMAP</span>
<span id="cb22-404"><a href="#cb22-404" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Run t-SNE or UMAP**: With appropriate parameters</span>
<span id="cb22-405"><a href="#cb22-405" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Explore parameters**: Try multiple perplexity/n_neighbors values</span>
<span id="cb22-406"><a href="#cb22-406" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>**Validate**: Compare with known labels or biological markers</span>
<span id="cb22-407"><a href="#cb22-407" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>**Iterate**: Adjust preprocessing and parameters as needed</span>
<span id="cb22-408"><a href="#cb22-408" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-411"><a href="#cb22-411" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb22-412"><a href="#cb22-412" aria-hidden="true" tabindex="-1"></a><span class="co">#| label: fig-pca-preprocessing</span></span>
<span id="cb22-413"><a href="#cb22-413" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-cap: "Using PCA as a preprocessing step before t-SNE speeds computation and can improve results by removing noise"</span></span>
<span id="cb22-414"><a href="#cb22-414" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-width: 8</span></span>
<span id="cb22-415"><a href="#cb22-415" aria-hidden="true" tabindex="-1"></a><span class="co">#| fig-height: 4</span></span>
<span id="cb22-416"><a href="#cb22-416" aria-hidden="true" tabindex="-1"></a><span class="co"># PCA preprocessing example</span></span>
<span id="cb22-417"><a href="#cb22-417" aria-hidden="true" tabindex="-1"></a><span class="co"># First reduce to 50 PCs, then run t-SNE</span></span>
<span id="cb22-418"><a href="#cb22-418" aria-hidden="true" tabindex="-1"></a>pca_50 <span class="ot">&lt;-</span> <span class="fu">prcomp</span>(x_sample)<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>]</span>
<span id="cb22-419"><a href="#cb22-419" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-420"><a href="#cb22-420" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb22-421"><a href="#cb22-421" aria-hidden="true" tabindex="-1"></a>tsne_after_pca <span class="ot">&lt;-</span> <span class="fu">Rtsne</span>(pca_50, <span class="at">perplexity =</span> <span class="dv">30</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>,</span>
<span id="cb22-422"><a href="#cb22-422" aria-hidden="true" tabindex="-1"></a>                         <span class="at">pca =</span> <span class="cn">FALSE</span>)  <span class="co"># Already did PCA</span></span>
<span id="cb22-423"><a href="#cb22-423" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-424"><a href="#cb22-424" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb22-425"><a href="#cb22-425" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_result<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-426"><a href="#cb22-426" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE (direct)"</span>)</span>
<span id="cb22-427"><a href="#cb22-427" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(tsne_after_pca<span class="sc">$</span>Y, <span class="at">col =</span> <span class="fu">rainbow</span>(<span class="dv">10</span>)[labels <span class="sc">+</span> <span class="dv">1</span>], <span class="at">pch =</span> <span class="dv">19</span>, <span class="at">cex =</span> <span class="fl">0.3</span>,</span>
<span id="cb22-428"><a href="#cb22-428" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"t-SNE 1"</span>, <span class="at">ylab =</span> <span class="st">"t-SNE 2"</span>, <span class="at">main =</span> <span class="st">"t-SNE (after PCA to 50D)"</span>)</span>
<span id="cb22-429"><a href="#cb22-429" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb22-430"><a href="#cb22-430" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-431"><a href="#cb22-431" aria-hidden="true" tabindex="-1"></a>::: {.callout-tip}</span>
<span id="cb22-432"><a href="#cb22-432" aria-hidden="true" tabindex="-1"></a><span class="fu">## Best Practices</span></span>
<span id="cb22-433"><a href="#cb22-433" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-434"><a href="#cb22-434" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>**Always set a random seed** for reproducibility</span>
<span id="cb22-435"><a href="#cb22-435" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>**Try multiple parameter values** and compare results</span>
<span id="cb22-436"><a href="#cb22-436" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>**Use PCA preprocessing** for large, high-dimensional datasets</span>
<span id="cb22-437"><a href="#cb22-437" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>**Don't over-interpret** cluster sizes or inter-cluster distances</span>
<span id="cb22-438"><a href="#cb22-438" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>**Run multiple times** to assess stability of results</span>
<span id="cb22-439"><a href="#cb22-439" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>**Validate findings** with independent methods or known biology</span>
<span id="cb22-440"><a href="#cb22-440" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-441"><a href="#cb22-441" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-442"><a href="#cb22-442" aria-hidden="true" tabindex="-1"></a><span class="fu">## Exercises</span></span>
<span id="cb22-443"><a href="#cb22-443" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-444"><a href="#cb22-444" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-445"><a href="#cb22-445" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.1: Comparing Methods</span></span>
<span id="cb22-446"><a href="#cb22-446" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-447"><a href="#cb22-447" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Load the <span class="in">`tissue_gene_expression`</span> dataset and apply PCA, t-SNE, and UMAP. Compare how well each method separates the different tissue types.</span>
<span id="cb22-448"><a href="#cb22-448" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-449"><a href="#cb22-449" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-450"><a href="#cb22-450" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-451"><a href="#cb22-451" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.2: t-SNE Perplexity</span></span>
<span id="cb22-452"><a href="#cb22-452" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-453"><a href="#cb22-453" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>For the t-SNE analysis, try perplexity values of 5, 15, 30, and 50. Which value produces the clearest separation of tissue types?</span>
<span id="cb22-454"><a href="#cb22-454" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-455"><a href="#cb22-455" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-456"><a href="#cb22-456" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-457"><a href="#cb22-457" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.3: UMAP Parameters</span></span>
<span id="cb22-458"><a href="#cb22-458" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-459"><a href="#cb22-459" aria-hidden="true" tabindex="-1"></a><span class="ss">3. </span>For UMAP, experiment with <span class="in">`n_neighbors`</span> values of 5, 15, and 50, and <span class="in">`min_dist`</span> values of 0.0, 0.1, and 0.5. How do these parameters affect the visualization?</span>
<span id="cb22-460"><a href="#cb22-460" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-461"><a href="#cb22-461" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-462"><a href="#cb22-462" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-463"><a href="#cb22-463" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.4: Stability Analysis</span></span>
<span id="cb22-464"><a href="#cb22-464" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-465"><a href="#cb22-465" aria-hidden="true" tabindex="-1"></a><span class="ss">4. </span>Run t-SNE on the MNIST digit dataset multiple times with different random seeds. How stable are the cluster positions? Do the same digits always cluster together?</span>
<span id="cb22-466"><a href="#cb22-466" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-467"><a href="#cb22-467" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-468"><a href="#cb22-468" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-469"><a href="#cb22-469" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.5: Computation Time</span></span>
<span id="cb22-470"><a href="#cb22-470" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-471"><a href="#cb22-471" aria-hidden="true" tabindex="-1"></a><span class="ss">5. </span>Time how long PCA, t-SNE, and UMAP take to run on increasing subsets of data (500, 1000, 2000, 5000 points). Plot computation time versus sample size.</span>
<span id="cb22-472"><a href="#cb22-472" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-473"><a href="#cb22-473" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-474"><a href="#cb22-474" aria-hidden="true" tabindex="-1"></a>::: {.callout-note}</span>
<span id="cb22-475"><a href="#cb22-475" aria-hidden="true" tabindex="-1"></a><span class="fu">### Exercise TU.6: PCA Preprocessing</span></span>
<span id="cb22-476"><a href="#cb22-476" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-477"><a href="#cb22-477" aria-hidden="true" tabindex="-1"></a><span class="ss">6. </span>Apply PCA preprocessing (reducing to 50 dimensions) before running t-SNE. Compare the results and computation time to running t-SNE directly on the full data.</span>
<span id="cb22-478"><a href="#cb22-478" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb22-479"><a href="#cb22-479" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-480"><a href="#cb22-480" aria-hidden="true" tabindex="-1"></a><span class="fu">## Summary</span></span>
<span id="cb22-481"><a href="#cb22-481" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-482"><a href="#cb22-482" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Non-linear dimensionality reduction** methods can reveal structure that linear methods like PCA miss</span>
<span id="cb22-483"><a href="#cb22-483" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**t-SNE** is designed for visualization and excels at preserving local neighborhood structure</span>
<span id="cb22-484"><a href="#cb22-484" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>The **perplexity** parameter controls the balance between local and global structure</span>
<span id="cb22-485"><a href="#cb22-485" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Cluster sizes and inter-cluster distances are not meaningful</span>
<span id="cb22-486"><a href="#cb22-486" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Results depend on random initialization—set a seed for reproducibility</span>
<span id="cb22-487"><a href="#cb22-487" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**UMAP** is a newer alternative that is often faster and may preserve more global structure</span>
<span id="cb22-488"><a href="#cb22-488" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Key parameters are **n_neighbors** and **min_dist**</span>
<span id="cb22-489"><a href="#cb22-489" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>Generally faster than t-SNE, especially on large datasets</span>
<span id="cb22-490"><a href="#cb22-490" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**PCA vs t-SNE vs UMAP**:</span>
<span id="cb22-491"><a href="#cb22-491" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>PCA: Fast, interpretable, preserves global variance, works well for linear structure</span>
<span id="cb22-492"><a href="#cb22-492" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>t-SNE: Best for visualizing local cluster structure, computationally expensive</span>
<span id="cb22-493"><a href="#cb22-493" aria-hidden="true" tabindex="-1"></a><span class="ss">  - </span>UMAP: Good balance of speed and quality, preserves some global structure</span>
<span id="cb22-494"><a href="#cb22-494" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**Best practices**: Use PCA preprocessing, try multiple parameter values, validate results, don't over-interpret</span>
<span id="cb22-495"><a href="#cb22-495" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-496"><a href="#cb22-496" aria-hidden="true" tabindex="-1"></a><span class="fu">## Additional Resources</span></span>
<span id="cb22-497"><a href="#cb22-497" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-498"><a href="#cb22-498" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>van der Maaten &amp; Hinton (2008). Visualizing Data using t-SNE. *JMLR*</span>
<span id="cb22-499"><a href="#cb22-499" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>McInnes et al. (2018). UMAP: Uniform Manifold Approximation and Projection. *arXiv*</span>
<span id="cb22-500"><a href="#cb22-500" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>Kobak &amp; Berens (2019). The art of using t-SNE for single-cell transcriptomics. *Nature Communications*</span>
<span id="cb22-501"><a href="#cb22-501" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>@james2023islr - Additional perspectives on dimensionality reduction</span>
</code><button title="Copy to Clipboard" class="code-copy-button" data-in-quarto-modal=""><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
<p>Statistics for Biosciences and Bioengineering</p>
</div>   
    <div class="nav-footer-center">
      &nbsp;
    </div>
    <div class="nav-footer-right">
<p>Built with <a href="https://quarto.org/">Quarto</a></p>
</div>
  </div>
</footer>




</body></html>